{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3c4b39dd",
   "metadata": {},
   "source": [
    "### basic model eval and compare - use static pretrained global embeddings + length + get cv preds\n",
    "* note:\n",
    "\n",
    "* all seqs: filtered for under length 15xx\n",
    "* viruses = Swissprot reviewed. + host is an animal (human, vertebrata, mammal)\n",
    "* human: SWP.\n",
    "\n",
    "* redundnants filtered. May be ur90 (was filtered also to dorp identicals in first 300 positions + 188 alphabets). Human seqs maps to same # of seqs in UR90 as here ,so may be ur90 already.\n",
    "* Train test splits done by UR50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3fc486d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "\n",
    "from sklearn.ensemble import IsolationForest\n",
    "from sklearn.preprocessing  import MinMaxScaler,StandardScaler\n",
    "from sklearn.pipeline import make_pipeline, Pipeline\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.metrics import classification_report, roc_auc_score, log_loss, precision_recall_fscore_support, accuracy_score\n",
    "from sklearn.model_selection import train_test_split, cross_val_predict\n",
    "from sklearn.linear_model import LogisticRegressionCV, LogisticRegression\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.ensemble import RandomForestClassifier,GradientBoostingClassifier   \n",
    "from sklearn.ensemble import HistGradientBoostingClassifier\n",
    "from sklearn.model_selection import StratifiedGroupKFold,GroupKFold\n",
    "from sklearn.compose import make_column_selector as selector\n",
    "from sklearn.feature_selection import chi2, SelectKBest\n",
    "from sklearn.preprocessing import FunctionTransformer\n",
    "from sklearn.compose import ColumnTransformer\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from sklearn.model_selection import cross_validate\n",
    "from sklearn.compose import make_column_selector as selector\n",
    "\n",
    "# from tqdm.autonotebook import tqdm\n",
    "import shap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8ce6a63d",
   "metadata": {},
   "outputs": [],
   "source": [
    "NUMERIC_ONLY = False # use only numbers, inore raw sequence\n",
    "FAST_RUN = False#True\n",
    "# FAST_RUN = True\n",
    "\n",
    "basic_keep_cols_list = ['Length',  'Sequence'] # \"av_mw\" # we exclude av_mw fom the data here, manually on import\n",
    "# 'Mass', \n",
    "SAVE_CV_MISTAKE_PREDS = False#True\n",
    "SAVE_EVAL_RES = False#True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b7899bc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def eval_model(clf,X_train,y_train,groups,numCV=3,X_test=None,y_test=None,return_preds=True,return_test_preds=True):\n",
    "    train_cv_preds = cross_val_predict(estimator=clf, X=X_train, y=y_train,\n",
    "                                       # cv=GroupKFold(n_splits=numCV), # ORIG\n",
    "                                       cv=StratifiedGroupKFold(n_splits=numCV),\n",
    "                                       groups=groups,method=\"predict_proba\",n_jobs=-2)\n",
    "    print(\"CV-Train res\")\n",
    "    print(classification_report(y_train,train_cv_preds[:,1]>=.5))\n",
    "    print(\"AUC\",roc_auc_score(y_train,train_cv_preds[:,1]).round(5))\n",
    "    \n",
    "    if X_test is not None:\n",
    "        clf.fit(X_train,y_train)\n",
    "        y_test_pred = clf.predict_proba(X_test)\n",
    "        print(\"TEST res\")\n",
    "        print(classification_report(y_test,y_test_pred[:,1]>=.5))\n",
    "        print(\"Test AUC\",roc_auc_score(y_test,y_test_pred[:,1]).round(5))\n",
    "    \n",
    "    if return_preds:\n",
    "        if return_test_preds:\n",
    "            return y_test_pred[:,1]\n",
    "        else:\n",
    "            return train_cv_preds[:,1]\n",
    "\n",
    "\n",
    "def get_model_results(model_name:str,y_test,y_pred):\n",
    "    \"\"\"\n",
    "    Note: returns macro averaged prec/recall/f1.\n",
    "    >>> df_res = pd.DataFrame()\n",
    "    >>> df_res = pd.concat([df_res, pd.DataFrame.from_dict(get_model_results(\"Length model\",y_test=y_test,y_pred=length_test_preds),orient=\"index\").T], ignore_index=True)\n",
    "    \"\"\"\n",
    "    # classification_report, roc_auc_score, log_loss, precision_recall_fscore_support, accuracy_score\n",
    "    AUC = roc_auc_score(y_test,y_pred)\n",
    "    # res = classification_report(y_test,y_pred>=0.5,output_dict=True)#.keys()\n",
    "    precision, recall, F1, _ = precision_recall_fscore_support(y_test,y_pred>=0.5,average=\"macro\") # micro used in DL eval\n",
    "    acc = accuracy_score(y_test,y_pred>=0.5)\n",
    "\n",
    "    return {\"model\":model_name,\n",
    "            \"eval_roc_auc\":AUC,\n",
    "    \"eval_accuracy\":acc,\n",
    "    \"eval_precision\":precision, \"eval_recall\":recall,\n",
    "            \"eval_loss\":log_loss(y_test,y_pred),\n",
    "            \"eval_f1\":F1}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3cf822bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20340, 1026)\n",
      "0.279\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>global_embed_0</th>\n",
       "      <th>global_embed_1</th>\n",
       "      <th>global_embed_2</th>\n",
       "      <th>global_embed_3</th>\n",
       "      <th>global_embed_4</th>\n",
       "      <th>global_embed_5</th>\n",
       "      <th>global_embed_6</th>\n",
       "      <th>global_embed_7</th>\n",
       "      <th>global_embed_8</th>\n",
       "      <th>global_embed_9</th>\n",
       "      <th>...</th>\n",
       "      <th>global_embed_1016</th>\n",
       "      <th>global_embed_1017</th>\n",
       "      <th>global_embed_1018</th>\n",
       "      <th>global_embed_1019</th>\n",
       "      <th>global_embed_1020</th>\n",
       "      <th>global_embed_1021</th>\n",
       "      <th>global_embed_1022</th>\n",
       "      <th>global_embed_1023</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.040560</td>\n",
       "      <td>0.012460</td>\n",
       "      <td>0.04860</td>\n",
       "      <td>0.006645</td>\n",
       "      <td>-0.006474</td>\n",
       "      <td>0.027070</td>\n",
       "      <td>0.019240</td>\n",
       "      <td>-0.10620</td>\n",
       "      <td>-0.005110</td>\n",
       "      <td>0.01721</td>\n",
       "      <td>...</td>\n",
       "      <td>0.018650</td>\n",
       "      <td>-0.004280</td>\n",
       "      <td>0.021000</td>\n",
       "      <td>0.002610</td>\n",
       "      <td>-0.004677</td>\n",
       "      <td>0.019590</td>\n",
       "      <td>-0.02592</td>\n",
       "      <td>-0.026840</td>\n",
       "      <td>102</td>\n",
       "      <td>MADFLKGLPVYNKSNFSRFHADSVCKASNRRPSVYLPTREYPSEQI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.007630</td>\n",
       "      <td>-0.004948</td>\n",
       "      <td>0.02852</td>\n",
       "      <td>-0.001821</td>\n",
       "      <td>-0.039600</td>\n",
       "      <td>0.066800</td>\n",
       "      <td>-0.020630</td>\n",
       "      <td>-0.16850</td>\n",
       "      <td>0.003940</td>\n",
       "      <td>-0.02774</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.059500</td>\n",
       "      <td>-0.019410</td>\n",
       "      <td>0.041660</td>\n",
       "      <td>0.047820</td>\n",
       "      <td>-0.034120</td>\n",
       "      <td>0.008250</td>\n",
       "      <td>-0.00011</td>\n",
       "      <td>0.052370</td>\n",
       "      <td>362</td>\n",
       "      <td>MPSTLQVLAKKVLALEHKENDHISREYYYHILKCCGLWWHEAPIIL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.037300</td>\n",
       "      <td>-0.061070</td>\n",
       "      <td>0.08730</td>\n",
       "      <td>0.005596</td>\n",
       "      <td>-0.008150</td>\n",
       "      <td>0.000575</td>\n",
       "      <td>0.011734</td>\n",
       "      <td>-0.16220</td>\n",
       "      <td>-0.013660</td>\n",
       "      <td>-0.02573</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.014534</td>\n",
       "      <td>-0.033780</td>\n",
       "      <td>0.033630</td>\n",
       "      <td>-0.093140</td>\n",
       "      <td>0.093300</td>\n",
       "      <td>0.026840</td>\n",
       "      <td>0.08120</td>\n",
       "      <td>0.024750</td>\n",
       "      <td>97</td>\n",
       "      <td>MASSAELDFNLQALLEQLSQDELSKFKSLIRTISLGKELQTVPQTE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.002218</td>\n",
       "      <td>0.002700</td>\n",
       "      <td>0.07270</td>\n",
       "      <td>-0.022480</td>\n",
       "      <td>0.001690</td>\n",
       "      <td>0.040280</td>\n",
       "      <td>0.001864</td>\n",
       "      <td>-0.17580</td>\n",
       "      <td>0.098200</td>\n",
       "      <td>0.03928</td>\n",
       "      <td>...</td>\n",
       "      <td>0.064760</td>\n",
       "      <td>-0.070200</td>\n",
       "      <td>-0.032100</td>\n",
       "      <td>0.000459</td>\n",
       "      <td>0.019100</td>\n",
       "      <td>0.055500</td>\n",
       "      <td>0.11730</td>\n",
       "      <td>-0.012960</td>\n",
       "      <td>186</td>\n",
       "      <td>MAAWGKKHAGKDPVRDECEERNRFTETREEDVTDEHGEREPFAETD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.069760</td>\n",
       "      <td>0.098600</td>\n",
       "      <td>0.04890</td>\n",
       "      <td>0.004288</td>\n",
       "      <td>0.022430</td>\n",
       "      <td>0.062700</td>\n",
       "      <td>-0.037500</td>\n",
       "      <td>-0.05383</td>\n",
       "      <td>-0.027130</td>\n",
       "      <td>-0.02348</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.058800</td>\n",
       "      <td>0.008675</td>\n",
       "      <td>0.039180</td>\n",
       "      <td>-0.004436</td>\n",
       "      <td>-0.029570</td>\n",
       "      <td>0.034200</td>\n",
       "      <td>-0.01112</td>\n",
       "      <td>-0.009920</td>\n",
       "      <td>1360</td>\n",
       "      <td>MASDSPARSLDEIDLSALRDPAGIFELVELVGNGTYGQVYKGRHVK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20335</th>\n",
       "      <td>-0.041050</td>\n",
       "      <td>-0.032000</td>\n",
       "      <td>0.03632</td>\n",
       "      <td>-0.013990</td>\n",
       "      <td>-0.042300</td>\n",
       "      <td>0.070900</td>\n",
       "      <td>-0.043600</td>\n",
       "      <td>-0.06573</td>\n",
       "      <td>0.044250</td>\n",
       "      <td>0.03620</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008830</td>\n",
       "      <td>-0.029920</td>\n",
       "      <td>-0.006527</td>\n",
       "      <td>0.036900</td>\n",
       "      <td>0.000668</td>\n",
       "      <td>-0.010840</td>\n",
       "      <td>-0.01273</td>\n",
       "      <td>0.018340</td>\n",
       "      <td>578</td>\n",
       "      <td>MDPDKQDALNSIENSIYRTAFKLQSVQTLCQLDLIDSSLIQQVLLR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20336</th>\n",
       "      <td>0.016450</td>\n",
       "      <td>-0.008040</td>\n",
       "      <td>0.04852</td>\n",
       "      <td>0.029430</td>\n",
       "      <td>-0.040300</td>\n",
       "      <td>0.040100</td>\n",
       "      <td>-0.041870</td>\n",
       "      <td>-0.11350</td>\n",
       "      <td>0.047270</td>\n",
       "      <td>0.03058</td>\n",
       "      <td>...</td>\n",
       "      <td>0.033300</td>\n",
       "      <td>-0.091400</td>\n",
       "      <td>0.001120</td>\n",
       "      <td>-0.027710</td>\n",
       "      <td>-0.014870</td>\n",
       "      <td>-0.000851</td>\n",
       "      <td>0.06420</td>\n",
       "      <td>0.044980</td>\n",
       "      <td>348</td>\n",
       "      <td>MLCPWRTANLGLLLILTIFLVAEAEGAAQPNNSLMLQTSKENHALA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20337</th>\n",
       "      <td>-0.072300</td>\n",
       "      <td>0.009674</td>\n",
       "      <td>0.08984</td>\n",
       "      <td>0.002928</td>\n",
       "      <td>-0.052500</td>\n",
       "      <td>0.059480</td>\n",
       "      <td>-0.031600</td>\n",
       "      <td>-0.07390</td>\n",
       "      <td>0.024300</td>\n",
       "      <td>0.03186</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.019200</td>\n",
       "      <td>-0.080140</td>\n",
       "      <td>0.008230</td>\n",
       "      <td>0.033750</td>\n",
       "      <td>-0.027820</td>\n",
       "      <td>0.004436</td>\n",
       "      <td>0.06400</td>\n",
       "      <td>0.001059</td>\n",
       "      <td>421</td>\n",
       "      <td>MCLRFFSPVPGSTSSATNVTMVVSAGPWSSEKAEMNILEINEKLRP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20338</th>\n",
       "      <td>0.009220</td>\n",
       "      <td>-0.034030</td>\n",
       "      <td>0.01591</td>\n",
       "      <td>-0.005898</td>\n",
       "      <td>0.019420</td>\n",
       "      <td>0.008280</td>\n",
       "      <td>-0.054140</td>\n",
       "      <td>-0.07000</td>\n",
       "      <td>0.005203</td>\n",
       "      <td>-0.04065</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023390</td>\n",
       "      <td>-0.068000</td>\n",
       "      <td>-0.036100</td>\n",
       "      <td>-0.056500</td>\n",
       "      <td>-0.011086</td>\n",
       "      <td>-0.031430</td>\n",
       "      <td>0.05466</td>\n",
       "      <td>0.043950</td>\n",
       "      <td>376</td>\n",
       "      <td>MASHAGQQHAPAFGQAARASGPTDGRAASRPSHRQGASEARGDPEL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20339</th>\n",
       "      <td>0.016800</td>\n",
       "      <td>0.052950</td>\n",
       "      <td>0.05692</td>\n",
       "      <td>0.040280</td>\n",
       "      <td>0.007310</td>\n",
       "      <td>0.067300</td>\n",
       "      <td>-0.032870</td>\n",
       "      <td>-0.02719</td>\n",
       "      <td>-0.013374</td>\n",
       "      <td>-0.00674</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.039000</td>\n",
       "      <td>0.028020</td>\n",
       "      <td>0.059170</td>\n",
       "      <td>0.014040</td>\n",
       "      <td>-0.034450</td>\n",
       "      <td>-0.019360</td>\n",
       "      <td>-0.05292</td>\n",
       "      <td>-0.007610</td>\n",
       "      <td>579</td>\n",
       "      <td>MSASSLLEQRPKGQGNKVQNGSVHQKDGLNDDDFEPYLSPQARPNN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20340 rows × 1026 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       global_embed_0  global_embed_1  global_embed_2  global_embed_3  \\\n",
       "0            0.040560        0.012460         0.04860        0.006645   \n",
       "1            0.007630       -0.004948         0.02852       -0.001821   \n",
       "2            0.037300       -0.061070         0.08730        0.005596   \n",
       "3           -0.002218        0.002700         0.07270       -0.022480   \n",
       "4            0.069760        0.098600         0.04890        0.004288   \n",
       "...               ...             ...             ...             ...   \n",
       "20335       -0.041050       -0.032000         0.03632       -0.013990   \n",
       "20336        0.016450       -0.008040         0.04852        0.029430   \n",
       "20337       -0.072300        0.009674         0.08984        0.002928   \n",
       "20338        0.009220       -0.034030         0.01591       -0.005898   \n",
       "20339        0.016800        0.052950         0.05692        0.040280   \n",
       "\n",
       "       global_embed_4  global_embed_5  global_embed_6  global_embed_7  \\\n",
       "0           -0.006474        0.027070        0.019240        -0.10620   \n",
       "1           -0.039600        0.066800       -0.020630        -0.16850   \n",
       "2           -0.008150        0.000575        0.011734        -0.16220   \n",
       "3            0.001690        0.040280        0.001864        -0.17580   \n",
       "4            0.022430        0.062700       -0.037500        -0.05383   \n",
       "...               ...             ...             ...             ...   \n",
       "20335       -0.042300        0.070900       -0.043600        -0.06573   \n",
       "20336       -0.040300        0.040100       -0.041870        -0.11350   \n",
       "20337       -0.052500        0.059480       -0.031600        -0.07390   \n",
       "20338        0.019420        0.008280       -0.054140        -0.07000   \n",
       "20339        0.007310        0.067300       -0.032870        -0.02719   \n",
       "\n",
       "       global_embed_8  global_embed_9  ...  global_embed_1016  \\\n",
       "0           -0.005110         0.01721  ...           0.018650   \n",
       "1            0.003940        -0.02774  ...          -0.059500   \n",
       "2           -0.013660        -0.02573  ...          -0.014534   \n",
       "3            0.098200         0.03928  ...           0.064760   \n",
       "4           -0.027130        -0.02348  ...          -0.058800   \n",
       "...               ...             ...  ...                ...   \n",
       "20335        0.044250         0.03620  ...           0.008830   \n",
       "20336        0.047270         0.03058  ...           0.033300   \n",
       "20337        0.024300         0.03186  ...          -0.019200   \n",
       "20338        0.005203        -0.04065  ...           0.023390   \n",
       "20339       -0.013374        -0.00674  ...          -0.039000   \n",
       "\n",
       "       global_embed_1017  global_embed_1018  global_embed_1019  \\\n",
       "0              -0.004280           0.021000           0.002610   \n",
       "1              -0.019410           0.041660           0.047820   \n",
       "2              -0.033780           0.033630          -0.093140   \n",
       "3              -0.070200          -0.032100           0.000459   \n",
       "4               0.008675           0.039180          -0.004436   \n",
       "...                  ...                ...                ...   \n",
       "20335          -0.029920          -0.006527           0.036900   \n",
       "20336          -0.091400           0.001120          -0.027710   \n",
       "20337          -0.080140           0.008230           0.033750   \n",
       "20338          -0.068000          -0.036100          -0.056500   \n",
       "20339           0.028020           0.059170           0.014040   \n",
       "\n",
       "       global_embed_1020  global_embed_1021  global_embed_1022  \\\n",
       "0              -0.004677           0.019590           -0.02592   \n",
       "1              -0.034120           0.008250           -0.00011   \n",
       "2               0.093300           0.026840            0.08120   \n",
       "3               0.019100           0.055500            0.11730   \n",
       "4              -0.029570           0.034200           -0.01112   \n",
       "...                  ...                ...                ...   \n",
       "20335           0.000668          -0.010840           -0.01273   \n",
       "20336          -0.014870          -0.000851            0.06420   \n",
       "20337          -0.027820           0.004436            0.06400   \n",
       "20338          -0.011086          -0.031430            0.05466   \n",
       "20339          -0.034450          -0.019360           -0.05292   \n",
       "\n",
       "       global_embed_1023  Length  \\\n",
       "0              -0.026840     102   \n",
       "1               0.052370     362   \n",
       "2               0.024750      97   \n",
       "3              -0.012960     186   \n",
       "4              -0.009920    1360   \n",
       "...                  ...     ...   \n",
       "20335           0.018340     578   \n",
       "20336           0.044980     348   \n",
       "20337           0.001059     421   \n",
       "20338           0.043950     376   \n",
       "20339          -0.007610     579   \n",
       "\n",
       "                                                Sequence  \n",
       "0      MADFLKGLPVYNKSNFSRFHADSVCKASNRRPSVYLPTREYPSEQI...  \n",
       "1      MPSTLQVLAKKVLALEHKENDHISREYYYHILKCCGLWWHEAPIIL...  \n",
       "2      MASSAELDFNLQALLEQLSQDELSKFKSLIRTISLGKELQTVPQTE...  \n",
       "3      MAAWGKKHAGKDPVRDECEERNRFTETREEDVTDEHGEREPFAETD...  \n",
       "4      MASDSPARSLDEIDLSALRDPAGIFELVELVGNGTYGQVYKGRHVK...  \n",
       "...                                                  ...  \n",
       "20335  MDPDKQDALNSIENSIYRTAFKLQSVQTLCQLDLIDSSLIQQVLLR...  \n",
       "20336  MLCPWRTANLGLLLILTIFLVAEAEGAAQPNNSLMLQTSKENHALA...  \n",
       "20337  MCLRFFSPVPGSTSSATNVTMVVSAGPWSSEKAEMNILEINEKLRP...  \n",
       "20338  MASHAGQQHAPAFGQAARASGPTDGRAASRPSHRQGASEARGDPEL...  \n",
       "20339  MSASSLLEQRPKGQGNKVQNGSVHQKDGLNDDDFEPYLSPQARPNN...  \n",
       "\n",
       "[20340 rows x 1026 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "X_train = pd.read_csv(\"hum_vir_swp-globalEmbed-train.csv.gz\",nrows= 700 if FAST_RUN else 1e9)\n",
    "# display(X_train.iloc[:,-10:].groupby([\"virus\"]).select_dtypes([\"number\"]).agg([\"mean\",\"max\",\"min\"]).round(2))\n",
    "y_train = X_train.pop(\"virus\")\n",
    "train_groups = X_train[\"Cluster name\"]\n",
    "X_train.drop(columns=['Mass', \"av_mw\"],errors=\"ignore\",inplace=True)\n",
    "keep_feature_cols_list = list(X_train.select_dtypes([\"number\",\"bool\"]).columns)\n",
    "if not NUMERIC_ONLY:\n",
    "    keep_feature_cols_list = keep_feature_cols_list + [\"Sequence\"]\n",
    "    \n",
    "X_train = X_train.filter(keep_feature_cols_list,axis=1)\n",
    "print(X_train.shape)\n",
    "print(y_train.mean().round(3))\n",
    "display(X_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "561b730b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test\n",
      "(4777, 1026)\n",
      "0.215\n"
     ]
    }
   ],
   "source": [
    "print(\"Test\")\n",
    "X_test = pd.read_csv(\"hum_vir_swp-globalEmbed-test.csv.gz\",nrows= 500 if FAST_RUN else 1e9)\n",
    "X_test.drop(columns=['Mass', \"av_mw\"],errors=\"ignore\",inplace=True)\n",
    "test_groups = X_test[\"Cluster name\"] # these aren't used , but nice to have to sanity check\n",
    "X_test.drop(columns=[\"Cluster name\"],errors=\"ignore\",inplace=True)\n",
    "y_test = X_test.pop(\"virus\")\n",
    "\n",
    "X_test = X_test.filter(keep_feature_cols_list,axis=1)\n",
    "\n",
    "print(X_test.shape)\n",
    "print(y_test.mean().round(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9877db69",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25117\n"
     ]
    }
   ],
   "source": [
    "print(X_train.shape[0]+X_test.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "29b4ed46-51b3-4440-b8e1-2199601fb1d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4488\n"
     ]
    }
   ],
   "source": [
    "print(test_groups.nunique())\n",
    "assert len ([i for i in train_groups if i in test_groups])==0 # no overlap between groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "030d2adc-1e08-4e9a-b4e0-cb5803511ef5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Cluster name\n",
       "Cluster: Nucleoprotein                                            20\n",
       "Cluster: Phosphoprotein                                           10\n",
       "Cluster: Envelope glycoprotein N                                   9\n",
       "Cluster: Matrix protein                                            9\n",
       "Cluster: Thymidine kinase                                          8\n",
       "                                                                  ..\n",
       "Cluster: Exosome component 10                                      1\n",
       "Cluster: Putative uncharacterized protein encoded by LINC00476     1\n",
       "Cluster: Protocadherin-9                                           1\n",
       "Cluster: Myc proto-oncogene protein                                1\n",
       "Cluster: Centrosomal protein of 76 kDa                             1\n",
       "Name: count, Length: 4488, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_groups.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "be2dc8a3-a014-4a46-ac6b-87c3ba33c44d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_test.isna().max(axis=0).sum() # no missing values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5ba569ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>global_embed_0</th>\n",
       "      <th>global_embed_1</th>\n",
       "      <th>global_embed_2</th>\n",
       "      <th>global_embed_3</th>\n",
       "      <th>global_embed_4</th>\n",
       "      <th>global_embed_5</th>\n",
       "      <th>global_embed_6</th>\n",
       "      <th>global_embed_7</th>\n",
       "      <th>global_embed_8</th>\n",
       "      <th>global_embed_9</th>\n",
       "      <th>...</th>\n",
       "      <th>global_embed_1016</th>\n",
       "      <th>global_embed_1017</th>\n",
       "      <th>global_embed_1018</th>\n",
       "      <th>global_embed_1019</th>\n",
       "      <th>global_embed_1020</th>\n",
       "      <th>global_embed_1021</th>\n",
       "      <th>global_embed_1022</th>\n",
       "      <th>global_embed_1023</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.025620</td>\n",
       "      <td>-0.002438</td>\n",
       "      <td>0.045720</td>\n",
       "      <td>0.04990</td>\n",
       "      <td>-0.003360</td>\n",
       "      <td>0.001394</td>\n",
       "      <td>-0.055600</td>\n",
       "      <td>-0.08990</td>\n",
       "      <td>0.032170</td>\n",
       "      <td>-0.015594</td>\n",
       "      <td>...</td>\n",
       "      <td>0.01883</td>\n",
       "      <td>-0.045720</td>\n",
       "      <td>-0.018170</td>\n",
       "      <td>-0.019180</td>\n",
       "      <td>-0.007690</td>\n",
       "      <td>0.015720</td>\n",
       "      <td>0.005882</td>\n",
       "      <td>-0.01941</td>\n",
       "      <td>102</td>\n",
       "      <td>MTTATPLGDTTFFSLNMTTRGEDFLYKSSGAIVAAVVVVVIIIFTV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.048920</td>\n",
       "      <td>0.053920</td>\n",
       "      <td>0.037050</td>\n",
       "      <td>0.02147</td>\n",
       "      <td>-0.003021</td>\n",
       "      <td>0.049130</td>\n",
       "      <td>-0.021260</td>\n",
       "      <td>-0.08130</td>\n",
       "      <td>0.018370</td>\n",
       "      <td>-0.018070</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.01804</td>\n",
       "      <td>0.007990</td>\n",
       "      <td>0.061000</td>\n",
       "      <td>-0.004280</td>\n",
       "      <td>-0.029430</td>\n",
       "      <td>0.002352</td>\n",
       "      <td>-0.036800</td>\n",
       "      <td>0.03096</td>\n",
       "      <td>907</td>\n",
       "      <td>MESAIAEGGASRFSASSGGGGSRGAPQHYPKTAGNSEFLGKTPGQN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.038100</td>\n",
       "      <td>0.071840</td>\n",
       "      <td>0.037840</td>\n",
       "      <td>0.02646</td>\n",
       "      <td>-0.009040</td>\n",
       "      <td>0.037170</td>\n",
       "      <td>-0.031280</td>\n",
       "      <td>-0.09094</td>\n",
       "      <td>0.030550</td>\n",
       "      <td>-0.015300</td>\n",
       "      <td>...</td>\n",
       "      <td>0.02495</td>\n",
       "      <td>-0.013080</td>\n",
       "      <td>-0.014190</td>\n",
       "      <td>-0.046360</td>\n",
       "      <td>0.005936</td>\n",
       "      <td>-0.009346</td>\n",
       "      <td>-0.012470</td>\n",
       "      <td>-0.03450</td>\n",
       "      <td>329</td>\n",
       "      <td>MGTAAAAAAAAAAAAAGEGARSPSPAAVSLGLGVAVVSSLVNGSTF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.015030</td>\n",
       "      <td>-0.016920</td>\n",
       "      <td>0.010445</td>\n",
       "      <td>0.01985</td>\n",
       "      <td>-0.016340</td>\n",
       "      <td>0.039060</td>\n",
       "      <td>-0.031070</td>\n",
       "      <td>-0.05377</td>\n",
       "      <td>0.024690</td>\n",
       "      <td>0.009270</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00788</td>\n",
       "      <td>-0.039860</td>\n",
       "      <td>0.047820</td>\n",
       "      <td>0.048770</td>\n",
       "      <td>0.029710</td>\n",
       "      <td>-0.018880</td>\n",
       "      <td>0.059140</td>\n",
       "      <td>0.00568</td>\n",
       "      <td>460</td>\n",
       "      <td>MFTIKLLLFIVPLVISSRIDQDNSSFDSLSPEPKSRFAMLDDVKIL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.000349</td>\n",
       "      <td>0.021680</td>\n",
       "      <td>0.151700</td>\n",
       "      <td>0.02565</td>\n",
       "      <td>-0.020340</td>\n",
       "      <td>0.090100</td>\n",
       "      <td>0.021700</td>\n",
       "      <td>-0.14670</td>\n",
       "      <td>-0.029890</td>\n",
       "      <td>0.134600</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.01367</td>\n",
       "      <td>-0.044130</td>\n",
       "      <td>-0.034180</td>\n",
       "      <td>-0.040900</td>\n",
       "      <td>-0.085140</td>\n",
       "      <td>0.024300</td>\n",
       "      <td>0.023480</td>\n",
       "      <td>0.03925</td>\n",
       "      <td>347</td>\n",
       "      <td>MPRGQKSKLRAREKRRKAREETQGLKVAHATAAEKEECPSSSPVLG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4772</th>\n",
       "      <td>-0.050000</td>\n",
       "      <td>-0.029130</td>\n",
       "      <td>0.075600</td>\n",
       "      <td>0.01599</td>\n",
       "      <td>-0.022140</td>\n",
       "      <td>0.042660</td>\n",
       "      <td>0.026350</td>\n",
       "      <td>-0.07434</td>\n",
       "      <td>-0.000616</td>\n",
       "      <td>0.006010</td>\n",
       "      <td>...</td>\n",
       "      <td>0.03177</td>\n",
       "      <td>-0.040620</td>\n",
       "      <td>-0.041930</td>\n",
       "      <td>0.026470</td>\n",
       "      <td>0.006435</td>\n",
       "      <td>0.059500</td>\n",
       "      <td>-0.001407</td>\n",
       "      <td>0.10913</td>\n",
       "      <td>164</td>\n",
       "      <td>MKWKALFTAAILQAQLPITEAQSFGLLDPKLCYLLDGILFIYGVIL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4773</th>\n",
       "      <td>0.008865</td>\n",
       "      <td>-0.047000</td>\n",
       "      <td>0.015396</td>\n",
       "      <td>0.04404</td>\n",
       "      <td>-0.038450</td>\n",
       "      <td>0.085270</td>\n",
       "      <td>-0.061100</td>\n",
       "      <td>-0.06018</td>\n",
       "      <td>0.085200</td>\n",
       "      <td>0.066100</td>\n",
       "      <td>...</td>\n",
       "      <td>0.01819</td>\n",
       "      <td>-0.001308</td>\n",
       "      <td>-0.001677</td>\n",
       "      <td>-0.011765</td>\n",
       "      <td>-0.010510</td>\n",
       "      <td>0.019680</td>\n",
       "      <td>0.049000</td>\n",
       "      <td>0.05966</td>\n",
       "      <td>1167</td>\n",
       "      <td>MKVNRETKRLYVGGLSQDISEADLQNQFSRFGEVSDVEIITRKDDQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4774</th>\n",
       "      <td>0.042570</td>\n",
       "      <td>0.050170</td>\n",
       "      <td>0.009445</td>\n",
       "      <td>0.02354</td>\n",
       "      <td>0.011020</td>\n",
       "      <td>0.076050</td>\n",
       "      <td>-0.026640</td>\n",
       "      <td>-0.06350</td>\n",
       "      <td>0.017720</td>\n",
       "      <td>0.000740</td>\n",
       "      <td>...</td>\n",
       "      <td>0.02773</td>\n",
       "      <td>0.016720</td>\n",
       "      <td>0.036040</td>\n",
       "      <td>-0.008484</td>\n",
       "      <td>-0.009030</td>\n",
       "      <td>0.015730</td>\n",
       "      <td>0.020370</td>\n",
       "      <td>0.04540</td>\n",
       "      <td>567</td>\n",
       "      <td>MAGVGDAAAPGEGGGGGVDGPQRDGRGEAEQPGGSGGQGPPPAPQL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4775</th>\n",
       "      <td>-0.017070</td>\n",
       "      <td>0.005505</td>\n",
       "      <td>0.044560</td>\n",
       "      <td>-0.01119</td>\n",
       "      <td>-0.005690</td>\n",
       "      <td>0.047180</td>\n",
       "      <td>0.032840</td>\n",
       "      <td>-0.07556</td>\n",
       "      <td>0.018000</td>\n",
       "      <td>-0.002062</td>\n",
       "      <td>...</td>\n",
       "      <td>0.04550</td>\n",
       "      <td>0.009690</td>\n",
       "      <td>-0.003290</td>\n",
       "      <td>-0.053900</td>\n",
       "      <td>0.005430</td>\n",
       "      <td>-0.048580</td>\n",
       "      <td>-0.021910</td>\n",
       "      <td>0.03400</td>\n",
       "      <td>283</td>\n",
       "      <td>MGASVSRGRAARVPAPEPEPEEALDLSQLPPELLLVVLSHVPPRTL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4776</th>\n",
       "      <td>-0.006810</td>\n",
       "      <td>0.000120</td>\n",
       "      <td>0.036560</td>\n",
       "      <td>0.06190</td>\n",
       "      <td>-0.002592</td>\n",
       "      <td>0.030730</td>\n",
       "      <td>-0.001801</td>\n",
       "      <td>-0.06854</td>\n",
       "      <td>0.033450</td>\n",
       "      <td>-0.036040</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.02457</td>\n",
       "      <td>-0.000325</td>\n",
       "      <td>0.028610</td>\n",
       "      <td>0.004750</td>\n",
       "      <td>-0.022050</td>\n",
       "      <td>-0.020160</td>\n",
       "      <td>-0.045560</td>\n",
       "      <td>-0.01116</td>\n",
       "      <td>678</td>\n",
       "      <td>MEVLIGDPITTCLSPSVYDIICNLGFQLRENCDINSIVTQNGEVCW...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4777 rows × 1026 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      global_embed_0  global_embed_1  global_embed_2  global_embed_3  \\\n",
       "0           0.025620       -0.002438        0.045720         0.04990   \n",
       "1           0.048920        0.053920        0.037050         0.02147   \n",
       "2           0.038100        0.071840        0.037840         0.02646   \n",
       "3          -0.015030       -0.016920        0.010445         0.01985   \n",
       "4          -0.000349        0.021680        0.151700         0.02565   \n",
       "...              ...             ...             ...             ...   \n",
       "4772       -0.050000       -0.029130        0.075600         0.01599   \n",
       "4773        0.008865       -0.047000        0.015396         0.04404   \n",
       "4774        0.042570        0.050170        0.009445         0.02354   \n",
       "4775       -0.017070        0.005505        0.044560        -0.01119   \n",
       "4776       -0.006810        0.000120        0.036560         0.06190   \n",
       "\n",
       "      global_embed_4  global_embed_5  global_embed_6  global_embed_7  \\\n",
       "0          -0.003360        0.001394       -0.055600        -0.08990   \n",
       "1          -0.003021        0.049130       -0.021260        -0.08130   \n",
       "2          -0.009040        0.037170       -0.031280        -0.09094   \n",
       "3          -0.016340        0.039060       -0.031070        -0.05377   \n",
       "4          -0.020340        0.090100        0.021700        -0.14670   \n",
       "...              ...             ...             ...             ...   \n",
       "4772       -0.022140        0.042660        0.026350        -0.07434   \n",
       "4773       -0.038450        0.085270       -0.061100        -0.06018   \n",
       "4774        0.011020        0.076050       -0.026640        -0.06350   \n",
       "4775       -0.005690        0.047180        0.032840        -0.07556   \n",
       "4776       -0.002592        0.030730       -0.001801        -0.06854   \n",
       "\n",
       "      global_embed_8  global_embed_9  ...  global_embed_1016  \\\n",
       "0           0.032170       -0.015594  ...            0.01883   \n",
       "1           0.018370       -0.018070  ...           -0.01804   \n",
       "2           0.030550       -0.015300  ...            0.02495   \n",
       "3           0.024690        0.009270  ...            0.00788   \n",
       "4          -0.029890        0.134600  ...           -0.01367   \n",
       "...              ...             ...  ...                ...   \n",
       "4772       -0.000616        0.006010  ...            0.03177   \n",
       "4773        0.085200        0.066100  ...            0.01819   \n",
       "4774        0.017720        0.000740  ...            0.02773   \n",
       "4775        0.018000       -0.002062  ...            0.04550   \n",
       "4776        0.033450       -0.036040  ...           -0.02457   \n",
       "\n",
       "      global_embed_1017  global_embed_1018  global_embed_1019  \\\n",
       "0             -0.045720          -0.018170          -0.019180   \n",
       "1              0.007990           0.061000          -0.004280   \n",
       "2             -0.013080          -0.014190          -0.046360   \n",
       "3             -0.039860           0.047820           0.048770   \n",
       "4             -0.044130          -0.034180          -0.040900   \n",
       "...                 ...                ...                ...   \n",
       "4772          -0.040620          -0.041930           0.026470   \n",
       "4773          -0.001308          -0.001677          -0.011765   \n",
       "4774           0.016720           0.036040          -0.008484   \n",
       "4775           0.009690          -0.003290          -0.053900   \n",
       "4776          -0.000325           0.028610           0.004750   \n",
       "\n",
       "      global_embed_1020  global_embed_1021  global_embed_1022  \\\n",
       "0             -0.007690           0.015720           0.005882   \n",
       "1             -0.029430           0.002352          -0.036800   \n",
       "2              0.005936          -0.009346          -0.012470   \n",
       "3              0.029710          -0.018880           0.059140   \n",
       "4             -0.085140           0.024300           0.023480   \n",
       "...                 ...                ...                ...   \n",
       "4772           0.006435           0.059500          -0.001407   \n",
       "4773          -0.010510           0.019680           0.049000   \n",
       "4774          -0.009030           0.015730           0.020370   \n",
       "4775           0.005430          -0.048580          -0.021910   \n",
       "4776          -0.022050          -0.020160          -0.045560   \n",
       "\n",
       "      global_embed_1023  Length  \\\n",
       "0              -0.01941     102   \n",
       "1               0.03096     907   \n",
       "2              -0.03450     329   \n",
       "3               0.00568     460   \n",
       "4               0.03925     347   \n",
       "...                 ...     ...   \n",
       "4772            0.10913     164   \n",
       "4773            0.05966    1167   \n",
       "4774            0.04540     567   \n",
       "4775            0.03400     283   \n",
       "4776           -0.01116     678   \n",
       "\n",
       "                                               Sequence  \n",
       "0     MTTATPLGDTTFFSLNMTTRGEDFLYKSSGAIVAAVVVVVIIIFTV...  \n",
       "1     MESAIAEGGASRFSASSGGGGSRGAPQHYPKTAGNSEFLGKTPGQN...  \n",
       "2     MGTAAAAAAAAAAAAAGEGARSPSPAAVSLGLGVAVVSSLVNGSTF...  \n",
       "3     MFTIKLLLFIVPLVISSRIDQDNSSFDSLSPEPKSRFAMLDDVKIL...  \n",
       "4     MPRGQKSKLRAREKRRKAREETQGLKVAHATAAEKEECPSSSPVLG...  \n",
       "...                                                 ...  \n",
       "4772  MKWKALFTAAILQAQLPITEAQSFGLLDPKLCYLLDGILFIYGVIL...  \n",
       "4773  MKVNRETKRLYVGGLSQDISEADLQNQFSRFGEVSDVEIITRKDDQ...  \n",
       "4774  MAGVGDAAAPGEGGGGGVDGPQRDGRGEAEQPGGSGGQGPPPAPQL...  \n",
       "4775  MGASVSRGRAARVPAPEPEPEEALDLSQLPPELLLVVLSHVPPRTL...  \n",
       "4776  MEVLIGDPITTCLSPSVYDIICNLGFQLRENCDINSIVTQNGEVCW...  \n",
       "\n",
       "[4777 rows x 1026 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_test"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ddba874-c272-4905-a6b2-1f40076ed079",
   "metadata": {},
   "source": [
    "### store test set results for models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "affd1167",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test_res_dict = {} # store results on test set, per model\n",
    "# \"eval_loss\",\"eval_accuracy\",\"eval_f1\",\"eval_precision\",\"eval_recall\",\"eval_roc_auc\"\n",
    "## store test set results for models. Will add DL models seperately/manually? \n",
    "df_res = pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "393465e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "##### was in \"not NUMERIC_ONLY\" orig\n",
    "# Identify numerical and text features\n",
    "\n",
    "## could replace num with make_column_selector(dtype_include=np.number)  # https://scikit-learn.org/stable/auto_examples/compose/plot_column_transformer_mixed_types.html\n",
    "num_features = [i for i in X_train.columns if i != 'Sequence']\n",
    "text_features = ['Sequence']\n",
    "\n",
    "# For numerical features, we use a simple identity function\n",
    "#     num_transformer = FunctionTransformer(func=lambda x: x, validate=True)\n",
    "num_transformer = StandardScaler()\n",
    "\n",
    "# For the text feature, we use TfidfVectorizer with char-level n-grams\n",
    "text_transformer = Pipeline(steps=[\n",
    "('tfidf',TfidfVectorizer(analyzer='char', ngram_range=(1, 3),\n",
    "                                   min_df=5, max_df=0.97,\n",
    "                                   lowercase=False,\n",
    "                         # max_features=50 if FAST_RUN else 5_000\n",
    "                        )),\n",
    "#     ('chi2', SelectKBest(chi2, k=3000))  # Select 20 best features\n",
    "])\n",
    "#####\n",
    "\n",
    "# LR = LogisticRegressionCV(n_jobs=-1,)#max_iter=200,class_weight=\"balanced\"))\n",
    "\n",
    "# Build ColumnTransformer for preprocessing step of clf\n",
    "preprocessor = ColumnTransformer(transformers=[\n",
    "#     ('num', num_transformer, num_features),# ORIG\n",
    "    (\"num\", num_transformer, selector(dtype_include=\"number\")), # new\n",
    "    ('text', text_transformer, text_features[0])\n",
    "],remainder=\"passthrough\",sparse_threshold=0)\n",
    "\n",
    "\n",
    "if NUMERIC_ONLY:\n",
    "    clf = HistGradientBoostingClassifier()\n",
    "    clf_lin = Pipeline(steps=[\n",
    "        ('preprocessor', StandardScaler()),\n",
    "         # ('classifier',LogisticRegression()#CV(n_jobs=-1)#,max_iter=200,class_weight=\"balanced\")\n",
    "          ('classifier',LogisticRegressionCV(n_jobs=-2)\n",
    "         )])\n",
    "else:  \n",
    "    # Create clf\n",
    "    clf_lin = Pipeline(steps=[\n",
    "        ('preprocessor', preprocessor),\n",
    "         ('classifier',\n",
    "          # LogisticRegression()#CV(n_jobs=-2)\n",
    "          LogisticRegressionCV(n_jobs=-2)\n",
    "         ) \n",
    "    ])\n",
    "\n",
    "    clf = Pipeline(steps=[\n",
    "    ('preprocessor', preprocessor),\n",
    "#         ('classifier', RandomForestClassifier(n_jobs=-1,)#GradientBoostingClassifier() ## gbc is slow \n",
    "     # min_samples_leaf=2,\n",
    "     # hist gbc - doesn't suppot sparse? \n",
    "     ('classifier',HistGradientBoostingClassifier()\n",
    "     )\n",
    "\n",
    "])\n",
    "\n",
    "## basic = for numeric only/length only\n",
    "clf_lin_basic = Pipeline(steps=[\n",
    "        ('preprocessor', StandardScaler()),\n",
    "         ('classifier',LogisticRegressionCV()#CV(n_jobs=-1)#,max_iter=200,class_weight=\"balanced\")\n",
    "         )])\n",
    "\n",
    "## AA text features + any numeric features left in, without scaling them \n",
    "aa_preprocessor = ColumnTransformer(transformers=[\n",
    "#         ('num', num_transformer, num_features),\n",
    "    (\"num\", num_transformer, selector(dtype_include=\"number\")),\n",
    "        ('text', text_transformer, text_features[0])\n",
    "    ],remainder=\"passthrough\",sparse_threshold=0)\n",
    "\n",
    "clf_aa = Pipeline(steps=[\n",
    "    ('preprocessor', aa_preprocessor),\n",
    "     ('classifier',HistGradientBoostingClassifier()\n",
    "     )\n",
    "\n",
    "])\n",
    "\n",
    "\n",
    "# # Fit the model\n",
    "# clf.fit(X_train, y_train)\n",
    "# print(\"Trained\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8a1a10e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%time\n",
    "# ##slow? \n",
    "\n",
    "# ## # Get the feature names extracted by TF-IDF\n",
    "# ## feature_names = num_features + preprocessor.transformers_[1][1].get_feature_names_out()\n",
    "# #ALT: (with FS)# Get the feature names extracted by TF-IDF and chi2\n",
    "# feature_names = num_features + [f'text_{i}' for i in range(text_transformer.named_steps['chi2'].k)]\n",
    "\n",
    "# # Create SHAP explainer\n",
    "# explainer = shap.TreeExplainer(clf.named_steps['classifier'])\n",
    "# print(\"Explainer\")\n",
    "# # Calculate SHAP values\n",
    "# shap_values = explainer.shap_values(preprocessor.transform(X_test.head(55)))\n",
    "\n",
    "# # Summarize the SHAP values in a pandas DataFrame for better visualization\n",
    "# shap_df = pd.DataFrame(shap_values[0], columns=feature_names)\n",
    "\n",
    "# # Print SHAP values\n",
    "# # print(shap_df)\n",
    "\n",
    "# shap_values = explainer(preprocessor.transform(X_test.head(55)),y_test.head(55))\n",
    "\n",
    "# # # visualize the first prediction's explanation\n",
    "# # shap.plots.waterfall(shap_values[0])\n",
    "\n",
    "# # shap.plots.beeswarm(shap_values)\n",
    "# ## hack for random forest:\n",
    "# shap.plots.beeswarm(shap_values[:,:,1]) ## laks column/feature names.."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f9ff752",
   "metadata": {},
   "source": [
    "#### length, AA baseline models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "645c2263",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CV-Train res\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.72      1.00      0.84     14668\n",
      "           1       0.00      0.00      0.00      5672\n",
      "\n",
      "    accuracy                           0.72     20340\n",
      "   macro avg       0.36      0.50      0.42     20340\n",
      "weighted avg       0.52      0.72      0.60     20340\n",
      "\n",
      "AUC 0.60538\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TEST res\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.79      1.00      0.88      3750\n",
      "           1       0.00      0.00      0.00      1027\n",
      "\n",
      "    accuracy                           0.79      4777\n",
      "   macro avg       0.39      0.50      0.44      4777\n",
      "weighted avg       0.62      0.79      0.69      4777\n",
      "\n",
      "Test AUC 0.61975\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n"
     ]
    }
   ],
   "source": [
    "length_test_preds = eval_model(clf=clf_lin_basic,X_train=X_train[\"Length\"].values.reshape(-1, 1),y_train=y_train,groups=train_groups,\n",
    "           numCV=3,X_test=X_test[\"Length\"].values.reshape(-1, 1),y_test=y_test)\n",
    "# length_test_preds = eval_model(clf=clf_lin_basic,X_train=X_train[\"Length\"],y_train=y_train,groups=train_groups,\n",
    "#            numCV=3,X_test=X_test[\"Length\"],y_test=y_test)\n",
    "\n",
    "\n",
    "df_res = pd.concat([df_res, pd.DataFrame.from_dict(get_model_results(\"Length model\",y_test=y_test,\n",
    "                                                                     y_pred=length_test_preds),orient=\"index\").T], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5eb5efe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# clf_aa.fit(X_train[basic_keep_cols_list],y_train) # works in fast run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f11de006",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CV-Train res\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.83      0.96      0.89     14668\n",
      "           1       0.84      0.49      0.62      5672\n",
      "\n",
      "    accuracy                           0.83     20340\n",
      "   macro avg       0.84      0.73      0.76     20340\n",
      "weighted avg       0.83      0.83      0.82     20340\n",
      "\n",
      "AUC 0.88871\n",
      "TEST res\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.90      0.96      0.93      3750\n",
      "           1       0.80      0.60      0.69      1027\n",
      "\n",
      "    accuracy                           0.88      4777\n",
      "   macro avg       0.85      0.78      0.81      4777\n",
      "weighted avg       0.88      0.88      0.88      4777\n",
      "\n",
      "Test AUC 0.91793\n",
      "CPU times: user 5min 16s, sys: 29.5 s, total: 5min 46s\n",
      "Wall time: 2min 50s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "## AA n-grams only model\n",
    "\n",
    "aa_test_cv_preds = eval_model(clf=clf_aa,X_train=X_train[basic_keep_cols_list],y_train=y_train,groups=train_groups,\n",
    "           numCV=2,X_test=X_test[basic_keep_cols_list],y_test=y_test)\n",
    "\n",
    "df_res = pd.concat([df_res,\n",
    "                    pd.DataFrame.from_dict(get_model_results(\"AA model\",y_test=y_test,\n",
    "                                                             y_pred=aa_test_cv_preds),orient=\"index\").T], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fd9395fa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CV-Train res\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.96      0.98      0.97     14668\n",
      "           1       0.95      0.90      0.92      5672\n",
      "\n",
      "    accuracy                           0.96     20340\n",
      "   macro avg       0.95      0.94      0.95     20340\n",
      "weighted avg       0.96      0.96      0.96     20340\n",
      "\n",
      "AUC 0.99102\n",
      "TEST res\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.99      0.98      0.98      3750\n",
      "           1       0.93      0.96      0.94      1027\n",
      "\n",
      "    accuracy                           0.98      4777\n",
      "   macro avg       0.96      0.97      0.96      4777\n",
      "weighted avg       0.98      0.98      0.98      4777\n",
      "\n",
      "Test AUC 0.99559\n",
      "CPU times: user 5min 44s, sys: 33.6 s, total: 6min 18s\n",
      "Wall time: 3min\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "test_cv_preds = eval_model(clf=clf,X_train=X_train,y_train=y_train,groups=train_groups,\n",
    "           numCV=2,X_test=X_test,y_test=y_test)\n",
    "\n",
    "df_res = pd.concat([df_res,\n",
    "                    pd.DataFrame.from_dict(get_model_results(\"Tree-AllFeat\",\n",
    "                                                             y_test=y_test,y_pred=test_cv_preds),orient=\"index\").T], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c7f86bc9-bec6-4a9e-9f67-bb4cf3d8558f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tree-DeepFeat\n",
      "CV-Train res\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.96      0.98      0.97     14668\n",
      "           1       0.95      0.89      0.92      5672\n",
      "\n",
      "    accuracy                           0.96     20340\n",
      "   macro avg       0.95      0.94      0.94     20340\n",
      "weighted avg       0.96      0.96      0.96     20340\n",
      "\n",
      "AUC 0.99093\n",
      "TEST res\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.99      0.98      0.98      3750\n",
      "           1       0.93      0.95      0.94      1027\n",
      "\n",
      "    accuracy                           0.98      4777\n",
      "   macro avg       0.96      0.97      0.96      4777\n",
      "weighted avg       0.98      0.98      0.98      4777\n",
      "\n",
      "Test AUC 0.99601\n",
      "CPU times: user 25.9 s, sys: 431 ms, total: 26.4 s\n",
      "Wall time: 13.6 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "print(\"Tree-DeepFeat\")\n",
    "test_cv_preds =eval_model(clf=HistGradientBoostingClassifier(),\n",
    "                          X_train=X_train.drop(columns=basic_keep_cols_list),y_train=y_train,groups=train_groups,\n",
    "           numCV=2,X_test=X_test.drop(columns=basic_keep_cols_list),y_test=y_test,return_preds=True)\n",
    "\n",
    "df_res = pd.concat([df_res,\n",
    "                    pd.DataFrame.from_dict(get_model_results(\"Tree-DeepFeat\",\n",
    "                                                             y_test=y_test,y_pred=test_cv_preds),orient=\"index\").T], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "53cad186",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CV-Train res\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.97      0.98      0.98     14668\n",
      "           1       0.96      0.93      0.94      5672\n",
      "\n",
      "    accuracy                           0.97     20340\n",
      "   macro avg       0.96      0.96      0.96     20340\n",
      "weighted avg       0.97      0.97      0.97     20340\n",
      "\n",
      "AUC 0.99361\n",
      "TEST res\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.99      0.98      0.98      3750\n",
      "           1       0.94      0.95      0.94      1027\n",
      "\n",
      "    accuracy                           0.98      4777\n",
      "   macro avg       0.96      0.97      0.96      4777\n",
      "weighted avg       0.98      0.98      0.98      4777\n",
      "\n",
      "Test AUC 0.99563\n",
      "CPU times: user 1min 8s, sys: 10.9 s, total: 1min 19s\n",
      "Wall time: 3min 37s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "test_cv_preds_linear =eval_model(clf=clf_lin,X_train=X_train,y_train=y_train,groups=train_groups,\n",
    "           numCV=2,X_test=X_test,y_test=y_test,return_preds=True)\n",
    "\n",
    "df_res = pd.concat([df_res,\n",
    "                    pd.DataFrame.from_dict(get_model_results(\"Linear-AllFeat\",\n",
    "                                                             y_test=y_test,y_pred=test_cv_preds_linear),orient=\"index\").T], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "595f0a8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lin model, without length and similars\n",
      "CV-Train res\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.97      0.98      0.98     14668\n",
      "           1       0.96      0.93      0.94      5672\n",
      "\n",
      "    accuracy                           0.97     20340\n",
      "   macro avg       0.96      0.96      0.96     20340\n",
      "weighted avg       0.97      0.97      0.97     20340\n",
      "\n",
      "AUC 0.9936\n",
      "TEST res\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.99      0.98      0.98      3750\n",
      "           1       0.94      0.95      0.94      1027\n",
      "\n",
      "    accuracy                           0.98      4777\n",
      "   macro avg       0.96      0.97      0.96      4777\n",
      "weighted avg       0.98      0.98      0.98      4777\n",
      "\n",
      "Test AUC 0.99561\n",
      "CPU times: user 2min 53s, sys: 56.7 s, total: 3min 49s\n",
      "Wall time: 23.2 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "print(\"Lin model, without length and similars\")\n",
    "test_cv_preds_linear =eval_model(clf=clf_lin_basic,X_train=X_train.drop(columns=basic_keep_cols_list),y_train=y_train,groups=train_groups,\n",
    "           numCV=2,X_test=X_test.drop(columns=basic_keep_cols_list),y_test=y_test,return_preds=True)\n",
    "\n",
    "df_res = pd.concat([df_res,\n",
    "                    pd.DataFrame.from_dict(get_model_results(\"Linear-DeepFeat\",\n",
    "                                                             y_test=y_test,y_pred=test_cv_preds_linear),orient=\"index\").T], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d2d93b50-9aa5-4358-bd70-1dfb993e4263",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>eval_roc_auc</th>\n",
       "      <th>eval_accuracy</th>\n",
       "      <th>eval_precision</th>\n",
       "      <th>eval_recall</th>\n",
       "      <th>eval_loss</th>\n",
       "      <th>eval_f1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Length model</td>\n",
       "      <td>61.97</td>\n",
       "      <td>78.5</td>\n",
       "      <td>39.25</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.53</td>\n",
       "      <td>43.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AA model</td>\n",
       "      <td>91.79</td>\n",
       "      <td>88.15</td>\n",
       "      <td>84.7</td>\n",
       "      <td>78.07</td>\n",
       "      <td>0.29</td>\n",
       "      <td>80.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tree-AllFeat</td>\n",
       "      <td>99.56</td>\n",
       "      <td>97.59</td>\n",
       "      <td>95.96</td>\n",
       "      <td>97.02</td>\n",
       "      <td>0.07</td>\n",
       "      <td>96.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linear-DeepFeat</td>\n",
       "      <td>99.56</td>\n",
       "      <td>97.57</td>\n",
       "      <td>96.28</td>\n",
       "      <td>96.54</td>\n",
       "      <td>0.06</td>\n",
       "      <td>96.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Linear-AllFeat</td>\n",
       "      <td>99.56</td>\n",
       "      <td>97.57</td>\n",
       "      <td>96.28</td>\n",
       "      <td>96.54</td>\n",
       "      <td>0.06</td>\n",
       "      <td>96.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tree-DeepFeat</td>\n",
       "      <td>99.6</td>\n",
       "      <td>97.57</td>\n",
       "      <td>96.08</td>\n",
       "      <td>96.79</td>\n",
       "      <td>0.06</td>\n",
       "      <td>96.43</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             model  eval_roc_auc  eval_accuracy  eval_precision  eval_recall  \\\n",
       "0     Length model         61.97           78.5           39.25         50.0   \n",
       "1         AA model         91.79          88.15            84.7        78.07   \n",
       "2     Tree-AllFeat         99.56          97.59           95.96        97.02   \n",
       "3  Linear-DeepFeat         99.56          97.57           96.28        96.54   \n",
       "4   Linear-AllFeat         99.56          97.57           96.28        96.54   \n",
       "5    Tree-DeepFeat          99.6          97.57           96.08        96.79   \n",
       "\n",
       "   eval_loss  eval_f1  \n",
       "0       0.53    43.98  \n",
       "1       0.29    80.68  \n",
       "2       0.07    96.47  \n",
       "3       0.06    96.41  \n",
       "4       0.06    96.41  \n",
       "5       0.06    96.43  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_res = 100*df_res.sort_values(\"eval_roc_auc\").convert_dtypes().set_index(\"model\")\n",
    "df_res[\"eval_loss\"] = df_res[\"eval_loss\"]/100\n",
    "df_res =df_res.convert_dtypes().round(2).reset_index()\n",
    "df_res.round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9062c46a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>eval_roc_auc</th>\n",
       "      <th>eval_accuracy</th>\n",
       "      <th>eval_precision</th>\n",
       "      <th>eval_recall</th>\n",
       "      <th>eval_loss</th>\n",
       "      <th>eval_f1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Length model</td>\n",
       "      <td>6197.0</td>\n",
       "      <td>7850.0</td>\n",
       "      <td>3925.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>0.53</td>\n",
       "      <td>4398.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AA model</td>\n",
       "      <td>9179.0</td>\n",
       "      <td>8815.0</td>\n",
       "      <td>8470.0</td>\n",
       "      <td>7807.0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>8068.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tree-AllFeat</td>\n",
       "      <td>9956.0</td>\n",
       "      <td>9759.0</td>\n",
       "      <td>9596.0</td>\n",
       "      <td>9702.0</td>\n",
       "      <td>0.07</td>\n",
       "      <td>9647.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linear-DeepFeat</td>\n",
       "      <td>9956.0</td>\n",
       "      <td>9757.0</td>\n",
       "      <td>9628.0</td>\n",
       "      <td>9654.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>9641.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Linear-AllFeat</td>\n",
       "      <td>9956.0</td>\n",
       "      <td>9757.0</td>\n",
       "      <td>9628.0</td>\n",
       "      <td>9654.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>9641.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tree-DeepFeat</td>\n",
       "      <td>9960.0</td>\n",
       "      <td>9757.0</td>\n",
       "      <td>9608.0</td>\n",
       "      <td>9679.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>9643.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             model  eval_roc_auc  eval_accuracy  eval_precision  eval_recall  \\\n",
       "0     Length model        6197.0         7850.0          3925.0       5000.0   \n",
       "1         AA model        9179.0         8815.0          8470.0       7807.0   \n",
       "2     Tree-AllFeat        9956.0         9759.0          9596.0       9702.0   \n",
       "3  Linear-DeepFeat        9956.0         9757.0          9628.0       9654.0   \n",
       "4   Linear-AllFeat        9956.0         9757.0          9628.0       9654.0   \n",
       "5    Tree-DeepFeat        9960.0         9757.0          9608.0       9679.0   \n",
       "\n",
       "   eval_loss  eval_f1  \n",
       "0       0.53   4398.0  \n",
       "1       0.29   8068.0  \n",
       "2       0.07   9647.0  \n",
       "3       0.06   9641.0  \n",
       "4       0.06   9641.0  \n",
       "5       0.06   9643.0  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_res = 100*df_res.sort_values(\"eval_roc_auc\").convert_dtypes().set_index(\"model\")\n",
    "df_res[\"eval_loss\"] = df_res[\"eval_loss\"]/100\n",
    "df_res =df_res.convert_dtypes().round(2).reset_index()\n",
    "df_res.round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5f4564d7-119d-4cfe-98f7-fc61ea9d12a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>eval_roc_auc</th>\n",
       "      <th>eval_accuracy</th>\n",
       "      <th>eval_precision</th>\n",
       "      <th>eval_recall</th>\n",
       "      <th>eval_loss</th>\n",
       "      <th>eval_f1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Length model</th>\n",
       "      <td>6197.0</td>\n",
       "      <td>7850.0</td>\n",
       "      <td>3925.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>0.53</td>\n",
       "      <td>4398.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA model</th>\n",
       "      <td>9179.0</td>\n",
       "      <td>8815.0</td>\n",
       "      <td>8470.0</td>\n",
       "      <td>7807.0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>8068.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tree-AllFeat</th>\n",
       "      <td>9956.0</td>\n",
       "      <td>9759.0</td>\n",
       "      <td>9596.0</td>\n",
       "      <td>9702.0</td>\n",
       "      <td>0.07</td>\n",
       "      <td>9647.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Linear-DeepFeat</th>\n",
       "      <td>9956.0</td>\n",
       "      <td>9757.0</td>\n",
       "      <td>9628.0</td>\n",
       "      <td>9654.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>9641.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Linear-AllFeat</th>\n",
       "      <td>9956.0</td>\n",
       "      <td>9757.0</td>\n",
       "      <td>9628.0</td>\n",
       "      <td>9654.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>9641.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tree-DeepFeat</th>\n",
       "      <td>9960.0</td>\n",
       "      <td>9757.0</td>\n",
       "      <td>9608.0</td>\n",
       "      <td>9679.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>9643.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 eval_roc_auc  eval_accuracy  eval_precision  eval_recall  \\\n",
       "model                                                                       \n",
       "Length model           6197.0         7850.0          3925.0       5000.0   \n",
       "AA model               9179.0         8815.0          8470.0       7807.0   \n",
       "Tree-AllFeat           9956.0         9759.0          9596.0       9702.0   \n",
       "Linear-DeepFeat        9956.0         9757.0          9628.0       9654.0   \n",
       "Linear-AllFeat         9956.0         9757.0          9628.0       9654.0   \n",
       "Tree-DeepFeat          9960.0         9757.0          9608.0       9679.0   \n",
       "\n",
       "                 eval_loss  eval_f1  \n",
       "model                                \n",
       "Length model          0.53   4398.0  \n",
       "AA model              0.29   8068.0  \n",
       "Tree-AllFeat          0.07   9647.0  \n",
       "Linear-DeepFeat       0.06   9641.0  \n",
       "Linear-AllFeat        0.06   9641.0  \n",
       "Tree-DeepFeat         0.06   9643.0  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_res.convert_dtypes().set_index(\"model\").round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "3d966ad1-9a8f-490c-9b9d-ea8a6bd5eaba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>eval_roc_auc</th>\n",
       "      <th>eval_accuracy</th>\n",
       "      <th>eval_precision</th>\n",
       "      <th>eval_recall</th>\n",
       "      <th>eval_loss</th>\n",
       "      <th>eval_f1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Length model</td>\n",
       "      <td>6197.0</td>\n",
       "      <td>7850.0</td>\n",
       "      <td>3925.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>0.53</td>\n",
       "      <td>4398.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AA model</td>\n",
       "      <td>9179.0</td>\n",
       "      <td>8815.0</td>\n",
       "      <td>8470.0</td>\n",
       "      <td>7807.0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>8068.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Tree-AllFeat</td>\n",
       "      <td>9956.0</td>\n",
       "      <td>9759.0</td>\n",
       "      <td>9596.0</td>\n",
       "      <td>9702.0</td>\n",
       "      <td>0.07</td>\n",
       "      <td>9647.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Linear-DeepFeat</td>\n",
       "      <td>9956.0</td>\n",
       "      <td>9757.0</td>\n",
       "      <td>9628.0</td>\n",
       "      <td>9654.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>9641.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Linear-AllFeat</td>\n",
       "      <td>9956.0</td>\n",
       "      <td>9757.0</td>\n",
       "      <td>9628.0</td>\n",
       "      <td>9654.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>9641.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Tree-DeepFeat</td>\n",
       "      <td>9960.0</td>\n",
       "      <td>9757.0</td>\n",
       "      <td>9608.0</td>\n",
       "      <td>9679.0</td>\n",
       "      <td>0.06</td>\n",
       "      <td>9643.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             model  eval_roc_auc  eval_accuracy  eval_precision  eval_recall  \\\n",
       "0     Length model        6197.0         7850.0          3925.0       5000.0   \n",
       "1         AA model        9179.0         8815.0          8470.0       7807.0   \n",
       "2     Tree-AllFeat        9956.0         9759.0          9596.0       9702.0   \n",
       "3  Linear-DeepFeat        9956.0         9757.0          9628.0       9654.0   \n",
       "4   Linear-AllFeat        9956.0         9757.0          9628.0       9654.0   \n",
       "5    Tree-DeepFeat        9960.0         9757.0          9608.0       9679.0   \n",
       "\n",
       "   eval_loss  eval_f1  \n",
       "0       0.53   4398.0  \n",
       "1       0.29   8068.0  \n",
       "2       0.07   9647.0  \n",
       "3       0.06   9641.0  \n",
       "4       0.06   9641.0  \n",
       "5       0.06   9643.0  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "ee1a39f1-2963-4451-aefc-358d31390da5",
   "metadata": {},
   "outputs": [],
   "source": [
    "if SAVE_EVAL_RES:\n",
    "    df_res.to_csv(\"./outputs/Table-2-modelsEval.csv\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90dd2cc5",
   "metadata": {},
   "source": [
    "* ensembl preds - doesn't improve"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5f5e8ac7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4777,)\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.99      0.98      0.99      3750\n",
      "           1       0.94      0.96      0.95      1027\n",
      "\n",
      "    accuracy                           0.98      4777\n",
      "   macro avg       0.96      0.97      0.97      4777\n",
      "weighted avg       0.98      0.98      0.98      4777\n",
      "\n",
      "AUC 0.9964\n"
     ]
    }
   ],
   "source": [
    "## similar or slightly better perf - negligible difference\n",
    "joint_preds = (test_cv_preds_linear + test_cv_preds)/2\n",
    "print(joint_preds.shape)\n",
    "print(classification_report(y_test,joint_preds>=.5))\n",
    "print(\"AUC\",roc_auc_score(y_test,joint_preds).round(4))\n",
    "\n",
    "# ## overwrite old preds\n",
    "# train_cv_preds = joint_preds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f78d67a",
   "metadata": {},
   "source": [
    "## CV on all - get output preds + erros\n",
    "### predictions error analysis\n",
    "\n",
    "* Q: Should we use best model (Linea + DL feats + Length ; or model without the added length feat (i.e like ap roper DL only model)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "32cfe293",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(25117, 1026)\n",
      "(25117,)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>global_embed_0</th>\n",
       "      <th>global_embed_1</th>\n",
       "      <th>global_embed_2</th>\n",
       "      <th>global_embed_3</th>\n",
       "      <th>global_embed_4</th>\n",
       "      <th>global_embed_5</th>\n",
       "      <th>global_embed_6</th>\n",
       "      <th>global_embed_7</th>\n",
       "      <th>global_embed_8</th>\n",
       "      <th>global_embed_9</th>\n",
       "      <th>...</th>\n",
       "      <th>global_embed_1016</th>\n",
       "      <th>global_embed_1017</th>\n",
       "      <th>global_embed_1018</th>\n",
       "      <th>global_embed_1019</th>\n",
       "      <th>global_embed_1020</th>\n",
       "      <th>global_embed_1021</th>\n",
       "      <th>global_embed_1022</th>\n",
       "      <th>global_embed_1023</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.002075</td>\n",
       "      <td>-0.013176</td>\n",
       "      <td>0.02216</td>\n",
       "      <td>0.013760</td>\n",
       "      <td>-0.047940</td>\n",
       "      <td>0.03510</td>\n",
       "      <td>-0.005756</td>\n",
       "      <td>-0.05280</td>\n",
       "      <td>0.04092</td>\n",
       "      <td>0.036530</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014010</td>\n",
       "      <td>-0.044340</td>\n",
       "      <td>0.004120</td>\n",
       "      <td>-0.003265</td>\n",
       "      <td>0.023090</td>\n",
       "      <td>0.027570</td>\n",
       "      <td>0.04825</td>\n",
       "      <td>0.021650</td>\n",
       "      <td>832</td>\n",
       "      <td>MAMRSGRHPSLLLLLVLLLWLLQVSIIDSVQQETDDLTKQTKEKIY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.033000</td>\n",
       "      <td>0.128700</td>\n",
       "      <td>0.02954</td>\n",
       "      <td>0.025380</td>\n",
       "      <td>0.049900</td>\n",
       "      <td>-0.10693</td>\n",
       "      <td>-0.012290</td>\n",
       "      <td>-0.06854</td>\n",
       "      <td>-0.01787</td>\n",
       "      <td>-0.010735</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.021700</td>\n",
       "      <td>-0.104900</td>\n",
       "      <td>0.129800</td>\n",
       "      <td>0.046260</td>\n",
       "      <td>0.030100</td>\n",
       "      <td>-0.003197</td>\n",
       "      <td>0.14530</td>\n",
       "      <td>0.070740</td>\n",
       "      <td>819</td>\n",
       "      <td>MMNLEKNFDKTTLFNHMRTDKRGKCSDLNEYGTSCDKTTAVEYNKV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.002344</td>\n",
       "      <td>0.018860</td>\n",
       "      <td>-0.03217</td>\n",
       "      <td>-0.009346</td>\n",
       "      <td>0.048770</td>\n",
       "      <td>0.02544</td>\n",
       "      <td>-0.026600</td>\n",
       "      <td>-0.02580</td>\n",
       "      <td>0.02396</td>\n",
       "      <td>-0.044280</td>\n",
       "      <td>...</td>\n",
       "      <td>0.073700</td>\n",
       "      <td>-0.072400</td>\n",
       "      <td>0.076000</td>\n",
       "      <td>0.076000</td>\n",
       "      <td>-0.024900</td>\n",
       "      <td>0.013810</td>\n",
       "      <td>-0.08330</td>\n",
       "      <td>-0.014710</td>\n",
       "      <td>486</td>\n",
       "      <td>MATFKDACYHYKRINKLNHTVLKLGVNDTWRSSPPTKYKGWCLDCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.014910</td>\n",
       "      <td>0.012170</td>\n",
       "      <td>0.05118</td>\n",
       "      <td>0.033970</td>\n",
       "      <td>-0.046660</td>\n",
       "      <td>0.06740</td>\n",
       "      <td>0.004925</td>\n",
       "      <td>-0.06200</td>\n",
       "      <td>0.04470</td>\n",
       "      <td>-0.041320</td>\n",
       "      <td>...</td>\n",
       "      <td>0.022250</td>\n",
       "      <td>-0.098270</td>\n",
       "      <td>0.004417</td>\n",
       "      <td>0.058600</td>\n",
       "      <td>0.001499</td>\n",
       "      <td>0.032070</td>\n",
       "      <td>-0.01502</td>\n",
       "      <td>0.002170</td>\n",
       "      <td>233</td>\n",
       "      <td>MSAKRAELKKTHLSKNYKAVCLELKPEPTKTFDYKAVKQEGRFTKA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.010605</td>\n",
       "      <td>0.080100</td>\n",
       "      <td>0.07580</td>\n",
       "      <td>0.011850</td>\n",
       "      <td>-0.018720</td>\n",
       "      <td>0.02084</td>\n",
       "      <td>-0.020360</td>\n",
       "      <td>-0.05804</td>\n",
       "      <td>0.06890</td>\n",
       "      <td>0.039030</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012240</td>\n",
       "      <td>0.009476</td>\n",
       "      <td>0.019840</td>\n",
       "      <td>-0.055540</td>\n",
       "      <td>-0.000411</td>\n",
       "      <td>0.003155</td>\n",
       "      <td>0.01721</td>\n",
       "      <td>-0.025160</td>\n",
       "      <td>1049</td>\n",
       "      <td>MGSRTPESPLHAVQLRWGPRRRPPLLPLLLLLLPPPPRVGGFNLDA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4772</th>\n",
       "      <td>0.024860</td>\n",
       "      <td>0.077940</td>\n",
       "      <td>0.02074</td>\n",
       "      <td>0.019450</td>\n",
       "      <td>-0.041080</td>\n",
       "      <td>0.03412</td>\n",
       "      <td>-0.033700</td>\n",
       "      <td>-0.12630</td>\n",
       "      <td>0.02231</td>\n",
       "      <td>0.026170</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.005222</td>\n",
       "      <td>-0.043330</td>\n",
       "      <td>0.040500</td>\n",
       "      <td>0.023210</td>\n",
       "      <td>0.005642</td>\n",
       "      <td>0.005108</td>\n",
       "      <td>0.01581</td>\n",
       "      <td>0.005238</td>\n",
       "      <td>541</td>\n",
       "      <td>MKSYTPYFILLWSAVGIAKAAKIIIVPPIMFESHMYIFKTLASALH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4773</th>\n",
       "      <td>0.045750</td>\n",
       "      <td>0.010086</td>\n",
       "      <td>-0.00791</td>\n",
       "      <td>0.024440</td>\n",
       "      <td>-0.043180</td>\n",
       "      <td>0.07916</td>\n",
       "      <td>0.003970</td>\n",
       "      <td>-0.03190</td>\n",
       "      <td>0.04657</td>\n",
       "      <td>0.005486</td>\n",
       "      <td>...</td>\n",
       "      <td>0.053960</td>\n",
       "      <td>0.008720</td>\n",
       "      <td>0.019260</td>\n",
       "      <td>-0.018370</td>\n",
       "      <td>-0.044070</td>\n",
       "      <td>-0.006134</td>\n",
       "      <td>0.02430</td>\n",
       "      <td>0.023700</td>\n",
       "      <td>238</td>\n",
       "      <td>MAARTGHTALRRVVSGCRPKSATAAGAQAPVRNGRYLASCGILMSR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4774</th>\n",
       "      <td>0.003473</td>\n",
       "      <td>0.036130</td>\n",
       "      <td>0.06660</td>\n",
       "      <td>0.022400</td>\n",
       "      <td>0.001187</td>\n",
       "      <td>0.00559</td>\n",
       "      <td>-0.039180</td>\n",
       "      <td>-0.05566</td>\n",
       "      <td>-0.02106</td>\n",
       "      <td>-0.003777</td>\n",
       "      <td>...</td>\n",
       "      <td>0.026120</td>\n",
       "      <td>-0.018050</td>\n",
       "      <td>0.030760</td>\n",
       "      <td>-0.022720</td>\n",
       "      <td>-0.030470</td>\n",
       "      <td>-0.053680</td>\n",
       "      <td>-0.00893</td>\n",
       "      <td>0.011220</td>\n",
       "      <td>384</td>\n",
       "      <td>MAAVQMDPELAKRLFFEGATVVILNMPKGTEFGIDYNSWEVGPKFR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4775</th>\n",
       "      <td>0.038540</td>\n",
       "      <td>0.054930</td>\n",
       "      <td>0.01980</td>\n",
       "      <td>0.018390</td>\n",
       "      <td>-0.040560</td>\n",
       "      <td>0.02724</td>\n",
       "      <td>0.000665</td>\n",
       "      <td>-0.08060</td>\n",
       "      <td>0.00828</td>\n",
       "      <td>-0.015720</td>\n",
       "      <td>...</td>\n",
       "      <td>0.012860</td>\n",
       "      <td>-0.069030</td>\n",
       "      <td>0.018260</td>\n",
       "      <td>0.030650</td>\n",
       "      <td>0.009010</td>\n",
       "      <td>0.002167</td>\n",
       "      <td>0.02405</td>\n",
       "      <td>0.000231</td>\n",
       "      <td>235</td>\n",
       "      <td>MRDLLQYIACFFAFFSAGFLIVATWTDCWMVNADDSLEVSTKCRGL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4776</th>\n",
       "      <td>-0.107900</td>\n",
       "      <td>0.127900</td>\n",
       "      <td>0.03094</td>\n",
       "      <td>-0.003967</td>\n",
       "      <td>0.002113</td>\n",
       "      <td>-0.17810</td>\n",
       "      <td>-0.058780</td>\n",
       "      <td>-0.17490</td>\n",
       "      <td>-0.05505</td>\n",
       "      <td>-0.111700</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.144300</td>\n",
       "      <td>-0.026960</td>\n",
       "      <td>-0.030530</td>\n",
       "      <td>0.107300</td>\n",
       "      <td>0.073400</td>\n",
       "      <td>0.043030</td>\n",
       "      <td>0.13810</td>\n",
       "      <td>-0.044950</td>\n",
       "      <td>254</td>\n",
       "      <td>MGAAGSDGRCCVRSGRAGTGGAGSKWVVMDLWTGGAGSRRAVLGPD...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25117 rows × 1026 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      global_embed_0  global_embed_1  global_embed_2  global_embed_3  \\\n",
       "0          -0.002075       -0.013176         0.02216        0.013760   \n",
       "1          -0.033000        0.128700         0.02954        0.025380   \n",
       "2          -0.002344        0.018860        -0.03217       -0.009346   \n",
       "3           0.014910        0.012170         0.05118        0.033970   \n",
       "4           0.010605        0.080100         0.07580        0.011850   \n",
       "...              ...             ...             ...             ...   \n",
       "4772        0.024860        0.077940         0.02074        0.019450   \n",
       "4773        0.045750        0.010086        -0.00791        0.024440   \n",
       "4774        0.003473        0.036130         0.06660        0.022400   \n",
       "4775        0.038540        0.054930         0.01980        0.018390   \n",
       "4776       -0.107900        0.127900         0.03094       -0.003967   \n",
       "\n",
       "      global_embed_4  global_embed_5  global_embed_6  global_embed_7  \\\n",
       "0          -0.047940         0.03510       -0.005756        -0.05280   \n",
       "1           0.049900        -0.10693       -0.012290        -0.06854   \n",
       "2           0.048770         0.02544       -0.026600        -0.02580   \n",
       "3          -0.046660         0.06740        0.004925        -0.06200   \n",
       "4          -0.018720         0.02084       -0.020360        -0.05804   \n",
       "...              ...             ...             ...             ...   \n",
       "4772       -0.041080         0.03412       -0.033700        -0.12630   \n",
       "4773       -0.043180         0.07916        0.003970        -0.03190   \n",
       "4774        0.001187         0.00559       -0.039180        -0.05566   \n",
       "4775       -0.040560         0.02724        0.000665        -0.08060   \n",
       "4776        0.002113        -0.17810       -0.058780        -0.17490   \n",
       "\n",
       "      global_embed_8  global_embed_9  ...  global_embed_1016  \\\n",
       "0            0.04092        0.036530  ...           0.014010   \n",
       "1           -0.01787       -0.010735  ...          -0.021700   \n",
       "2            0.02396       -0.044280  ...           0.073700   \n",
       "3            0.04470       -0.041320  ...           0.022250   \n",
       "4            0.06890        0.039030  ...           0.012240   \n",
       "...              ...             ...  ...                ...   \n",
       "4772         0.02231        0.026170  ...          -0.005222   \n",
       "4773         0.04657        0.005486  ...           0.053960   \n",
       "4774        -0.02106       -0.003777  ...           0.026120   \n",
       "4775         0.00828       -0.015720  ...           0.012860   \n",
       "4776        -0.05505       -0.111700  ...          -0.144300   \n",
       "\n",
       "      global_embed_1017  global_embed_1018  global_embed_1019  \\\n",
       "0             -0.044340           0.004120          -0.003265   \n",
       "1             -0.104900           0.129800           0.046260   \n",
       "2             -0.072400           0.076000           0.076000   \n",
       "3             -0.098270           0.004417           0.058600   \n",
       "4              0.009476           0.019840          -0.055540   \n",
       "...                 ...                ...                ...   \n",
       "4772          -0.043330           0.040500           0.023210   \n",
       "4773           0.008720           0.019260          -0.018370   \n",
       "4774          -0.018050           0.030760          -0.022720   \n",
       "4775          -0.069030           0.018260           0.030650   \n",
       "4776          -0.026960          -0.030530           0.107300   \n",
       "\n",
       "      global_embed_1020  global_embed_1021  global_embed_1022  \\\n",
       "0              0.023090           0.027570            0.04825   \n",
       "1              0.030100          -0.003197            0.14530   \n",
       "2             -0.024900           0.013810           -0.08330   \n",
       "3              0.001499           0.032070           -0.01502   \n",
       "4             -0.000411           0.003155            0.01721   \n",
       "...                 ...                ...                ...   \n",
       "4772           0.005642           0.005108            0.01581   \n",
       "4773          -0.044070          -0.006134            0.02430   \n",
       "4774          -0.030470          -0.053680           -0.00893   \n",
       "4775           0.009010           0.002167            0.02405   \n",
       "4776           0.073400           0.043030            0.13810   \n",
       "\n",
       "      global_embed_1023  Length  \\\n",
       "0              0.021650     832   \n",
       "1              0.070740     819   \n",
       "2             -0.014710     486   \n",
       "3              0.002170     233   \n",
       "4             -0.025160    1049   \n",
       "...                 ...     ...   \n",
       "4772           0.005238     541   \n",
       "4773           0.023700     238   \n",
       "4774           0.011220     384   \n",
       "4775           0.000231     235   \n",
       "4776          -0.044950     254   \n",
       "\n",
       "                                               Sequence  \n",
       "0     MAMRSGRHPSLLLLLVLLLWLLQVSIIDSVQQETDDLTKQTKEKIY...  \n",
       "1     MMNLEKNFDKTTLFNHMRTDKRGKCSDLNEYGTSCDKTTAVEYNKV...  \n",
       "2     MATFKDACYHYKRINKLNHTVLKLGVNDTWRSSPPTKYKGWCLDCC...  \n",
       "3     MSAKRAELKKTHLSKNYKAVCLELKPEPTKTFDYKAVKQEGRFTKA...  \n",
       "4     MGSRTPESPLHAVQLRWGPRRRPPLLPLLLLLLPPPPRVGGFNLDA...  \n",
       "...                                                 ...  \n",
       "4772  MKSYTPYFILLWSAVGIAKAAKIIIVPPIMFESHMYIFKTLASALH...  \n",
       "4773  MAARTGHTALRRVVSGCRPKSATAAGAQAPVRNGRYLASCGILMSR...  \n",
       "4774  MAAVQMDPELAKRLFFEGATVVILNMPKGTEFGIDYNSWEVGPKFR...  \n",
       "4775  MRDLLQYIACFFAFFSAGFLIVATWTDCWMVNADDSLEVSTKCRGL...  \n",
       "4776  MGAAGSDGRCCVRSGRAGTGGAGSKWVVMDLWTGGAGSRRAVLGPD...  \n",
       "\n",
       "[25117 rows x 1026 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X = pd.concat([X_train,X_test])\n",
    "print(X.shape)\n",
    "y = pd.concat([y_train,y_test])\n",
    "print(y.shape)\n",
    "groups = pd.concat([train_groups,test_groups])\n",
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d4f27a19",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CV-Train res\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.98      0.98      0.98     18418\n",
      "           1       0.96      0.94      0.95      6699\n",
      "\n",
      "    accuracy                           0.97     25117\n",
      "   macro avg       0.97      0.96      0.97     25117\n",
      "weighted avg       0.97      0.97      0.97     25117\n",
      "\n",
      "AUC 0.99424\n",
      "CPU times: user 6.21 s, sys: 325 ms, total: 6.54 s\n",
      "Wall time: 1min 33s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([5.59194262e-04, 8.64401986e-07, 9.99998690e-01, ...,\n",
       "       2.81372484e-03, 5.38118073e-03, 2.52926900e-02])"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "all_preds = eval_model(clf=clf_lin_basic,X_train=X.drop(columns=[\"Sequence\"]) #.set_index(\"Sequence\")\n",
    "                       ,y_train=y,groups=groups,\n",
    "           numCV=4,return_preds=True,return_test_preds=False)\n",
    "all_preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e14a15a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.9942415023913257"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "roc_auc_score(y,all_preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "904bfa19",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mistakes/mismatches\n",
      "sum     682.000000\n",
      "mean      0.027153\n",
      "Name: mistake, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "df_preds = X.filter(X.select_dtypes([\"O\"]).columns.to_list() + [\"Length\",\"virus\"],axis=1).copy()\n",
    "df_preds[\"virus\"] = y\n",
    "df_preds[\"model_pred_proba_vir\"] = all_preds.round(4)\n",
    "df_preds[\"model_pred_vir\"] = (all_preds>=.5).astype(int)\n",
    "print(\"mistakes/mismatches\")\n",
    "df_preds[\"mistake\"] = (df_preds[\"model_pred_vir\"]!=df_preds[\"virus\"])\n",
    "\n",
    "# ## keep only mismatch cases\n",
    "# df_preds = df_preds.loc[df_preds[\"model_pred\"]!=df_preds[\"virus\"]].sort_values([\"model_pred_proba\"])\n",
    "print(df_preds[\"mistake\"].agg([\"sum\",\"mean\"]))\n",
    "# df_preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "297f95e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(25119, 16)\n",
      "nunique:\n",
      "Sequence              25119\n",
      "Baltimore                 6\n",
      "Family                   50\n",
      "Genome Composition        9\n",
      "Genus                   171\n",
      "Keywords              15980\n",
      "Mass                  22130\n",
      "Organism               1136\n",
      "Protein names         21394\n",
      "Taxonomic lineage       561\n",
      "UR50_Cluster ID       20886\n",
      "Virus hosts             291\n",
      "av_mw                  2110\n",
      "genus_human_host          2\n",
      "group                 23473\n",
      "human_host                2\n",
      "dtype: int64\n",
      "count    25119.0\n",
      "mean       446.7\n",
      "std        307.8\n",
      "min         11.0\n",
      "25%        213.0\n",
      "50%        371.0\n",
      "75%        592.0\n",
      "max       1534.0\n",
      "Name: Sequence, dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Baltimore</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genome Composition</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Keywords</th>\n",
       "      <th>Mass</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Taxonomic lineage</th>\n",
       "      <th>UR50_Cluster ID</th>\n",
       "      <th>Virus hosts</th>\n",
       "      <th>av_mw</th>\n",
       "      <th>genus_human_host</th>\n",
       "      <th>group</th>\n",
       "      <th>human_host</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AARADGTMGFSSELCSPQGHGAEQQMQEAELRLLEGMRKWMAQRVK...</td>\n",
       "      <td>ssRNA-RT</td>\n",
       "      <td>Retroviridae</td>\n",
       "      <td>ssRNA-RT</td>\n",
       "      <td>Gammaretrovirus</td>\n",
       "      <td>ATP-binding . Kinase . Nucleotide-binding . On...</td>\n",
       "      <td>68769</td>\n",
       "      <td>Feline sarcoma virus (strain Gardner-Arnstein)...</td>\n",
       "      <td>Tyrosine-protein kinase transforming protein F...</td>\n",
       "      <td>Viruses, Riboviria, Pararnavirae (kingdom), Ar...</td>\n",
       "      <td>P00542</td>\n",
       "      <td>Felidae (cat family) [TaxID: 9681]</td>\n",
       "      <td>112.92</td>\n",
       "      <td>False</td>\n",
       "      <td>P00542</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACSTPHYDLPLICAATWTATCVPRQTSAARCGPCPRHDPHPSAPLP...</td>\n",
       "      <td>dsDNA</td>\n",
       "      <td>Herpesviridae</td>\n",
       "      <td>dsDNA</td>\n",
       "      <td>Cytomegalovirus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12796</td>\n",
       "      <td>Human cytomegalovirus (strain AD169) (HHV-5) (...</td>\n",
       "      <td>Uncharacterized protein UL81</td>\n",
       "      <td>Viruses, Duplodnaviria, Heunggongvirae (kingdo...</td>\n",
       "      <td>P16825</td>\n",
       "      <td>Homo sapiens (Human) [TaxID: 9606]</td>\n",
       "      <td>110.31</td>\n",
       "      <td>True</td>\n",
       "      <td>P16825</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            Sequence Baltimore         Family  \\\n",
       "0  AARADGTMGFSSELCSPQGHGAEQQMQEAELRLLEGMRKWMAQRVK...  ssRNA-RT   Retroviridae   \n",
       "1  ACSTPHYDLPLICAATWTATCVPRQTSAARCGPCPRHDPHPSAPLP...     dsDNA  Herpesviridae   \n",
       "\n",
       "  Genome Composition            Genus  \\\n",
       "0           ssRNA-RT  Gammaretrovirus   \n",
       "1              dsDNA  Cytomegalovirus   \n",
       "\n",
       "                                            Keywords   Mass  \\\n",
       "0  ATP-binding . Kinase . Nucleotide-binding . On...  68769   \n",
       "1                                                NaN  12796   \n",
       "\n",
       "                                            Organism  \\\n",
       "0  Feline sarcoma virus (strain Gardner-Arnstein)...   \n",
       "1  Human cytomegalovirus (strain AD169) (HHV-5) (...   \n",
       "\n",
       "                                       Protein names  \\\n",
       "0  Tyrosine-protein kinase transforming protein F...   \n",
       "1                       Uncharacterized protein UL81   \n",
       "\n",
       "                                   Taxonomic lineage UR50_Cluster ID  \\\n",
       "0  Viruses, Riboviria, Pararnavirae (kingdom), Ar...          P00542   \n",
       "1  Viruses, Duplodnaviria, Heunggongvirae (kingdo...          P16825   \n",
       "\n",
       "                          Virus hosts   av_mw genus_human_host   group  \\\n",
       "0  Felidae (cat family) [TaxID: 9681]  112.92            False  P00542   \n",
       "1  Homo sapiens (Human) [TaxID: 9606]  110.31             True  P16825   \n",
       "\n",
       "  human_host  \n",
       "0      False  \n",
       "1       True  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv(\"SWP_humVir_meta_v2.csv.gz\")\n",
    "print(df_meta.shape)\n",
    "print(\"nunique:\")\n",
    "print(df_meta.nunique())\n",
    "print(df_meta.Sequence.str.len().describe().round(1))\n",
    "df_meta.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "cd813eaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Length</th>\n",
       "      <th>virus</th>\n",
       "      <th>model_pred_proba_vir</th>\n",
       "      <th>model_pred_vir</th>\n",
       "      <th>mistake</th>\n",
       "      <th>Baltimore</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genome Composition</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Keywords</th>\n",
       "      <th>Mass</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Taxonomic lineage</th>\n",
       "      <th>UR50_Cluster ID</th>\n",
       "      <th>Virus hosts</th>\n",
       "      <th>av_mw</th>\n",
       "      <th>genus_human_host</th>\n",
       "      <th>human_host</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MAMRSGRHPSLLLLLVLLLWLLQVSIIDSVQQETDDLTKQTKEKIY...</td>\n",
       "      <td>832</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0006</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alternative splicing . Calcium . Cell adhesion...</td>\n",
       "      <td>92416</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Cadherin-like protein 26 (Cadherin-like protei...</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>Q8IXH8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111.08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MMNLEKNFDKTTLFNHMRTDKRGKCSDLNEYGTSCDKTTAVEYNKV...</td>\n",
       "      <td>819</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DNA-binding . Metal-binding . Nucleus . Refere...</td>\n",
       "      <td>94331</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Zinc finger protein 658B</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>Q4V348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>115.18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MATFKDACYHYKRINKLNHTVLKLGVNDTWRSSPPTKYKGWCLDCC...</td>\n",
       "      <td>486</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0000</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>dsRNA</td>\n",
       "      <td>Sedoreoviridae</td>\n",
       "      <td>dsRNA</td>\n",
       "      <td>Rotavirus</td>\n",
       "      <td>Host cytoplasm . Host cytoskeleton . Host-viru...</td>\n",
       "      <td>57672</td>\n",
       "      <td>Rotavirus A (strain RVA/Human/United States/P/...</td>\n",
       "      <td>Non-structural protein 1 (NSP1) (NCVP2) (Non-s...</td>\n",
       "      <td>Viruses, Riboviria, Orthornavirae (kingdom), D...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Homo sapiens (Human) [TaxID: 9606]</td>\n",
       "      <td>118.67</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSAKRAELKKTHLSKNYKAVCLELKPEPTKTFDYKAVKQEGRFTKA...</td>\n",
       "      <td>233</td>\n",
       "      <td>0</td>\n",
       "      <td>0.2340</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alternative splicing . Coiled coil . Reference...</td>\n",
       "      <td>26518</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Testis-expressed protein 35</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>Q5T0J7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113.81</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGSRTPESPLHAVQLRWGPRRRPPLLPLLLLLLPPPPRVGGFNLDA...</td>\n",
       "      <td>1049</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3D-structure . Calcium . Cell adhesion . Cell ...</td>\n",
       "      <td>114536</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Integrin alpha-5 (CD49 antigen-like family mem...</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>P08648</td>\n",
       "      <td>NaN</td>\n",
       "      <td>109.19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25112</th>\n",
       "      <td>MKSYTPYFILLWSAVGIAKAAKIIIVPPIMFESHMYIFKTLASALH...</td>\n",
       "      <td>541</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Endoplasmic reticulum . Glycoprotein . Glycosy...</td>\n",
       "      <td>61438</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>2-hydroxyacylsphingosine 1-beta-galactosyltran...</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>Q16880</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113.56</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25113</th>\n",
       "      <td>MAARTGHTALRRVVSGCRPKSATAAGAQAPVRNGRYLASCGILMSR...</td>\n",
       "      <td>238</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0001</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alternative splicing . Membrane . Mitochondrio...</td>\n",
       "      <td>27071</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Coenzyme Q-binding protein COQ10 homolog B, mi...</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>Q9H8M1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113.74</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25114</th>\n",
       "      <td>MAAVQMDPELAKRLFFEGATVVILNMPKGTEFGIDYNSWEVGPKFR...</td>\n",
       "      <td>384</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0028</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3D-structure . Acetylation . mRNA processing ....</td>\n",
       "      <td>43472</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Protein AAR2 homolog (AAR2 splicing factor hom...</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>Q9Y312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113.21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25115</th>\n",
       "      <td>MRDLLQYIACFFAFFSAGFLIVATWTDCWMVNADDSLEVSTKCRGL...</td>\n",
       "      <td>235</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0054</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Cell junction . Cell membrane . Disease varian...</td>\n",
       "      <td>26078</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Claudin-16 (Paracellin-1) (PCLN-1)</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>Q9Y5I7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>110.97</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25116</th>\n",
       "      <td>MGAAGSDGRCCVRSGRAGTGGAGSKWVVMDLWTGGAGSRRAVLGPD...</td>\n",
       "      <td>254</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0253</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Alternative splicing . Reference proteome . Re...</td>\n",
       "      <td>26128</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>Putative uncharacterized protein AFDN-DT (AFDN...</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>Q9Y6Z5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>102.87</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25117 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Sequence  Length  virus  \\\n",
       "0      MAMRSGRHPSLLLLLVLLLWLLQVSIIDSVQQETDDLTKQTKEKIY...     832      0   \n",
       "1      MMNLEKNFDKTTLFNHMRTDKRGKCSDLNEYGTSCDKTTAVEYNKV...     819      0   \n",
       "2      MATFKDACYHYKRINKLNHTVLKLGVNDTWRSSPPTKYKGWCLDCC...     486      1   \n",
       "3      MSAKRAELKKTHLSKNYKAVCLELKPEPTKTFDYKAVKQEGRFTKA...     233      0   \n",
       "4      MGSRTPESPLHAVQLRWGPRRRPPLLPLLLLLLPPPPRVGGFNLDA...    1049      0   \n",
       "...                                                  ...     ...    ...   \n",
       "25112  MKSYTPYFILLWSAVGIAKAAKIIIVPPIMFESHMYIFKTLASALH...     541      0   \n",
       "25113  MAARTGHTALRRVVSGCRPKSATAAGAQAPVRNGRYLASCGILMSR...     238      0   \n",
       "25114  MAAVQMDPELAKRLFFEGATVVILNMPKGTEFGIDYNSWEVGPKFR...     384      0   \n",
       "25115  MRDLLQYIACFFAFFSAGFLIVATWTDCWMVNADDSLEVSTKCRGL...     235      0   \n",
       "25116  MGAAGSDGRCCVRSGRAGTGGAGSKWVVMDLWTGGAGSRRAVLGPD...     254      0   \n",
       "\n",
       "       model_pred_proba_vir  model_pred_vir  mistake Baltimore  \\\n",
       "0                    0.0006               0    False       NaN   \n",
       "1                    0.0000               0    False       NaN   \n",
       "2                    1.0000               1    False     dsRNA   \n",
       "3                    0.2340               0    False       NaN   \n",
       "4                    0.0001               0    False       NaN   \n",
       "...                     ...             ...      ...       ...   \n",
       "25112                0.0025               0    False       NaN   \n",
       "25113                0.0001               0    False       NaN   \n",
       "25114                0.0028               0    False       NaN   \n",
       "25115                0.0054               0    False       NaN   \n",
       "25116                0.0253               0    False       NaN   \n",
       "\n",
       "               Family Genome Composition      Genus  \\\n",
       "0                 NaN                NaN        NaN   \n",
       "1                 NaN                NaN        NaN   \n",
       "2      Sedoreoviridae              dsRNA  Rotavirus   \n",
       "3                 NaN                NaN        NaN   \n",
       "4                 NaN                NaN        NaN   \n",
       "...               ...                ...        ...   \n",
       "25112             NaN                NaN        NaN   \n",
       "25113             NaN                NaN        NaN   \n",
       "25114             NaN                NaN        NaN   \n",
       "25115             NaN                NaN        NaN   \n",
       "25116             NaN                NaN        NaN   \n",
       "\n",
       "                                                Keywords    Mass  \\\n",
       "0      Alternative splicing . Calcium . Cell adhesion...   92416   \n",
       "1      DNA-binding . Metal-binding . Nucleus . Refere...   94331   \n",
       "2      Host cytoplasm . Host cytoskeleton . Host-viru...   57672   \n",
       "3      Alternative splicing . Coiled coil . Reference...   26518   \n",
       "4      3D-structure . Calcium . Cell adhesion . Cell ...  114536   \n",
       "...                                                  ...     ...   \n",
       "25112  Endoplasmic reticulum . Glycoprotein . Glycosy...   61438   \n",
       "25113  Alternative splicing . Membrane . Mitochondrio...   27071   \n",
       "25114  3D-structure . Acetylation . mRNA processing ....   43472   \n",
       "25115  Cell junction . Cell membrane . Disease varian...   26078   \n",
       "25116  Alternative splicing . Reference proteome . Re...   26128   \n",
       "\n",
       "                                                Organism  \\\n",
       "0                                   Homo sapiens (Human)   \n",
       "1                                   Homo sapiens (Human)   \n",
       "2      Rotavirus A (strain RVA/Human/United States/P/...   \n",
       "3                                   Homo sapiens (Human)   \n",
       "4                                   Homo sapiens (Human)   \n",
       "...                                                  ...   \n",
       "25112                               Homo sapiens (Human)   \n",
       "25113                               Homo sapiens (Human)   \n",
       "25114                               Homo sapiens (Human)   \n",
       "25115                               Homo sapiens (Human)   \n",
       "25116                               Homo sapiens (Human)   \n",
       "\n",
       "                                           Protein names  \\\n",
       "0      Cadherin-like protein 26 (Cadherin-like protei...   \n",
       "1                               Zinc finger protein 658B   \n",
       "2      Non-structural protein 1 (NSP1) (NCVP2) (Non-s...   \n",
       "3                            Testis-expressed protein 35   \n",
       "4      Integrin alpha-5 (CD49 antigen-like family mem...   \n",
       "...                                                  ...   \n",
       "25112  2-hydroxyacylsphingosine 1-beta-galactosyltran...   \n",
       "25113  Coenzyme Q-binding protein COQ10 homolog B, mi...   \n",
       "25114  Protein AAR2 homolog (AAR2 splicing factor hom...   \n",
       "25115                 Claudin-16 (Paracellin-1) (PCLN-1)   \n",
       "25116  Putative uncharacterized protein AFDN-DT (AFDN...   \n",
       "\n",
       "                                       Taxonomic lineage UR50_Cluster ID  \\\n",
       "0      cellular organisms, Eukaryota, Opisthokonta, M...          Q8IXH8   \n",
       "1      cellular organisms, Eukaryota, Opisthokonta, M...          Q4V348   \n",
       "2      Viruses, Riboviria, Orthornavirae (kingdom), D...             NaN   \n",
       "3      cellular organisms, Eukaryota, Opisthokonta, M...          Q5T0J7   \n",
       "4      cellular organisms, Eukaryota, Opisthokonta, M...          P08648   \n",
       "...                                                  ...             ...   \n",
       "25112  cellular organisms, Eukaryota, Opisthokonta, M...          Q16880   \n",
       "25113  cellular organisms, Eukaryota, Opisthokonta, M...          Q9H8M1   \n",
       "25114  cellular organisms, Eukaryota, Opisthokonta, M...          Q9Y312   \n",
       "25115  cellular organisms, Eukaryota, Opisthokonta, M...          Q9Y5I7   \n",
       "25116  cellular organisms, Eukaryota, Opisthokonta, M...          Q9Y6Z5   \n",
       "\n",
       "                              Virus hosts   av_mw genus_human_host human_host  \n",
       "0                                     NaN  111.08              NaN        NaN  \n",
       "1                                     NaN  115.18              NaN        NaN  \n",
       "2      Homo sapiens (Human) [TaxID: 9606]  118.67             True       True  \n",
       "3                                     NaN  113.81              NaN        NaN  \n",
       "4                                     NaN  109.19              NaN        NaN  \n",
       "...                                   ...     ...              ...        ...  \n",
       "25112                                 NaN  113.56              NaN        NaN  \n",
       "25113                                 NaN  113.74              NaN        NaN  \n",
       "25114                                 NaN  113.21              NaN        NaN  \n",
       "25115                                 NaN  110.97              NaN        NaN  \n",
       "25116                                 NaN  102.87              NaN        NaN  \n",
       "\n",
       "[25117 rows x 20 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_preds = df_preds.merge(df_meta, on=\"Sequence\",\n",
    "                 how= \"inner\",#'outer',\n",
    "                          suffixes=('', '_y'))\n",
    "\n",
    "df_preds.drop(df_preds.filter(regex='_y$').columns, axis=1, inplace=True)\n",
    "df_preds.drop(columns=[\"group\"],errors=\"ignore\",inplace=True)\n",
    "df_preds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6804db41-4931-4d82-b982-ffc6325829ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "if SAVE_CV_MISTAKE_PREDS:\n",
    "    df_preds.to_csv(\"mistake_preds_4cv_650+len_lr_2024.csv\",index=False)\n",
    "    df_preds.to_csv(\"./outputs/mistake_preds_4cv_650+len_lr_2024.csv.gz\",index=False,compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "be78c072",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Length</th>\n",
       "      <th>virus</th>\n",
       "      <th>model_pred_proba_vir</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22581</th>\n",
       "      <td>MQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ...</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17012</th>\n",
       "      <td>MDSYSAKIRANLVCRRSTDPSIRVTFSSRSLGSLPAFAMFRSSRPS...</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "      <td>0.9997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8265</th>\n",
       "      <td>MGCIQSITCKARIRRENIVVYDVCATIDQCPTRIEETSPIVLRYKT...</td>\n",
       "      <td>261</td>\n",
       "      <td>0</td>\n",
       "      <td>0.9997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3121</th>\n",
       "      <td>MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...</td>\n",
       "      <td>247</td>\n",
       "      <td>0</td>\n",
       "      <td>0.9995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2458</th>\n",
       "      <td>MSTSRKLKSHGMRRSKSRSPHKGVKRGGSKRKYRKGNLKSRKRGDD...</td>\n",
       "      <td>55</td>\n",
       "      <td>0</td>\n",
       "      <td>0.9994</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6567</th>\n",
       "      <td>MGDPNSRKKQALNRLRAQLRKKKESLADQFDFKMYIAFVFKEKKKK...</td>\n",
       "      <td>229</td>\n",
       "      <td>0</td>\n",
       "      <td>0.9989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5364</th>\n",
       "      <td>MYGKIIFVLLLSGIVSISASSTTGVAMHTSTSSSVTKSYISSQTNG...</td>\n",
       "      <td>78</td>\n",
       "      <td>0</td>\n",
       "      <td>0.9987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17134</th>\n",
       "      <td>MAECPTLGEAVTDHPDRLWAWEKFVYLDEKQHAWLPLTIEIKDRLQ...</td>\n",
       "      <td>114</td>\n",
       "      <td>0</td>\n",
       "      <td>0.9986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2498</th>\n",
       "      <td>MANTQLDHLHYTTEFTRNDLLIICKKFNLMLMDEDIISLLAIFIKM...</td>\n",
       "      <td>102</td>\n",
       "      <td>0</td>\n",
       "      <td>0.9984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24540</th>\n",
       "      <td>MLGARVAAHLDALGPLVPYVPPPLLPSMFYVGLFFVNVLILYYAFL...</td>\n",
       "      <td>95</td>\n",
       "      <td>0</td>\n",
       "      <td>0.9983</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Sequence  Length  virus  \\\n",
       "22581  MQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ...      80      0   \n",
       "17012  MDSYSAKIRANLVCRRSTDPSIRVTFSSRSLGSLPAFAMFRSSRPS...      81      0   \n",
       "8265   MGCIQSITCKARIRRENIVVYDVCATIDQCPTRIEETSPIVLRYKT...     261      0   \n",
       "3121   MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...     247      0   \n",
       "2458   MSTSRKLKSHGMRRSKSRSPHKGVKRGGSKRKYRKGNLKSRKRGDD...      55      0   \n",
       "6567   MGDPNSRKKQALNRLRAQLRKKKESLADQFDFKMYIAFVFKEKKKK...     229      0   \n",
       "5364   MYGKIIFVLLLSGIVSISASSTTGVAMHTSTSSSVTKSYISSQTNG...      78      0   \n",
       "17134  MAECPTLGEAVTDHPDRLWAWEKFVYLDEKQHAWLPLTIEIKDRLQ...     114      0   \n",
       "2498   MANTQLDHLHYTTEFTRNDLLIICKKFNLMLMDEDIISLLAIFIKM...     102      0   \n",
       "24540  MLGARVAAHLDALGPLVPYVPPPLLPSMFYVGLFFVNVLILYYAFL...      95      0   \n",
       "\n",
       "       model_pred_proba_vir  \n",
       "22581                1.0000  \n",
       "17012                0.9997  \n",
       "8265                 0.9997  \n",
       "3121                 0.9995  \n",
       "2458                 0.9994  \n",
       "6567                 0.9989  \n",
       "5364                 0.9987  \n",
       "17134                0.9986  \n",
       "2498                 0.9984  \n",
       "24540                0.9983  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(df_preds[df_preds[\"mistake\"]].sort_values([\"model_pred_proba_vir\"],ascending=False).iloc[0:10,0:4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0a0d2785",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Baltimore\n",
       "dsDNA       0.060394\n",
       "dsDNA-RT    0.027778\n",
       "dsRNA       0.005587\n",
       "ssDNA       0.062016\n",
       "ssRNA       0.047198\n",
       "ssRNA-RT    0.108108\n",
       "Name: mistake, dtype: float64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_preds.groupby(\"Baltimore\")[\"mistake\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc91b48a-64e1-4f82-94d1-a9f460bb408d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
