{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "3_5w39f_fnY9"
   },
   "source": [
    "Get preds over all data\n",
    "\n",
    "esm based dynamic model (not using static embeds).\n",
    "\n",
    "\n",
    "Use TF:\n",
    "*  ðŸ‡°https://colab.research.google.com/github/huggingface/notebooks/blob/main/examples/protein_language_modeling-tf.ipynb#scrollTo=de8419b5\n",
    "* Torch based /Trainer example:  https://colab.research.google.com/github/huggingface/notebooks/blob/main/examples/protein_language_modeling.ipynb#scrollTo=49dcba23\n",
    "\n",
    "\n",
    "\n",
    "* NOte for pytorch training could use trainer maybe, and mixed precision - https://huggingface.co/docs/transformers/v4.18.0/en/performance#fp16-training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "kalvvWZZgM0E",
    "outputId": "f575c531-2ec5-4880-fba6-73597c8e58cb"
   },
   "outputs": [],
   "source": [
    "# from google.colab import drive\n",
    "# drive.mount('/content/drive')\n",
    "\n",
    "# # drive.mount(\"/content/drive/MyDrive/Research/proteins\")\n",
    "# # drive.mount(\"/content/drive/MyDrive/proteins\")\n",
    "\n",
    "# !pip3 install tensorflow datasets transformers accelerate fair-esm -U #-q --user\n",
    "# !pip install torch accelerate transformers fair-esm datasets  -U -q\n",
    "\n",
    "# import os\n",
    "# for dirname, _, filenames in os.walk('/kaggle/input'):\n",
    "#     for filename in filenames:\n",
    "#         print(os.path.join(dirname, filename))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Z_cRmHN0j709"
   },
   "source": [
    "* Use the unirpot fasta file I downloaded and uploaded to my drive\n",
    "\n",
    "`/content/drive/MyDrive/Research/biodata/proteins/Transmembrane_human_90.fasta`\n",
    "\n",
    "* Download fasta from: `https://www.uniprot.org/uniref/?query=uniprot:(keyword%3A%22Transmembrane+%5BKW-0812%5D%22+AND+organism%3A%22Homo+sapiens+%28Human%29+%5B9606%5D%22)+identity:0.9`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "kwgVw-CN1OgJ"
   },
   "outputs": [],
   "source": [
    "# DATA_PATH = \"/content/drive/MyDrive/Research/biodata/proteins/membrane_100_whole_protbert_embed.snappy.parquet\"\n",
    "# DATA_PATH = \"/content/drive/MyDrive/Research/CIDR-Protein Anomalies project/protein_anomalies_data/swp_human_viri_all_embed_esm.parquet\" ## ESM1B embedding (max len 1022)\n",
    "# DATA_PATH = \"/content/drive/MyDrive/proteins/New Protein-Virus anom project/hum_vir_swp-globalEmbed-train.csv.gz\"## TRAIN\n",
    "DATA_PATH = \"hum_vir_swp-train.csv.gz\"## TRAIN\n",
    "# DATA_PATH =\"/kaggle/input/human-virus-proteins/hum_vir_swp-train.csv\"\n",
    "\n",
    "## TEST data:\n",
    "# TEST_DATA_PATH = \"/content/drive/MyDrive/proteins/New Protein-Virus anom project/hum_vir_swp-globalEmbed-test.csv.gz\"## TRAIN\n",
    "TEST_DATA_PATH = \"hum_vir_swp-test.csv.gz\"## TEST\n",
    "# TEST_DATA_PATH = \"/kaggle/input/human-virus-proteins/hum_vir_swp-test.csv\"\n",
    "# ## metadata for all reviewed/swissprot human + virus proteins\n",
    "# METADATA_PATH = \"/content/drive/MyDrive/Research/CIDR-Protein Anomalies project/protein_anomalies_data/SWP_human_viruses_all.xlsx\"\n",
    "\n",
    "TARGET_COL = \"virus\" ## use for filtering data into 1 class\n",
    "\n",
    "MAX_LEN = 612 # exclude sequences longer than this. (Not merely truncate)\n",
    "\n",
    "FAST_RUN = False#True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Xa-7WUuSfnZC"
   },
   "source": [
    "# Embed sequences in a FASTA file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "id": "hKw7JD3_fnZD",
    "outputId": "79159032-8099-4c16-b409-94eb74b0d371"
   },
   "outputs": [],
   "source": [
    "import torch\n",
    "# import esm\n",
    "import tensorflow\n",
    "import tensorflow as tf\n",
    "from tensorflow.keras.metrics import AUC as tf_auc\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "from sklearn.preprocessing  import MinMaxScaler\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.metrics import classification_report, roc_auc_score\n",
    "from sklearn.model_selection import train_test_split, cross_val_predict\n",
    "# from sklearn.linear_model import LogisticRegressionCV, LogisticRegression\n",
    "# # from sklearn.experimental import enable_hist_gradient_boosting\n",
    "# from sklearn.ensemble import HistGradientBoostingClassifier\n",
    "from tqdm.autonotebook import tqdm\n",
    "\n",
    "# from sklearn.svm import SVC\n",
    "from sklearn.pipeline import make_pipeline, Pipeline\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.metrics import roc_auc_score\n",
    "# from sklearn.decomposition import PCA\n",
    "\n",
    "from transformers import TrainingArguments, Trainer, logging\n",
    "# from accelerate import Accelerator\n",
    "# from torch.utils.data.dataloader import DataLoader\n",
    "## https://huggingface.co/docs/transformers/perf_train_gpu_one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "id": "Eggiq94CubAw",
    "outputId": "f4902346-b160-4efc-98cd-0bf7dad2de16"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>virus</th>\n",
       "      <th>Length</th>\n",
       "      <th>Cluster name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MADFLKGLPVYNKSNFSRFHADSVCKASNRRPSVYLPTREYPSEQI...</td>\n",
       "      <td>0</td>\n",
       "      <td>102</td>\n",
       "      <td>DET1- and DDB1-associated protein 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MPSTLQVLAKKVLALEHKENDHISREYYYHILKCCGLWWHEAPIIL...</td>\n",
       "      <td>1</td>\n",
       "      <td>362</td>\n",
       "      <td>Protein MGF 360-19R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MASSAELDFNLQALLEQLSQDELSKFKSLIRTISLGKELQTVPQTE...</td>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>Pyrin domain-containing protein 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAAWGKKHAGKDPVRDECEERNRFTETREEDVTDEHGEREPFAETD...</td>\n",
       "      <td>0</td>\n",
       "      <td>186</td>\n",
       "      <td>Protein FAM9B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MADPAAAQNYPLLKLLHSYTPTTPPRPIPKPAPWAPQKPRRQITND...</td>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "      <td>Protein E4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18996</th>\n",
       "      <td>MDPDKQDALNSIENSIYRTAFKLQSVQTLCQLDLIDSSLIQQVLLR...</td>\n",
       "      <td>0</td>\n",
       "      <td>578</td>\n",
       "      <td>Dystrotelin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18997</th>\n",
       "      <td>MLCPWRTANLGLLLILTIFLVAEAEGAAQPNNSLMLQTSKENHALA...</td>\n",
       "      <td>0</td>\n",
       "      <td>348</td>\n",
       "      <td>Cell surface glycoprotein CD200 receptor 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18998</th>\n",
       "      <td>MCLRFFSPVPGSTSSATNVTMVVSAGPWSSEKAEMNILEINEKLRP...</td>\n",
       "      <td>0</td>\n",
       "      <td>421</td>\n",
       "      <td>Putative neuroblastoma breakpoint family member 7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18999</th>\n",
       "      <td>MASHAGQQHAPAFGQAARASGPTDGRAASRPSHRQGASEARGDPEL...</td>\n",
       "      <td>1</td>\n",
       "      <td>376</td>\n",
       "      <td>Thymidine kinase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19000</th>\n",
       "      <td>MSASSLLEQRPKGQGNKVQNGSVHQKDGLNDDDFEPYLSPQARPNN...</td>\n",
       "      <td>0</td>\n",
       "      <td>579</td>\n",
       "      <td>YTH domain-containing family protein 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19001 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Sequence  virus  Length  \\\n",
       "0      MADFLKGLPVYNKSNFSRFHADSVCKASNRRPSVYLPTREYPSEQI...      0     102   \n",
       "1      MPSTLQVLAKKVLALEHKENDHISREYYYHILKCCGLWWHEAPIIL...      1     362   \n",
       "2      MASSAELDFNLQALLEQLSQDELSKFKSLIRTISLGKELQTVPQTE...      0      97   \n",
       "3      MAAWGKKHAGKDPVRDECEERNRFTETREEDVTDEHGEREPFAETD...      0     186   \n",
       "4      MADPAAAQNYPLLKLLHSYTPTTPPRPIPKPAPWAPQKPRRQITND...      1      93   \n",
       "...                                                  ...    ...     ...   \n",
       "18996  MDPDKQDALNSIENSIYRTAFKLQSVQTLCQLDLIDSSLIQQVLLR...      0     578   \n",
       "18997  MLCPWRTANLGLLLILTIFLVAEAEGAAQPNNSLMLQTSKENHALA...      0     348   \n",
       "18998  MCLRFFSPVPGSTSSATNVTMVVSAGPWSSEKAEMNILEINEKLRP...      0     421   \n",
       "18999  MASHAGQQHAPAFGQAARASGPTDGRAASRPSHRQGASEARGDPEL...      1     376   \n",
       "19000  MSASSLLEQRPKGQGNKVQNGSVHQKDGLNDDDFEPYLSPQARPNN...      0     579   \n",
       "\n",
       "                                            Cluster name  \n",
       "0                    DET1- and DDB1-associated protein 1  \n",
       "1                                    Protein MGF 360-19R  \n",
       "2                      Pyrin domain-containing protein 2  \n",
       "3                                          Protein FAM9B  \n",
       "4                                             Protein E4  \n",
       "...                                                  ...  \n",
       "18996                                        Dystrotelin  \n",
       "18997         Cell surface glycoprotein CD200 receptor 1  \n",
       "18998  Putative neuroblastoma breakpoint family member 7  \n",
       "18999                                   Thymidine kinase  \n",
       "19000             YTH domain-containing family protein 2  \n",
       "\n",
       "[19001 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df = pd.read_parquet(DATA_PATH) # numpy to pandas\n",
    "df = pd.read_csv(DATA_PATH,usecols=[\"Sequence\",\"virus\",\"Length\",\"Cluster name\"])\n",
    "df_test = pd.read_csv(TEST_DATA_PATH,usecols=[\"Sequence\",\"virus\",\"Length\",\"Cluster name\"])\n",
    " ## lengths of all the seqs\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "id": "fkeYBgfafVoy"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>virus</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13587.0</td>\n",
       "      <td>413.60</td>\n",
       "      <td>230.28</td>\n",
       "      <td>11.0</td>\n",
       "      <td>230.5</td>\n",
       "      <td>373.0</td>\n",
       "      <td>558.00</td>\n",
       "      <td>1024.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5414.0</td>\n",
       "      <td>338.62</td>\n",
       "      <td>232.85</td>\n",
       "      <td>11.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>275.0</td>\n",
       "      <td>490.75</td>\n",
       "      <td>1023.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         count    mean     std   min    25%    50%     75%     max\n",
       "virus                                                             \n",
       "0      13587.0  413.60  230.28  11.0  230.5  373.0  558.00  1024.0\n",
       "1       5414.0  338.62  232.85  11.0  146.0  275.0  490.75  1023.0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby([\"virus\"])[\"Length\"].describe().round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "id": "rm2Ms4BmQ4wn",
    "outputId": "3cd408f3-b45e-4bcd-ec67-604d441fe4e2"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mean         0.28\n",
      "sum       5414.00\n",
      "count    19001.00\n",
      "Name: virus, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "print(df[\"virus\"].agg([\"mean\",\"sum\",\"count\"]).round(2))\n",
    "# df[\"Length\"].describe().round(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "id": "aNDVeR2vQ9YZ",
    "outputId": "6b144341-817f-4ea9-82e5-242fb54639ff"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    4491.0\n",
       "mean      393.8\n",
       "std       233.7\n",
       "min        18.0\n",
       "25%       204.0\n",
       "50%       350.0\n",
       "75%       538.5\n",
       "max      1023.0\n",
       "Name: Length, dtype: float64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_test[\"Length\"].describe().round(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "id": "F8VNnHUwRQRm",
    "outputId": "d46999a0-aadf-4a25-8b74-af46363b26e7"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count    15535.0\n",
      "mean       306.0\n",
      "std        151.7\n",
      "min         11.0\n",
      "25%        176.0\n",
      "50%        302.0\n",
      "75%        428.0\n",
      "max        612.0\n",
      "Name: Length, dtype: float64\n",
      "mean         0.3\n",
      "sum       4670.0\n",
      "count    15535.0\n",
      "Name: virus, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "df = df.loc[df[\"Length\"]<=MAX_LEN].reset_index(drop=True).copy()\n",
    "df_test = df_test.loc[df_test[\"Length\"]<=MAX_LEN].reset_index(drop=True).copy()\n",
    "\n",
    "print(df[\"Length\"].describe().round(1))\n",
    "print(df[\"virus\"].agg([\"mean\",\"sum\",\"count\"]).round(2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "if FAST_RUN:\n",
    "    df = df[df[\"Length\"]>450].sample(25).reset_index(drop=True).copy()\n",
    "    df_test = df_test[df_test[\"Length\"]>=500].sample(17).reset_index(drop=True).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "id": "PcMwJBkUov07"
   },
   "outputs": [],
   "source": [
    "# # ## metadata about all sequences, can be used to identify and to define targets/labels\n",
    "### Old cleaning/gen  wrangle code: \n",
    "# # new : # more correct and coverage ? (by ~200 cases? despite being smaller file)\n",
    "# df_meta = pd.read_csv(\"SWP_human_or_virus-animal-data.csv\",\n",
    "#                      usecols=[ 'Protein names', 'Organism', 'Sequence',\n",
    "#         'Mass',\"av_mw\", 'Keywords', 'Virus hosts', 'Taxonomic lineage',\"UR50_Cluster ID\"]) \n",
    "# print(df_meta.nunique())\n",
    "# print(df_meta[\"UR50_Cluster ID\"].isna().mean()) # 16% of cases missing cluster for some reason. Can't use generic \"missing\" name, so use prot name? First part of entry name?  organism? \n",
    "# ## df_meta[\"Entry Name\"].str.split(\"_\",expand=True)[0].nunique() # 20540\n",
    "# df_meta[\"group\"] = df_meta[\"UR50_Cluster ID\"].fillna(df_meta['Protein names'].str.split(\"_\",expand=True)[0].str.strip())\n",
    "\n",
    "# # df_meta[\"av_mw\"] = df_meta[\"Mass\"].div(df_meta[\"Length\"]).round(2)\n",
    "# df_meta[\"Keywords\"] = df_meta[\"Keywords\"].str.replace(\";\",\" . \").str.strip()\n",
    "# df_meta[\"Taxonomic lineage\"] = df_meta[\"Taxonomic lineage\"].str.replace(\" (no rank)\",\"\").str.replace(\" (superkingdom)\",\"\").str.strip()\n",
    "\n",
    "# display(df_meta)\n",
    "\n",
    "# df_meta.to_csv(\"SWP_humVir_meta.csv.gz\",index=False,compression=\"gzip\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence             25119\n",
      "Protein names        21394\n",
      "Organism              1136\n",
      "Mass                 22130\n",
      "Keywords             15980\n",
      "av_mw                 2110\n",
      "Virus hosts            291\n",
      "Taxonomic lineage      561\n",
      "UR50_Cluster ID      20886\n",
      "group                23473\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Mass</th>\n",
       "      <th>Keywords</th>\n",
       "      <th>av_mw</th>\n",
       "      <th>Virus hosts</th>\n",
       "      <th>Taxonomic lineage</th>\n",
       "      <th>UR50_Cluster ID</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGLEALVPLAMIVAIFLLLVDLMHRHQRWAARYPPGPLPLPGLGNL...</td>\n",
       "      <td>Putative cytochrome P450 2D7 (EC 1.14.14.1)</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>57489</td>\n",
       "      <td>Cytoplasm . Glycoprotein . Heme . Iron . Membr...</td>\n",
       "      <td>111.63</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Putative cytochrome P450 2D7 (EC 1.14.14.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFRRLTFAQLLFATVLGIAGGVYIFQPVFEQYAKDQKELKEKMQLV...</td>\n",
       "      <td>Protein PIGBOS1 (PIGB opposite strand protein 1)</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>6313</td>\n",
       "      <td>Direct protein sequencing . Membrane . Mitocho...</td>\n",
       "      <td>116.91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>A0A0B4J2F0</td>\n",
       "      <td>A0A0B4J2F0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MRWQEMGYIFYPRKLR</td>\n",
       "      <td>Mitochondrial-derived peptide MOTS-c (Mitochon...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>2175</td>\n",
       "      <td>DNA-binding . Mitochondrion . Nucleus . Osteog...</td>\n",
       "      <td>135.94</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MTQRAGAAMLPSALLLLCVPGCLTVSGPSTVMGAVGESLSVQCRYE...</td>\n",
       "      <td>Protein CD300H (CD300 antigen-like family memb...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>21806</td>\n",
       "      <td>Alternative splicing . Disulfide bond . Glycop...</td>\n",
       "      <td>108.49</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>A0A0K2S4Q6</td>\n",
       "      <td>A0A0K2S4Q6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGDQPCASGRSTLPPGNAREAKPPKKRCLLAPRWDYPEGTPNGGST...</td>\n",
       "      <td>Negative regulator of P-body association (P-bo...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>7025</td>\n",
       "      <td>Cytoplasm . mRNA processing . Reference proteome</td>\n",
       "      <td>103.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>A0A0U1RRE5</td>\n",
       "      <td>A0A0U1RRE5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25114</th>\n",
       "      <td>MKTSPRLSVIDGVGEEVRPDLFQLVHVVPPNTCEPPEQWLRVTVLL...</td>\n",
       "      <td>Uncharacterized protein ORF6</td>\n",
       "      <td>Beak and feather disease virus (BFDV)</td>\n",
       "      <td>9846</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111.89</td>\n",
       "      <td>Gracula [TaxID: 116991]; Psittaciformes [TaxID...</td>\n",
       "      <td>Viruses, Monodnaviria, Shotokuvirae (kingdom),...</td>\n",
       "      <td>Q9YUD1</td>\n",
       "      <td>Q9YUD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25115</th>\n",
       "      <td>MGGAYITSRYWLVPAHVTSRLRLTSFTGHRGVARVDGPMSSLGLNI...</td>\n",
       "      <td>Uncharacterized protein ORF5</td>\n",
       "      <td>Beak and feather disease virus (BFDV)</td>\n",
       "      <td>10863</td>\n",
       "      <td>NaN</td>\n",
       "      <td>107.55</td>\n",
       "      <td>Gracula [TaxID: 116991]; Psittaciformes [TaxID...</td>\n",
       "      <td>Viruses, Monodnaviria, Shotokuvirae (kingdom),...</td>\n",
       "      <td>Q9YUD2</td>\n",
       "      <td>Q9YUD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25116</th>\n",
       "      <td>MPYVDFWELSRDFLHFASGHHGSNSSVERAVTVSRHNAQGEYNVPL...</td>\n",
       "      <td>Uncharacterized protein ORF3</td>\n",
       "      <td>Beak and feather disease virus (BFDV)</td>\n",
       "      <td>17738</td>\n",
       "      <td>NaN</td>\n",
       "      <td>110.86</td>\n",
       "      <td>Gracula [TaxID: 116991]; Psittaciformes [TaxID...</td>\n",
       "      <td>Viruses, Monodnaviria, Shotokuvirae (kingdom),...</td>\n",
       "      <td>Q9YUD4</td>\n",
       "      <td>Q9YUD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25117</th>\n",
       "      <td>MNNGQATITRNGGRFEIRCRHLDRDYTMPLPNATSNDNFLDCIKFI...</td>\n",
       "      <td>Non-structural protein 3 (NS3)</td>\n",
       "      <td>Banna virus (strain Indonesia/JKT-6423/1980) (...</td>\n",
       "      <td>34991</td>\n",
       "      <td>Reference proteome</td>\n",
       "      <td>114.35</td>\n",
       "      <td>Bos taurus (Bovine) [TaxID: 9913]; Culex annul...</td>\n",
       "      <td>Viruses, Riboviria, Orthornavirae (kingdom), D...</td>\n",
       "      <td>Q9YWQ0</td>\n",
       "      <td>Q9YWQ0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25118</th>\n",
       "      <td>MGRKEHESPSQPHMCGWEDSQKPSVPSHGPKTPSCKGVKAPHSSRP...</td>\n",
       "      <td>Putative uncharacterized protein GUCA1ANB (GUC...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>14531</td>\n",
       "      <td>Reference proteome</td>\n",
       "      <td>114.42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>X6R8D5</td>\n",
       "      <td>X6R8D5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25119 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Sequence  \\\n",
       "0      MGLEALVPLAMIVAIFLLLVDLMHRHQRWAARYPPGPLPLPGLGNL...   \n",
       "1      MFRRLTFAQLLFATVLGIAGGVYIFQPVFEQYAKDQKELKEKMQLV...   \n",
       "2                                       MRWQEMGYIFYPRKLR   \n",
       "3      MTQRAGAAMLPSALLLLCVPGCLTVSGPSTVMGAVGESLSVQCRYE...   \n",
       "4      MGDQPCASGRSTLPPGNAREAKPPKKRCLLAPRWDYPEGTPNGGST...   \n",
       "...                                                  ...   \n",
       "25114  MKTSPRLSVIDGVGEEVRPDLFQLVHVVPPNTCEPPEQWLRVTVLL...   \n",
       "25115  MGGAYITSRYWLVPAHVTSRLRLTSFTGHRGVARVDGPMSSLGLNI...   \n",
       "25116  MPYVDFWELSRDFLHFASGHHGSNSSVERAVTVSRHNAQGEYNVPL...   \n",
       "25117  MNNGQATITRNGGRFEIRCRHLDRDYTMPLPNATSNDNFLDCIKFI...   \n",
       "25118  MGRKEHESPSQPHMCGWEDSQKPSVPSHGPKTPSCKGVKAPHSSRP...   \n",
       "\n",
       "                                           Protein names  \\\n",
       "0            Putative cytochrome P450 2D7 (EC 1.14.14.1)   \n",
       "1       Protein PIGBOS1 (PIGB opposite strand protein 1)   \n",
       "2      Mitochondrial-derived peptide MOTS-c (Mitochon...   \n",
       "3      Protein CD300H (CD300 antigen-like family memb...   \n",
       "4      Negative regulator of P-body association (P-bo...   \n",
       "...                                                  ...   \n",
       "25114                       Uncharacterized protein ORF6   \n",
       "25115                       Uncharacterized protein ORF5   \n",
       "25116                       Uncharacterized protein ORF3   \n",
       "25117                     Non-structural protein 3 (NS3)   \n",
       "25118  Putative uncharacterized protein GUCA1ANB (GUC...   \n",
       "\n",
       "                                                Organism   Mass  \\\n",
       "0                                   Homo sapiens (Human)  57489   \n",
       "1                                   Homo sapiens (Human)   6313   \n",
       "2                                   Homo sapiens (Human)   2175   \n",
       "3                                   Homo sapiens (Human)  21806   \n",
       "4                                   Homo sapiens (Human)   7025   \n",
       "...                                                  ...    ...   \n",
       "25114              Beak and feather disease virus (BFDV)   9846   \n",
       "25115              Beak and feather disease virus (BFDV)  10863   \n",
       "25116              Beak and feather disease virus (BFDV)  17738   \n",
       "25117  Banna virus (strain Indonesia/JKT-6423/1980) (...  34991   \n",
       "25118                               Homo sapiens (Human)  14531   \n",
       "\n",
       "                                                Keywords   av_mw  \\\n",
       "0      Cytoplasm . Glycoprotein . Heme . Iron . Membr...  111.63   \n",
       "1      Direct protein sequencing . Membrane . Mitocho...  116.91   \n",
       "2      DNA-binding . Mitochondrion . Nucleus . Osteog...  135.94   \n",
       "3      Alternative splicing . Disulfide bond . Glycop...  108.49   \n",
       "4       Cytoplasm . mRNA processing . Reference proteome  103.31   \n",
       "...                                                  ...     ...   \n",
       "25114                                                NaN  111.89   \n",
       "25115                                                NaN  107.55   \n",
       "25116                                                NaN  110.86   \n",
       "25117                                 Reference proteome  114.35   \n",
       "25118                                 Reference proteome  114.42   \n",
       "\n",
       "                                             Virus hosts  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                                                    NaN   \n",
       "...                                                  ...   \n",
       "25114  Gracula [TaxID: 116991]; Psittaciformes [TaxID...   \n",
       "25115  Gracula [TaxID: 116991]; Psittaciformes [TaxID...   \n",
       "25116  Gracula [TaxID: 116991]; Psittaciformes [TaxID...   \n",
       "25117  Bos taurus (Bovine) [TaxID: 9913]; Culex annul...   \n",
       "25118                                                NaN   \n",
       "\n",
       "                                       Taxonomic lineage UR50_Cluster ID  \\\n",
       "0      cellular organisms, Eukaryota, Opisthokonta, M...             NaN   \n",
       "1      cellular organisms, Eukaryota, Opisthokonta, M...      A0A0B4J2F0   \n",
       "2      cellular organisms, Eukaryota, Opisthokonta, M...      A0A0C5B5G6   \n",
       "3      cellular organisms, Eukaryota, Opisthokonta, M...      A0A0K2S4Q6   \n",
       "4      cellular organisms, Eukaryota, Opisthokonta, M...      A0A0U1RRE5   \n",
       "...                                                  ...             ...   \n",
       "25114  Viruses, Monodnaviria, Shotokuvirae (kingdom),...          Q9YUD1   \n",
       "25115  Viruses, Monodnaviria, Shotokuvirae (kingdom),...          Q9YUD2   \n",
       "25116  Viruses, Monodnaviria, Shotokuvirae (kingdom),...          Q9YUD4   \n",
       "25117  Viruses, Riboviria, Orthornavirae (kingdom), D...          Q9YWQ0   \n",
       "25118  cellular organisms, Eukaryota, Opisthokonta, M...          X6R8D5   \n",
       "\n",
       "                                             group  \n",
       "0      Putative cytochrome P450 2D7 (EC 1.14.14.1)  \n",
       "1                                       A0A0B4J2F0  \n",
       "2                                       A0A0C5B5G6  \n",
       "3                                       A0A0K2S4Q6  \n",
       "4                                       A0A0U1RRE5  \n",
       "...                                            ...  \n",
       "25114                                       Q9YUD1  \n",
       "25115                                       Q9YUD2  \n",
       "25116                                       Q9YUD4  \n",
       "25117                                       Q9YWQ0  \n",
       "25118                                       X6R8D5  \n",
       "\n",
       "[25119 rows x 10 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv(\"SWP_humVir_meta.csv.gz\")\n",
    "print(df_meta.nunique())\n",
    "df_meta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MSXy_OzjhIB7"
   },
   "source": [
    "### example pretrained fb/torch:\n",
    "* https://github.com/facebookresearch/esm#getting-started-with-this-repo-\n",
    "\n",
    "* Transformers + trainer example : (mlm case): https://github.com/facebookresearch/esm/discussions/556\n",
    "* keras models supported / via HF?\n",
    "  * https://huggingface.co/docs/transformers/model_doc/esm#transformers.EsmForSequenceClassification.forward.example\n",
    "\n",
    "\n",
    "  TF finetuning example (sequence evel?):\n",
    "  https://colab.research.google.com/github/huggingface/notebooks/blob/main/examples/protein_language_modeling-tf.ipynb#scrollTo=4b26b828"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "id": "CDdMQ98rii9t"
   },
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "cannot import name 'Adafactor' from 'tensorflow.keras.optimizers' (/Users/oferd2/anaconda3/envs/ot/lib/python3.10/site-packages/keras/api/_v2/keras/optimizers/__init__.py)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[19], line 4\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdatasets\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m Dataset\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtransformers\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AdamWeightDecay\n\u001b[0;32m----> 4\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtensorflow\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mkeras\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01moptimizers\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m Adafactor, Adam \u001b[38;5;66;03m# more memory effecient than adamWD\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mtensorflow\u001b[39;00m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m## https://huggingface.co/docs/transformers/model_doc/esm#transformers.EsmForSequenceClassification.forward.example\u001b[39;00m\n",
      "\u001b[0;31mImportError\u001b[0m: cannot import name 'Adafactor' from 'tensorflow.keras.optimizers' (/Users/oferd2/anaconda3/envs/ot/lib/python3.10/site-packages/keras/api/_v2/keras/optimizers/__init__.py)"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "from datasets import Dataset\n",
    "from transformers import AdamWeightDecay\n",
    "from tensorflow.keras.optimizers import Adafactor, Adam # more memory effecient than adamWD\n",
    "import tensorflow\n",
    "\n",
    "## https://huggingface.co/docs/transformers/model_doc/esm#transformers.EsmForSequenceClassification.forward.example\n",
    "from transformers import AutoTokenizer, EsmForSequenceClassification, Trainer, TFAutoModel\n",
    "from transformers import TFAutoModelForTokenClassification, TFAutoModelForSequenceClassification ,TFEsmForSequenceClassification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "id": "rR9-cKgPjzW2"
   },
   "outputs": [],
   "source": [
    "train_sequences = df[\"Sequence\"].tolist()\n",
    "train_labels = df[\"virus\"].tolist()\n",
    "train_groups = df[\"Cluster name\"].tolist()\n",
    "\n",
    "# train_sequences, test_sequences, train_labels, test_labels = train_test_split(sequences, labels, test_size=0.25, shuffle=True)#,stratify=labels)\n",
    "\n",
    "test_sequences = df_test[\"Sequence\"].tolist()\n",
    "test_labels = df_test[\"virus\"].tolist()\n",
    "\n",
    "df_all = pd.concat([df,df_test])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>virus</th>\n",
       "      <th>Length</th>\n",
       "      <th>Cluster name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MADFLKGLPVYNKSNFSRFHADSVCKASNRRPSVYLPTREYPSEQI...</td>\n",
       "      <td>0</td>\n",
       "      <td>102</td>\n",
       "      <td>DET1- and DDB1-associated protein 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MPSTLQVLAKKVLALEHKENDHISREYYYHILKCCGLWWHEAPIIL...</td>\n",
       "      <td>1</td>\n",
       "      <td>362</td>\n",
       "      <td>Protein MGF 360-19R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MASSAELDFNLQALLEQLSQDELSKFKSLIRTISLGKELQTVPQTE...</td>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>Pyrin domain-containing protein 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAAWGKKHAGKDPVRDECEERNRFTETREEDVTDEHGEREPFAETD...</td>\n",
       "      <td>0</td>\n",
       "      <td>186</td>\n",
       "      <td>Protein FAM9B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MADPAAAQNYPLLKLLHSYTPTTPPRPIPKPAPWAPQKPRRQITND...</td>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "      <td>Protein E4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3677</th>\n",
       "      <td>MAAGDGDVKLGTLGSGSESSNDGGSESPGDAGAAAEGGGWAAAALA...</td>\n",
       "      <td>0</td>\n",
       "      <td>223</td>\n",
       "      <td>Membrane-associated progesterone receptor comp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3678</th>\n",
       "      <td>MTLPASLKGFETLDIFNISSVRNICELNKHAEHVSIKNLPLIDISV...</td>\n",
       "      <td>1</td>\n",
       "      <td>256</td>\n",
       "      <td>Cytoplasmic envelopment protein 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3679</th>\n",
       "      <td>MKWKALFTAAILQAQLPITEAQSFGLLDPKLCYLLDGILFIYGVIL...</td>\n",
       "      <td>0</td>\n",
       "      <td>164</td>\n",
       "      <td>T-cell surface glycoprotein CD3 zeta chain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3680</th>\n",
       "      <td>MAGVGDAAAPGEGGGGGVDGPQRDGRGEAEQPGGSGGQGPPPAPQL...</td>\n",
       "      <td>0</td>\n",
       "      <td>567</td>\n",
       "      <td>Transmembrane anterior posterior transformatio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3681</th>\n",
       "      <td>MGASVSRGRAARVPAPEPEPEEALDLSQLPPELLLVVLSHVPPRTL...</td>\n",
       "      <td>0</td>\n",
       "      <td>283</td>\n",
       "      <td>F-box only protein 27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19217 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Sequence  virus  Length  \\\n",
       "0     MADFLKGLPVYNKSNFSRFHADSVCKASNRRPSVYLPTREYPSEQI...      0     102   \n",
       "1     MPSTLQVLAKKVLALEHKENDHISREYYYHILKCCGLWWHEAPIIL...      1     362   \n",
       "2     MASSAELDFNLQALLEQLSQDELSKFKSLIRTISLGKELQTVPQTE...      0      97   \n",
       "3     MAAWGKKHAGKDPVRDECEERNRFTETREEDVTDEHGEREPFAETD...      0     186   \n",
       "4     MADPAAAQNYPLLKLLHSYTPTTPPRPIPKPAPWAPQKPRRQITND...      1      93   \n",
       "...                                                 ...    ...     ...   \n",
       "3677  MAAGDGDVKLGTLGSGSESSNDGGSESPGDAGAAAEGGGWAAAALA...      0     223   \n",
       "3678  MTLPASLKGFETLDIFNISSVRNICELNKHAEHVSIKNLPLIDISV...      1     256   \n",
       "3679  MKWKALFTAAILQAQLPITEAQSFGLLDPKLCYLLDGILFIYGVIL...      0     164   \n",
       "3680  MAGVGDAAAPGEGGGGGVDGPQRDGRGEAEQPGGSGGQGPPPAPQL...      0     567   \n",
       "3681  MGASVSRGRAARVPAPEPEPEEALDLSQLPPELLLVVLSHVPPRTL...      0     283   \n",
       "\n",
       "                                           Cluster name  \n",
       "0                   DET1- and DDB1-associated protein 1  \n",
       "1                                   Protein MGF 360-19R  \n",
       "2                     Pyrin domain-containing protein 2  \n",
       "3                                         Protein FAM9B  \n",
       "4                                            Protein E4  \n",
       "...                                                 ...  \n",
       "3677  Membrane-associated progesterone receptor comp...  \n",
       "3678                  Cytoplasmic envelopment protein 1  \n",
       "3679         T-cell surface glycoprotein CD3 zeta chain  \n",
       "3680  Transmembrane anterior posterior transformatio...  \n",
       "3681                              F-box only protein 27  \n",
       "\n",
       "[19217 rows x 4 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T10:55:44.980369Z",
     "iopub.status.busy": "2023-08-29T10:55:44.979947Z",
     "iopub.status.idle": "2023-08-29T10:55:44.988938Z",
     "shell.execute_reply": "2023-08-29T10:55:44.987935Z",
     "shell.execute_reply.started": "2023-08-29T10:55:44.980274Z"
    },
    "id": "k_KUug1Lh5bH",
    "outputId": "8249a98b-2849-49bc-8610-c7216fb44b9e"
   },
   "outputs": [],
   "source": [
    "# # Load ESM-2 model\n",
    "# ## smallest: esm2_t6_8M_UR50D\n",
    "# ## 2d smallest\n",
    "# ## large: esm2_t33_650M_UR50D\n",
    "# # model_checkpoint =  \"facebook/esm2_t12_35M_UR50D\"#\"facebook/esm2_t6_8M_UR50D\"\n",
    "# model_checkpoint =  \"facebook/esm2_t30_150M_UR50D\"\n",
    "#  # ElnaggarLab/ankh-base\n",
    "#  ### https://github.com/agemagician/Ankh/blob/main/examples/binary_classification_solubility_task.ipynb - different model?\n",
    "#  #  https://github.com/agemagician/Ankh#models   - 450M model size\n",
    "# # model_checkpoint =   \"ElnaggarLab/ankh-base\"\n",
    "\n",
    "# model_max_len = min(800,MAX_LEN)\n",
    "\n",
    "# tokenizer = AutoTokenizer.from_pretrained(model_checkpoint, padding=True, truncation=True,max_length=model_max_len)\n",
    "\n",
    "# num_labels = max(train_labels + test_labels) #+ 1  # Add 1 since 0 can be a label\n",
    "# print(\"Num labels:\", num_labels)\n",
    "# model = TFAutoModelForSequenceClassification.from_pretrained(model_checkpoint, num_labels=num_labels,problem_type=\"single_label_classification\") # worked, orig\n",
    "# # model = EsmForSequenceClassification.from_pretrained(model_checkpoint, num_labels=num_labels)\n",
    "# # model = TFEsmForSequenceClassification.from_pretrained(model_checkpoint, num_labels=num_labels,problem_type=\"single_label_classification\")\n",
    "\n",
    "# # model.eval()  # disables dropout for deterministic results # from original - in infer/embed only example\n",
    "\n",
    "# # last_layer_num = model.num_layers ## 33 for esm2_t33_650M_UR50D\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T10:55:44.992875Z",
     "iopub.status.busy": "2023-08-29T10:55:44.992042Z",
     "iopub.status.idle": "2023-08-29T10:55:45.165118Z",
     "shell.execute_reply": "2023-08-29T10:55:45.164061Z",
     "shell.execute_reply.started": "2023-08-29T10:55:44.992836Z"
    }
   },
   "outputs": [],
   "source": [
    "from transformers import AutoTokenizer, TFAutoModel\n",
    "import tensorflow as tf\n",
    "from tensorflow.keras.layers import Input, GlobalMaxPooling1D, GlobalAveragePooling1D, Concatenate, Dense, Dropout, SpatialDropout1D\n",
    "from tensorflow.keras.models import Model\n",
    "\n",
    "\n",
    "# # Load pre-trained model and tokenizer\n",
    "model_checkpoint = \"facebook/esm2_t30_150M_UR50D\"#\"facebook/esm2_t6_8M_UR50D\"  # Replace with your pre-trained model checkpoint\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)\n",
    "\n",
    "# # Parameters\n",
    "model_max_len = MAX_LEN\n",
    "batch_size_train = 4#5\n",
    "batch_size_test = 4#5\n",
    "learning_rate = 1e-4 #8e-5\n",
    "num_labels = 1  # Replace with the number of unique labels in your dataset\n",
    "\n",
    "# # Tokenize the sequences\n",
    "# train_tokenized = tokenizer(train_sequences, padding=True, truncation=True, max_length=model_max_len)\n",
    "# test_tokenized = tokenizer(test_sequences, padding=True, truncation=True, max_length=model_max_len)\n",
    "\n",
    "# # Convert to TensorFlow Dataset\n",
    "# train_dataset = tf.data.Dataset.from_tensor_slices((\n",
    "#     dict(train_tokenized),\n",
    "#     train_labels\n",
    "# )).batch(batch_size_train).shuffle(buffer_size=len(train_sequences))\n",
    "\n",
    "# test_dataset = tf.data.Dataset.from_tensor_slices((\n",
    "#     dict(test_tokenized),\n",
    "#     test_labels\n",
    "# )).batch(batch_size_test)\n",
    "\n",
    "# ## added. (Awhat about add col labels?)\n",
    "# train_dataset = train_dataset.map(lambda x, y: ({'input_layer': x['input_ids'], 'attention_mask': x['attention_mask']}, y))\n",
    "# test_dataset = test_dataset.map(lambda x, y: ({'input_layer': x['input_ids'], 'attention_mask': x['attention_mask']}, y))\n",
    "\n",
    "# # Load the pre-trained transformer model\n",
    "# base_model = TFAutoModel.from_pretrained(model_checkpoint)\n",
    "\n",
    "# # Freeze all layers except the last few\n",
    "# for layer in base_model.layers[:-4]: #[1:-3]:\n",
    "#     layer.trainable = False\n",
    "\n",
    "# # Create new model on top\n",
    "# input_layer = Input(shape=(model_max_len,), dtype=tf.int32, name='input_layer')\n",
    "# sequence_output = base_model(input_layer)[0]\n",
    "# sequence_output = SpatialDropout1D(0.15)(sequence_output)\n",
    "# # Add GlobalMeanPooling1D and GlobalMaxPooling1D\n",
    "# avg_pool = GlobalAveragePooling1D()(sequence_output)\n",
    "# # avg_pool = Dropout(0.15)(avg_pool)\n",
    "# max_pool = GlobalMaxPooling1D()(sequence_output)\n",
    "# # max_pool = Dropout(0.15)(max_pool)\n",
    "# concat = Concatenate()([avg_pool, max_pool])\n",
    "# concat = Dropout(0.1)(concat)\n",
    "\n",
    "# # # Initialize mixed precision training\n",
    "# # from tensorflow.keras import mixed_precision\n",
    "# # policy = mixed_precision.Policy('mixed_float16')\n",
    "# # mixed_precision.set_global_policy(policy)\n",
    "\n",
    "\n",
    "# # Add output layer for binary classification\n",
    "# output = Dense(1, activation='sigmoid', dtype='float32')(concat)  # Ensuring the dtype is float32 for the output layer\n",
    "\n",
    "# # Compile model\n",
    "# final_model = Model(inputs=input_layer, outputs=output)\n",
    "# optimizer = tf.keras.optimizers.Adam(learning_rate=learning_rate)\n",
    "# loss = \"BinaryFocalCrossentropy\"#tf.keras.losses.BinaryCrossentropy(from_logits=False)\n",
    "\n",
    "# final_model.compile(optimizer=optimizer, loss=loss, metrics=[\"accuracy\",tf.keras.metrics.AUC(num_thresholds=12,from_logits=True)])\n",
    "\n",
    "# # Train the model\n",
    "# final_model.fit(train_dataset, validation_data=test_dataset, epochs=4)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CV all points version (Slower!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T10:55:45.167122Z",
     "iopub.status.busy": "2023-08-29T10:55:45.166737Z",
     "iopub.status.idle": "2023-08-29T10:55:45.172645Z",
     "shell.execute_reply": "2023-08-29T10:55:45.171495Z",
     "shell.execute_reply.started": "2023-08-29T10:55:45.167079Z"
    }
   },
   "outputs": [],
   "source": [
    "assert len(train_sequences+test_sequences) == len(train_labels+test_labels)\n",
    "assert len(train_labels+test_labels)>len(train_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T11:09:11.865566Z",
     "iopub.status.busy": "2023-08-29T11:09:11.865195Z",
     "iopub.status.idle": "2023-08-29T11:14:40.848355Z",
     "shell.execute_reply": "2023-08-29T11:14:40.847337Z",
     "shell.execute_reply.started": "2023-08-29T11:09:11.865536Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-08-29 14:48:06.767268: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:982] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node\n",
      "Your kernel may have been built without NUMA support.\n",
      "2023-08-29 14:48:06.770109: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:982] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node\n",
      "Your kernel may have been built without NUMA support.\n",
      "2023-08-29 14:48:06.770127: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:982] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node\n",
      "Your kernel may have been built without NUMA support.\n",
      "2023-08-29 14:48:06.772242: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:982] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node\n",
      "Your kernel may have been built without NUMA support.\n",
      "2023-08-29 14:48:06.772262: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:982] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node\n",
      "Your kernel may have been built without NUMA support.\n",
      "2023-08-29 14:48:06.772270: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:982] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node\n",
      "Your kernel may have been built without NUMA support.\n",
      "2023-08-29 14:48:08.497857: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:982] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node\n",
      "Your kernel may have been built without NUMA support.\n",
      "2023-08-29 14:48:08.497934: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:982] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node\n",
      "Your kernel may have been built without NUMA support.\n",
      "2023-08-29 14:48:08.497940: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1722] Could not identify NUMA node of platform GPU id 0, defaulting to 0.  Your kernel may not have been built with NUMA support.\n",
      "2023-08-29 14:48:08.497960: I tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc:982] could not open file to read NUMA node: /sys/bus/pci/devices/0000:01:00.0/numa_node\n",
      "Your kernel may have been built without NUMA support.\n",
      "2023-08-29 14:48:08.498015: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1635] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 12860 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 4080, pci bus id: 0000:01:00.0, compute capability: 8.9\n",
      "2023-08-29 14:48:17.948797: I tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:637] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once.\n",
      "Some weights of the PyTorch model were not used when initializing the TF 2.0 model TFEsmModel: ['esm.embeddings.position_ids', 'esm.embeddings.position_embeddings.weight', 'lm_head.bias', 'lm_head.layer_norm.weight', 'lm_head.dense.bias', 'lm_head.layer_norm.bias', 'lm_head.dense.weight']\n",
      "- This IS expected if you are initializing TFEsmModel from a PyTorch model trained on another task or with another architecture (e.g. initializing a TFBertForSequenceClassification model from a BertForPreTraining model).\n",
      "- This IS NOT expected if you are initializing TFEsmModel from a PyTorch model that you expect to be exactly identical (e.g. initializing a TFBertForSequenceClassification model from a BertForSequenceClassification model).\n",
      "Some weights or buffers of the TF 2.0 model TFEsmModel were not initialized from the PyTorch model and are newly initialized: ['esm.pooler.dense.weight', 'esm.pooler.dense.bias']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-08-29 14:48:23.457979: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'Placeholder/_2' with dtype int64 and shape [14412]\n",
      "\t [[{{node Placeholder/_2}}]]\n",
      "2023-08-29 14:48:23.458135: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'Placeholder/_1' with dtype int32 and shape [14412,612]\n",
      "\t [[{{node Placeholder/_1}}]]\n",
      "/home/ddofer/anaconda3/lib/python3.10/site-packages/keras/engine/functional.py:639: UserWarning: Input dict contained keys ['attention_mask'] which did not match any model input. They will be ignored by the model.\n",
      "  inputs = self._flatten_to_reference_inputs(inputs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model/esm/pooler/dense/kernel:0', 'tf_esm_model/esm/pooler/dense/bias:0', 'tf_esm_model/esm/contact_head/regression/kernel:0', 'tf_esm_model/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model/esm/pooler/dense/kernel:0', 'tf_esm_model/esm/pooler/dense/bias:0', 'tf_esm_model/esm/contact_head/regression/kernel:0', 'tf_esm_model/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model/esm/pooler/dense/kernel:0', 'tf_esm_model/esm/pooler/dense/bias:0', 'tf_esm_model/esm/contact_head/regression/kernel:0', 'tf_esm_model/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model/esm/pooler/dense/kernel:0', 'tf_esm_model/esm/pooler/dense/bias:0', 'tf_esm_model/esm/contact_head/regression/kernel:0', 'tf_esm_model/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-08-29 14:48:48.914691: I tensorflow/compiler/xla/service/service.cc:169] XLA service 0x7f0778042c30 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:\n",
      "2023-08-29 14:48:48.914718: I tensorflow/compiler/xla/service/service.cc:177]   StreamExecutor device (0): NVIDIA GeForce RTX 4080, Compute Capability 8.9\n",
      "2023-08-29 14:48:48.940393: I tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.cc:269] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.\n",
      "2023-08-29 14:48:49.126643: I tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:424] Loaded cuDNN version 8700\n",
      "2023-08-29 14:48:49.261339: I tensorflow/tsl/platform/default/subprocess.cc:304] Start cannot spawn child process: Permission denied\n",
      "2023-08-29 14:48:49.342976: I ./tensorflow/compiler/jit/device_compiler.h:180] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3603/3603 [==============================] - 1190s 318ms/step - loss: 0.0712 - accuracy: 0.8773 - auc: 0.8396\n",
      "Epoch 2/3\n",
      "3603/3603 [==============================] - 1142s 317ms/step - loss: 0.0418 - accuracy: 0.9371 - auc: 0.9154\n",
      "Epoch 3/3\n",
      "3603/3603 [==============================] - 1143s 317ms/step - loss: 0.0251 - accuracy: 0.9656 - auc: 0.9599\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-08-29 15:46:18.550357: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'Placeholder/_2' with dtype int64 and shape [4805]\n",
      "\t [[{{node Placeholder/_2}}]]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1202/1202 [==============================] - 177s 144ms/step\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of the PyTorch model were not used when initializing the TF 2.0 model TFEsmModel: ['esm.embeddings.position_ids', 'esm.embeddings.position_embeddings.weight', 'lm_head.bias', 'lm_head.layer_norm.weight', 'lm_head.dense.bias', 'lm_head.layer_norm.bias', 'lm_head.dense.weight']\n",
      "- This IS expected if you are initializing TFEsmModel from a PyTorch model trained on another task or with another architecture (e.g. initializing a TFBertForSequenceClassification model from a BertForPreTraining model).\n",
      "- This IS NOT expected if you are initializing TFEsmModel from a PyTorch model that you expect to be exactly identical (e.g. initializing a TFBertForSequenceClassification model from a BertForSequenceClassification model).\n",
      "Some weights or buffers of the TF 2.0 model TFEsmModel were not initialized from the PyTorch model and are newly initialized: ['esm.pooler.dense.weight', 'esm.pooler.dense.bias']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-08-29 15:49:37.328576: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'Placeholder/_2' with dtype int64 and shape [14413]\n",
      "\t [[{{node Placeholder/_2}}]]\n",
      "2023-08-29 15:49:37.328739: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'Placeholder/_2' with dtype int64 and shape [14413]\n",
      "\t [[{{node Placeholder/_2}}]]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model_1/esm/pooler/dense/kernel:0', 'tf_esm_model_1/esm/pooler/dense/bias:0', 'tf_esm_model_1/esm/contact_head/regression/kernel:0', 'tf_esm_model_1/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model_1/esm/pooler/dense/kernel:0', 'tf_esm_model_1/esm/pooler/dense/bias:0', 'tf_esm_model_1/esm/contact_head/regression/kernel:0', 'tf_esm_model_1/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model_1/esm/pooler/dense/kernel:0', 'tf_esm_model_1/esm/pooler/dense/bias:0', 'tf_esm_model_1/esm/contact_head/regression/kernel:0', 'tf_esm_model_1/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model_1/esm/pooler/dense/kernel:0', 'tf_esm_model_1/esm/pooler/dense/bias:0', 'tf_esm_model_1/esm/contact_head/regression/kernel:0', 'tf_esm_model_1/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "3604/3604 [==============================] - 1186s 317ms/step - loss: 0.0741 - accuracy: 0.8700 - auc_1: 0.8412\n",
      "Epoch 2/3\n",
      "3604/3604 [==============================] - 1159s 322ms/step - loss: 0.0463 - accuracy: 0.9286 - auc_1: 0.9112\n",
      "Epoch 3/3\n",
      "3604/3604 [==============================] - 1153s 320ms/step - loss: 0.0281 - accuracy: 0.9600 - auc_1: 0.9545\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-08-29 16:47:55.756473: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'Placeholder/_2' with dtype int64 and shape [4804]\n",
      "\t [[{{node Placeholder/_2}}]]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1201/1201 [==============================] - 183s 149ms/step\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of the PyTorch model were not used when initializing the TF 2.0 model TFEsmModel: ['esm.embeddings.position_ids', 'esm.embeddings.position_embeddings.weight', 'lm_head.bias', 'lm_head.layer_norm.weight', 'lm_head.dense.bias', 'lm_head.layer_norm.bias', 'lm_head.dense.weight']\n",
      "- This IS expected if you are initializing TFEsmModel from a PyTorch model trained on another task or with another architecture (e.g. initializing a TFBertForSequenceClassification model from a BertForPreTraining model).\n",
      "- This IS NOT expected if you are initializing TFEsmModel from a PyTorch model that you expect to be exactly identical (e.g. initializing a TFBertForSequenceClassification model from a BertForSequenceClassification model).\n",
      "Some weights or buffers of the TF 2.0 model TFEsmModel were not initialized from the PyTorch model and are newly initialized: ['esm.pooler.dense.weight', 'esm.pooler.dense.bias']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-08-29 16:51:20.780365: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'Placeholder/_1' with dtype int32 and shape [14413,612]\n",
      "\t [[{{node Placeholder/_1}}]]\n",
      "2023-08-29 16:51:20.780519: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'Placeholder/_2' with dtype int64 and shape [14413]\n",
      "\t [[{{node Placeholder/_2}}]]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model_2/esm/pooler/dense/kernel:0', 'tf_esm_model_2/esm/pooler/dense/bias:0', 'tf_esm_model_2/esm/contact_head/regression/kernel:0', 'tf_esm_model_2/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model_2/esm/pooler/dense/kernel:0', 'tf_esm_model_2/esm/pooler/dense/bias:0', 'tf_esm_model_2/esm/contact_head/regression/kernel:0', 'tf_esm_model_2/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model_2/esm/pooler/dense/kernel:0', 'tf_esm_model_2/esm/pooler/dense/bias:0', 'tf_esm_model_2/esm/contact_head/regression/kernel:0', 'tf_esm_model_2/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model_2/esm/pooler/dense/kernel:0', 'tf_esm_model_2/esm/pooler/dense/bias:0', 'tf_esm_model_2/esm/contact_head/regression/kernel:0', 'tf_esm_model_2/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "3604/3604 [==============================] - 1233s 329ms/step - loss: 0.0735 - accuracy: 0.8726 - auc_2: 0.8427\n",
      "Epoch 2/3\n",
      "3604/3604 [==============================] - 1156s 321ms/step - loss: 0.0426 - accuracy: 0.9374 - auc_2: 0.9137\n",
      "Epoch 3/3\n",
      "3604/3604 [==============================] - 1144s 317ms/step - loss: 0.0273 - accuracy: 0.9626 - auc_2: 0.9579\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-08-29 17:50:14.269492: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'Placeholder/_2' with dtype int64 and shape [4804]\n",
      "\t [[{{node Placeholder/_2}}]]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1201/1201 [==============================] - 176s 144ms/step\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of the PyTorch model were not used when initializing the TF 2.0 model TFEsmModel: ['esm.embeddings.position_ids', 'esm.embeddings.position_embeddings.weight', 'lm_head.bias', 'lm_head.layer_norm.weight', 'lm_head.dense.bias', 'lm_head.layer_norm.bias', 'lm_head.dense.weight']\n",
      "- This IS expected if you are initializing TFEsmModel from a PyTorch model trained on another task or with another architecture (e.g. initializing a TFBertForSequenceClassification model from a BertForPreTraining model).\n",
      "- This IS NOT expected if you are initializing TFEsmModel from a PyTorch model that you expect to be exactly identical (e.g. initializing a TFBertForSequenceClassification model from a BertForSequenceClassification model).\n",
      "Some weights or buffers of the TF 2.0 model TFEsmModel were not initialized from the PyTorch model and are newly initialized: ['esm.pooler.dense.weight', 'esm.pooler.dense.bias']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-08-29 17:53:32.560312: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'Placeholder/_1' with dtype int32 and shape [14413,612]\n",
      "\t [[{{node Placeholder/_1}}]]\n",
      "2023-08-29 17:53:32.560453: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'Placeholder/_1' with dtype int32 and shape [14413,612]\n",
      "\t [[{{node Placeholder/_1}}]]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model_3/esm/pooler/dense/kernel:0', 'tf_esm_model_3/esm/pooler/dense/bias:0', 'tf_esm_model_3/esm/contact_head/regression/kernel:0', 'tf_esm_model_3/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model_3/esm/pooler/dense/kernel:0', 'tf_esm_model_3/esm/pooler/dense/bias:0', 'tf_esm_model_3/esm/contact_head/regression/kernel:0', 'tf_esm_model_3/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model_3/esm/pooler/dense/kernel:0', 'tf_esm_model_3/esm/pooler/dense/bias:0', 'tf_esm_model_3/esm/contact_head/regression/kernel:0', 'tf_esm_model_3/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "WARNING:tensorflow:Gradients do not exist for variables ['tf_esm_model_3/esm/pooler/dense/kernel:0', 'tf_esm_model_3/esm/pooler/dense/bias:0', 'tf_esm_model_3/esm/contact_head/regression/kernel:0', 'tf_esm_model_3/esm/contact_head/regression/bias:0'] when minimizing the loss. If you're using `model.compile()`, did you forget to provide a `loss` argument?\n",
      "3604/3604 [==============================] - 1188s 317ms/step - loss: 0.0756 - accuracy: 0.8666 - auc_3: 0.8256\n",
      "Epoch 2/3\n",
      "3604/3604 [==============================] - 1143s 317ms/step - loss: 0.0410 - accuracy: 0.9387 - auc_3: 0.9174\n",
      "Epoch 3/3\n",
      "3604/3604 [==============================] - 1151s 319ms/step - loss: 0.0260 - accuracy: 0.9650 - auc_3: 0.9579\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-08-29 18:51:34.709612: I tensorflow/core/common_runtime/executor.cc:1197] [/device:CPU:0] (DEBUG INFO) Executor start aborting (this does not indicate an error and you can ignore this message): INVALID_ARGUMENT: You must feed a value for placeholder tensor 'Placeholder/_2' with dtype int64 and shape [4804]\n",
      "\t [[{{node Placeholder/_2}}]]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1201/1201 [==============================] - 180s 147ms/step\n",
      "Cross-validation complete. Predictions are stored in 'all_predictions'.\n"
     ]
    }
   ],
   "source": [
    "from sklearn.model_selection import StratifiedKFold\n",
    "import numpy as np\n",
    "import tensorflow as tf\n",
    "\n",
    "# Let's assume `all_sequences` and `all_labels` contain all your data points, including train and test.\n",
    "# 'groups' contains group labels for each data point to ensure stratification by groups.\n",
    "\n",
    "n_splits = 4  # Number of folds\n",
    "skf = StratifiedKFold(n_splits=n_splits)\n",
    "\n",
    "all_sequences = df_all[\"Sequence\"].tolist()\n",
    "all_labels = df_all[\"virus\"].tolist()\n",
    "groups = df_all[\"Cluster name\"].tolist()\n",
    "\n",
    "all_sequences = np.array(all_sequences)\n",
    "all_labels = np.array(all_labels)\n",
    "\n",
    "\n",
    "# Placeholder for predictions\n",
    "all_predictions = np.zeros(len(all_labels))\n",
    "\n",
    "# Loop through each fold\n",
    "for train_index, test_index in skf.split(all_sequences, all_labels, groups):\n",
    "    print(\"\\n\\n FOLD \\n\\n\")\n",
    "#     train_sequences, test_sequences = all_sequences[train_index], all_sequences[test_index]\n",
    "    train_labels, test_labels = all_labels[train_index], all_labels[test_index]\n",
    "    train_sequences = [all_sequences[i] for i in train_index]\n",
    "    test_sequences = [all_sequences[i] for i in test_index]\n",
    "\n",
    "    # Tokenize and prepare datasets\n",
    "    train_tokenized = tokenizer(train_sequences, padding=True, truncation=True, max_length=model_max_len)\n",
    "    test_tokenized = tokenizer(test_sequences, padding=True, truncation=True, max_length=model_max_len)\n",
    "\n",
    "    train_dataset = tf.data.Dataset.from_tensor_slices((\n",
    "        dict(train_tokenized),\n",
    "        train_labels\n",
    "    )).batch(batch_size_train).shuffle(buffer_size=len(train_sequences))\n",
    "\n",
    "    test_dataset = tf.data.Dataset.from_tensor_slices((\n",
    "        dict(test_tokenized),\n",
    "        test_labels\n",
    "    )).batch(batch_size_test)\n",
    "    \n",
    "    train_dataset = train_dataset.map(lambda x, y: ({'input_layer': x['input_ids'], 'attention_mask': x['attention_mask']}, y))\n",
    "    test_dataset = test_dataset.map(lambda x, y: ({'input_layer': x['input_ids'], 'attention_mask': x['attention_mask']}, y))\n",
    "\n",
    "    print(\"Dataset ready\")\n",
    "    # Load the pre-trained transformer model\n",
    "    base_model = TFAutoModel.from_pretrained(model_checkpoint)\n",
    "\n",
    "    # Freeze all layers except the last few\n",
    "    for layer in base_model.layers[:-8]: #[2:-5]:\n",
    "        layer.trainable = False\n",
    "\n",
    "    # Create new model on top\n",
    "    input_layer = Input(shape=(model_max_len,), dtype=tf.int32, name='input_layer')\n",
    "    sequence_output = base_model(input_layer)[0]\n",
    "    sequence_output = SpatialDropout1D(0.1)(sequence_output)\n",
    "    # Add GlobalMeanPooling1D and GlobalMaxPooling1D\n",
    "    avg_pool = GlobalAveragePooling1D()(sequence_output)\n",
    "    max_pool = GlobalMaxPooling1D()(sequence_output)\n",
    "    concat = Concatenate()([avg_pool, max_pool])\n",
    "#     concat = Dropout(0.05)(concat)\n",
    "    \n",
    "    # Add output layer for binary classification\n",
    "    output = Dense(1, activation='sigmoid', dtype='float32')(concat)\n",
    "\n",
    "    # Compile model\n",
    "    final_model = Model(inputs=input_layer, outputs=output)\n",
    "    optimizer = tf.keras.optimizers.Adam(learning_rate=learning_rate)\n",
    "    loss = \"BinaryFocalCrossentropy\"\n",
    "\n",
    "    final_model.compile(optimizer=optimizer, loss=loss, metrics=[\"accuracy\",tf.keras.metrics.AUC(num_thresholds=8,from_logits=True)])\n",
    "\n",
    "    # Train on the fold\n",
    "    final_model.fit(train_dataset, epochs=3)#, verbose=0)  # Turn off verbosity if needed\n",
    "\n",
    "    # Predict on the test fold\n",
    "    predictions = final_model.predict(test_dataset)\n",
    "    all_predictions[test_index] = predictions.squeeze()\n",
    "\n",
    "print(\"Cross-validation complete. Predictions are stored in 'all_predictions'.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T11:14:50.469955Z",
     "iopub.status.busy": "2023-08-29T11:14:50.469412Z",
     "iopub.status.idle": "2023-08-29T11:14:50.479888Z",
     "shell.execute_reply": "2023-08-29T11:14:50.478926Z",
     "shell.execute_reply.started": "2023-08-29T11:14:50.469913Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19217,)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_predictions.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T10:34:45.612789Z",
     "iopub.status.busy": "2023-08-29T10:34:45.612072Z",
     "iopub.status.idle": "2023-08-29T10:37:45.401049Z",
     "shell.execute_reply": "2023-08-29T10:37:45.399900Z",
     "shell.execute_reply.started": "2023-08-29T10:34:45.612754Z"
    },
    "id": "TgVhQQzYgy23"
   },
   "outputs": [],
   "source": [
    "# # model.save(\"trained_esm.keras\")\n",
    "# model.save(\"trained_esm_pool.tf\",save_format=\"tf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T10:40:41.405137Z",
     "iopub.status.busy": "2023-08-29T10:40:41.404750Z",
     "iopub.status.idle": "2023-08-29T10:40:41.421339Z",
     "shell.execute_reply": "2023-08-29T10:40:41.420323Z",
     "shell.execute_reply.started": "2023-08-29T10:40:41.405105Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.94      0.96      0.95     13691\n",
      "           1       0.90      0.84      0.87      5526\n",
      "\n",
      "    accuracy                           0.93     19217\n",
      "   macro avg       0.92      0.90      0.91     19217\n",
      "weighted avg       0.93      0.93      0.93     19217\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# print(classification_report(test_labels,y_test_pred.logits>=0.5))\n",
    "print(classification_report(df_all[\"virus\"],all_predictions>=0.5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T10:40:49.973855Z",
     "iopub.status.busy": "2023-08-29T10:40:49.972900Z",
     "iopub.status.idle": "2023-08-29T10:40:49.988950Z",
     "shell.execute_reply": "2023-08-29T10:40:49.987818Z",
     "shell.execute_reply.started": "2023-08-29T10:40:49.973811Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.973929221066181"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# roc_auc_score(test_labels,y_test_pred.logits)\n",
    "roc_auc_score(df_all[\"virus\"],all_predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all.drop(columns=[\"pred_score\"],inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T11:15:17.860319Z",
     "iopub.status.busy": "2023-08-29T11:15:17.859847Z",
     "iopub.status.idle": "2023-08-29T11:15:17.875817Z",
     "shell.execute_reply": "2023-08-29T11:15:17.874596Z",
     "shell.execute_reply.started": "2023-08-29T11:15:17.860278Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1398\n",
      "19217\n"
     ]
    }
   ],
   "source": [
    "y_test_binary_preds = (all_predictions>=0.5).astype(int).ravel()\n",
    "\n",
    "test_mistakes_mask = df_all[\"virus\"]!=y_test_binary_preds\n",
    "print(test_mistakes_mask.sum())\n",
    "print(len(test_mistakes_mask))\n",
    "df_all[\"mistake\"] = test_mistakes_mask\n",
    "\n",
    "df_all[\"pred_score_1\"] = all_predictions.round(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>virus</th>\n",
       "      <th>Length</th>\n",
       "      <th>Cluster name</th>\n",
       "      <th>mistake</th>\n",
       "      <th>pred_score_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1884</th>\n",
       "      <td>MISWLPSVAMGSRLLCCVALCLLGAGPADSGLTQTPRHLVKARGQQ...</td>\n",
       "      <td>1</td>\n",
       "      <td>321</td>\n",
       "      <td>Viral T-cell receptor beta chain-like T17T-22</td>\n",
       "      <td>True</td>\n",
       "      <td>0.012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1500</th>\n",
       "      <td>MKMAGWKKKLCPGHHLWALGCYMLLAVVSLRLSLRFKCDVDSLDLE...</td>\n",
       "      <td>1</td>\n",
       "      <td>440</td>\n",
       "      <td>Beta-1,3-galactosyl-O-glycosyl-glycoprotein be...</td>\n",
       "      <td>True</td>\n",
       "      <td>0.012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1949</th>\n",
       "      <td>NRTDVQCQHRWQKVLNPELNKGPWTKEEDQRVIEHVQKYGPKRWSD...</td>\n",
       "      <td>1</td>\n",
       "      <td>382</td>\n",
       "      <td>Transforming protein Myb</td>\n",
       "      <td>True</td>\n",
       "      <td>0.014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2744</th>\n",
       "      <td>MPLSVSLPSKNYDYDYDSVQPYFYFEEEEENFYLAAQQRSSELQPP...</td>\n",
       "      <td>1</td>\n",
       "      <td>412</td>\n",
       "      <td>Viral myc transforming protein</td>\n",
       "      <td>True</td>\n",
       "      <td>0.015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>MPLSASLPSKNYDYDYDSVQPYFYFEEEEENFYLAAQQRGSELQPP...</td>\n",
       "      <td>1</td>\n",
       "      <td>416</td>\n",
       "      <td>Viral myc transforming protein</td>\n",
       "      <td>True</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1598</th>\n",
       "      <td>MELQVVPVDFTAKATSQSLLPLLIALTVCFLSIIILIFVSEFLLYS...</td>\n",
       "      <td>1</td>\n",
       "      <td>94</td>\n",
       "      <td>Probable protein E5</td>\n",
       "      <td>True</td>\n",
       "      <td>0.497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12449</th>\n",
       "      <td>MSLQGIQLSDLSYKHAILKESQYTIKRDVGTTPAVTPSSLQREVSL...</td>\n",
       "      <td>1</td>\n",
       "      <td>394</td>\n",
       "      <td>Nucleoprotein</td>\n",
       "      <td>True</td>\n",
       "      <td>0.497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9485</th>\n",
       "      <td>MMICFVFLCVLTFVRGEIYPSTCPALGAGNGEAVRSGEMLLEISAY...</td>\n",
       "      <td>1</td>\n",
       "      <td>433</td>\n",
       "      <td>U21 glycoprotein</td>\n",
       "      <td>True</td>\n",
       "      <td>0.499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11993</th>\n",
       "      <td>MAFYLPDWSCCGLWLFGRPRNRYSQLPEEPETFECPDRWRAEIDLG...</td>\n",
       "      <td>1</td>\n",
       "      <td>218</td>\n",
       "      <td>Tegument protein UL51 homolog</td>\n",
       "      <td>True</td>\n",
       "      <td>0.499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3043</th>\n",
       "      <td>MKIVGADGQEQEETDIPFRLWRKFAARRKLQYQSWEEGKEVLLNKL...</td>\n",
       "      <td>1</td>\n",
       "      <td>217</td>\n",
       "      <td>U exon protein</td>\n",
       "      <td>True</td>\n",
       "      <td>0.500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>886 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Sequence  virus  Length  \\\n",
       "1884   MISWLPSVAMGSRLLCCVALCLLGAGPADSGLTQTPRHLVKARGQQ...      1     321   \n",
       "1500   MKMAGWKKKLCPGHHLWALGCYMLLAVVSLRLSLRFKCDVDSLDLE...      1     440   \n",
       "1949   NRTDVQCQHRWQKVLNPELNKGPWTKEEDQRVIEHVQKYGPKRWSD...      1     382   \n",
       "2744   MPLSVSLPSKNYDYDYDSVQPYFYFEEEEENFYLAAQQRSSELQPP...      1     412   \n",
       "817    MPLSASLPSKNYDYDYDSVQPYFYFEEEEENFYLAAQQRGSELQPP...      1     416   \n",
       "...                                                  ...    ...     ...   \n",
       "1598   MELQVVPVDFTAKATSQSLLPLLIALTVCFLSIIILIFVSEFLLYS...      1      94   \n",
       "12449  MSLQGIQLSDLSYKHAILKESQYTIKRDVGTTPAVTPSSLQREVSL...      1     394   \n",
       "9485   MMICFVFLCVLTFVRGEIYPSTCPALGAGNGEAVRSGEMLLEISAY...      1     433   \n",
       "11993  MAFYLPDWSCCGLWLFGRPRNRYSQLPEEPETFECPDRWRAEIDLG...      1     218   \n",
       "3043   MKIVGADGQEQEETDIPFRLWRKFAARRKLQYQSWEEGKEVLLNKL...      1     217   \n",
       "\n",
       "                                            Cluster name  mistake  \\\n",
       "1884       Viral T-cell receptor beta chain-like T17T-22     True   \n",
       "1500   Beta-1,3-galactosyl-O-glycosyl-glycoprotein be...     True   \n",
       "1949                            Transforming protein Myb     True   \n",
       "2744                      Viral myc transforming protein     True   \n",
       "817                       Viral myc transforming protein     True   \n",
       "...                                                  ...      ...   \n",
       "1598                                 Probable protein E5     True   \n",
       "12449                                      Nucleoprotein     True   \n",
       "9485                                    U21 glycoprotein     True   \n",
       "11993                      Tegument protein UL51 homolog     True   \n",
       "3043                                      U exon protein     True   \n",
       "\n",
       "       pred_score_1  \n",
       "1884          0.012  \n",
       "1500          0.012  \n",
       "1949          0.014  \n",
       "2744          0.015  \n",
       "817           0.016  \n",
       "...             ...  \n",
       "1598          0.497  \n",
       "12449         0.497  \n",
       "9485          0.499  \n",
       "11993         0.499  \n",
       "3043          0.500  \n",
       "\n",
       "[886 rows x 6 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all[test_mistakes_mask & (df_all[\"virus\"]==1)].sort_values(\"pred_score_1\",ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all.loc[df_all[\"virus\"]==0,\"pred_score\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>virus</th>\n",
       "      <th>Length</th>\n",
       "      <th>Cluster name</th>\n",
       "      <th>mistake</th>\n",
       "      <th>pred_score_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>MECLYYFLGFLLLAARLPLDAAKRFHDVLGNERPSAYMREHNQLNG...</td>\n",
       "      <td>0</td>\n",
       "      <td>572</td>\n",
       "      <td>Transmembrane glycoprotein NMB</td>\n",
       "      <td>True</td>\n",
       "      <td>0.933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9662</th>\n",
       "      <td>MEAAAQFFVESPDVVYGPEAIEAQYEYRTTRVSREGGVLKVHPTST...</td>\n",
       "      <td>0</td>\n",
       "      <td>558</td>\n",
       "      <td>Inositol-3-phosphate synthase 1</td>\n",
       "      <td>True</td>\n",
       "      <td>0.926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1705</th>\n",
       "      <td>MHVAEVAVNVILLLSMGWTSDSLCSPTIFYRDCWIRRFPGLLINLE...</td>\n",
       "      <td>0</td>\n",
       "      <td>340</td>\n",
       "      <td>MANSC domain-containing protein 4</td>\n",
       "      <td>True</td>\n",
       "      <td>0.917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9450</th>\n",
       "      <td>MWTVPSFTNDSYQVYNVFSTNSFQLLTVKRTPHEAWRVPLTTKTNK...</td>\n",
       "      <td>0</td>\n",
       "      <td>482</td>\n",
       "      <td>Endogenous retrovirus group K member 13-1 Env ...</td>\n",
       "      <td>True</td>\n",
       "      <td>0.914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2334</th>\n",
       "      <td>MQTLRQEAARPCIPSGTLEASFPAPLYSDDYLSLEGSRWPPAIRQA...</td>\n",
       "      <td>0</td>\n",
       "      <td>209</td>\n",
       "      <td>Uncharacterized protein C19orf71</td>\n",
       "      <td>True</td>\n",
       "      <td>0.913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9187</th>\n",
       "      <td>MGVPRTPSRTVLFERERTGLTYRVPSLLPVPPGPTLLAFVEQRLSP...</td>\n",
       "      <td>0</td>\n",
       "      <td>484</td>\n",
       "      <td>Sialidase-4</td>\n",
       "      <td>True</td>\n",
       "      <td>0.503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1348</th>\n",
       "      <td>MADLEEQLSDEEKVRIAAKFIIHAPPGEFNEVFNDVRLLLNNDNLL...</td>\n",
       "      <td>0</td>\n",
       "      <td>286</td>\n",
       "      <td>F-actin-capping protein subunit alpha-2</td>\n",
       "      <td>True</td>\n",
       "      <td>0.503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14436</th>\n",
       "      <td>MGSILSRRIAGVEDIDIQANSAYRYPPKSGNYFASHFFMGGEKFDT...</td>\n",
       "      <td>0</td>\n",
       "      <td>552</td>\n",
       "      <td>E3 ubiquitin-protein ligase MGRN1</td>\n",
       "      <td>True</td>\n",
       "      <td>0.502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7299</th>\n",
       "      <td>MDTQTHSLPITHTQLHSNSQPQSRTCTRHCQTFSQSCRQSHRGSRS...</td>\n",
       "      <td>0</td>\n",
       "      <td>138</td>\n",
       "      <td>Nuclear transition protein 2</td>\n",
       "      <td>True</td>\n",
       "      <td>0.502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15227</th>\n",
       "      <td>MGKIALQLKATLENITNLRPVGEDFRWYLKMKCGNCGEISDKWQYI...</td>\n",
       "      <td>0</td>\n",
       "      <td>160</td>\n",
       "      <td>CXXC motif containing zinc binding protein</td>\n",
       "      <td>True</td>\n",
       "      <td>0.501</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>512 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Sequence  virus  Length  \\\n",
       "80     MECLYYFLGFLLLAARLPLDAAKRFHDVLGNERPSAYMREHNQLNG...      0     572   \n",
       "9662   MEAAAQFFVESPDVVYGPEAIEAQYEYRTTRVSREGGVLKVHPTST...      0     558   \n",
       "1705   MHVAEVAVNVILLLSMGWTSDSLCSPTIFYRDCWIRRFPGLLINLE...      0     340   \n",
       "9450   MWTVPSFTNDSYQVYNVFSTNSFQLLTVKRTPHEAWRVPLTTKTNK...      0     482   \n",
       "2334   MQTLRQEAARPCIPSGTLEASFPAPLYSDDYLSLEGSRWPPAIRQA...      0     209   \n",
       "...                                                  ...    ...     ...   \n",
       "9187   MGVPRTPSRTVLFERERTGLTYRVPSLLPVPPGPTLLAFVEQRLSP...      0     484   \n",
       "1348   MADLEEQLSDEEKVRIAAKFIIHAPPGEFNEVFNDVRLLLNNDNLL...      0     286   \n",
       "14436  MGSILSRRIAGVEDIDIQANSAYRYPPKSGNYFASHFFMGGEKFDT...      0     552   \n",
       "7299   MDTQTHSLPITHTQLHSNSQPQSRTCTRHCQTFSQSCRQSHRGSRS...      0     138   \n",
       "15227  MGKIALQLKATLENITNLRPVGEDFRWYLKMKCGNCGEISDKWQYI...      0     160   \n",
       "\n",
       "                                            Cluster name  mistake  \\\n",
       "80                        Transmembrane glycoprotein NMB     True   \n",
       "9662                     Inositol-3-phosphate synthase 1     True   \n",
       "1705                   MANSC domain-containing protein 4     True   \n",
       "9450   Endogenous retrovirus group K member 13-1 Env ...     True   \n",
       "2334                    Uncharacterized protein C19orf71     True   \n",
       "...                                                  ...      ...   \n",
       "9187                                         Sialidase-4     True   \n",
       "1348             F-actin-capping protein subunit alpha-2     True   \n",
       "14436                  E3 ubiquitin-protein ligase MGRN1     True   \n",
       "7299                        Nuclear transition protein 2     True   \n",
       "15227         CXXC motif containing zinc binding protein     True   \n",
       "\n",
       "       pred_score_1  \n",
       "80            0.933  \n",
       "9662          0.926  \n",
       "1705          0.917  \n",
       "9450          0.914  \n",
       "2334          0.913  \n",
       "...             ...  \n",
       "9187          0.503  \n",
       "1348          0.503  \n",
       "14436         0.502  \n",
       "7299          0.502  \n",
       "15227         0.501  \n",
       "\n",
       "[512 rows x 6 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all[test_mistakes_mask & (df_all[\"virus\"]==0)].sort_values(\"pred_score_1\",ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T11:15:18.037823Z",
     "iopub.status.busy": "2023-08-29T11:15:18.037436Z",
     "iopub.status.idle": "2023-08-29T11:15:18.049974Z",
     "shell.execute_reply": "2023-08-29T11:15:18.048990Z",
     "shell.execute_reply.started": "2023-08-29T11:15:18.037791Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>virus</th>\n",
       "      <th>Length</th>\n",
       "      <th>Cluster name</th>\n",
       "      <th>mistake</th>\n",
       "      <th>pred_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MNPSEMQRKAPPRRRRHRNRAPSSHKMNKMMMSEEQMKLPSTNKAE...</td>\n",
       "      <td>0</td>\n",
       "      <td>105</td>\n",
       "      <td>Endogenous retrovirus group K member 16 Rec pr...</td>\n",
       "      <td>True</td>\n",
       "      <td>0.7947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MEGEKNWIVVPTWRIPGRLEKWHSLVKYLKHRTKELQQVSYVPHHK...</td>\n",
       "      <td>1</td>\n",
       "      <td>216</td>\n",
       "      <td>Virion infectivity factor</td>\n",
       "      <td>True</td>\n",
       "      <td>0.2859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MDHGKYLLTIFLNDDDSFFKYLAAQEDDVAMSDVHTIVDYLNFLLA...</td>\n",
       "      <td>1</td>\n",
       "      <td>148</td>\n",
       "      <td>MF7 protein</td>\n",
       "      <td>True</td>\n",
       "      <td>0.3750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>MFDKIVIPRQFHHQHKWDKLLSKCPHPGVGFYIIGNKKALSARMRT...</td>\n",
       "      <td>1</td>\n",
       "      <td>205</td>\n",
       "      <td>Probable DNA-binding protein</td>\n",
       "      <td>True</td>\n",
       "      <td>0.0309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MDYVSLLNQFWQKQIKSYKETPSQYHYLYPPRFFYKPVLGNLQHPT...</td>\n",
       "      <td>1</td>\n",
       "      <td>119</td>\n",
       "      <td>Non-structural protein 3b</td>\n",
       "      <td>True</td>\n",
       "      <td>0.3432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3579</th>\n",
       "      <td>MNIRGAPDLGQPSDDPSSGGERERIRQRMKMVIGQLEGILRELKEV...</td>\n",
       "      <td>0</td>\n",
       "      <td>293</td>\n",
       "      <td>Inhibitory synaptic factor 1</td>\n",
       "      <td>True</td>\n",
       "      <td>0.5399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3580</th>\n",
       "      <td>MKSIRQLLSLAKKEKKREKKSNHGSHSMEWESPPSYNEIKSPSAPI...</td>\n",
       "      <td>1</td>\n",
       "      <td>229</td>\n",
       "      <td>Matrix protein</td>\n",
       "      <td>True</td>\n",
       "      <td>0.2237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3594</th>\n",
       "      <td>MSFLGGFFGPICEIDIVLNDGETRKMAEMKTEDGKVEKHYLFYDGE...</td>\n",
       "      <td>0</td>\n",
       "      <td>327</td>\n",
       "      <td>Vacuolar protein sorting-associated protein 26A</td>\n",
       "      <td>True</td>\n",
       "      <td>0.6473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3633</th>\n",
       "      <td>MYKCSQGAMNTEKVMEKFVIQSRFREMYPDKAKAIAGMTVPARYAD...</td>\n",
       "      <td>1</td>\n",
       "      <td>90</td>\n",
       "      <td>Uncharacterized protein 045R</td>\n",
       "      <td>True</td>\n",
       "      <td>0.4516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3648</th>\n",
       "      <td>MADTQCCPPPCEFISSAGTDLALGMGWDATLCLLPFTGFGKCAGIW...</td>\n",
       "      <td>0</td>\n",
       "      <td>198</td>\n",
       "      <td>Putative uncharacterized protein encoded by AG...</td>\n",
       "      <td>True</td>\n",
       "      <td>0.7281</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1398 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Sequence  virus  Length  \\\n",
       "7     MNPSEMQRKAPPRRRRHRNRAPSSHKMNKMMMSEEQMKLPSTNKAE...      0     105   \n",
       "9     MEGEKNWIVVPTWRIPGRLEKWHSLVKYLKHRTKELQQVSYVPHHK...      1     216   \n",
       "14    MDHGKYLLTIFLNDDDSFFKYLAAQEDDVAMSDVHTIVDYLNFLLA...      1     148   \n",
       "19    MFDKIVIPRQFHHQHKWDKLLSKCPHPGVGFYIIGNKKALSARMRT...      1     205   \n",
       "21    MDYVSLLNQFWQKQIKSYKETPSQYHYLYPPRFFYKPVLGNLQHPT...      1     119   \n",
       "...                                                 ...    ...     ...   \n",
       "3579  MNIRGAPDLGQPSDDPSSGGERERIRQRMKMVIGQLEGILRELKEV...      0     293   \n",
       "3580  MKSIRQLLSLAKKEKKREKKSNHGSHSMEWESPPSYNEIKSPSAPI...      1     229   \n",
       "3594  MSFLGGFFGPICEIDIVLNDGETRKMAEMKTEDGKVEKHYLFYDGE...      0     327   \n",
       "3633  MYKCSQGAMNTEKVMEKFVIQSRFREMYPDKAKAIAGMTVPARYAD...      1      90   \n",
       "3648  MADTQCCPPPCEFISSAGTDLALGMGWDATLCLLPFTGFGKCAGIW...      0     198   \n",
       "\n",
       "                                           Cluster name  mistake  pred_score  \n",
       "7     Endogenous retrovirus group K member 16 Rec pr...     True      0.7947  \n",
       "9                             Virion infectivity factor     True      0.2859  \n",
       "14                                          MF7 protein     True      0.3750  \n",
       "19                         Probable DNA-binding protein     True      0.0309  \n",
       "21                            Non-structural protein 3b     True      0.3432  \n",
       "...                                                 ...      ...         ...  \n",
       "3579                       Inhibitory synaptic factor 1     True      0.5399  \n",
       "3580                                     Matrix protein     True      0.2237  \n",
       "3594    Vacuolar protein sorting-associated protein 26A     True      0.6473  \n",
       "3633                       Uncharacterized protein 045R     True      0.4516  \n",
       "3648  Putative uncharacterized protein encoded by AG...     True      0.7281  \n",
       "\n",
       "[1398 rows x 6 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all[test_mistakes_mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T11:15:18.207595Z",
     "iopub.status.busy": "2023-08-29T11:15:18.206596Z",
     "iopub.status.idle": "2023-08-29T11:15:18.216701Z",
     "shell.execute_reply": "2023-08-29T11:15:18.215699Z",
     "shell.execute_reply.started": "2023-08-29T11:15:18.207562Z"
    }
   },
   "outputs": [],
   "source": [
    "df_all.to_csv(\"data_with_mistakes.csv.gz\",index=False,compression=\"gzip\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-08-29T11:15:20.486146Z",
     "iopub.status.busy": "2023-08-29T11:15:20.485763Z",
     "iopub.status.idle": "2023-08-29T11:15:20.504541Z",
     "shell.execute_reply": "2023-08-29T11:15:20.503476Z",
     "shell.execute_reply.started": "2023-08-29T11:15:20.486115Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>virus</th>\n",
       "      <th>Length</th>\n",
       "      <th>Cluster name</th>\n",
       "      <th>mistake</th>\n",
       "      <th>pred_score_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MADFLKGLPVYNKSNFSRFHADSVCKASNRRPSVYLPTREYPSEQI...</td>\n",
       "      <td>0</td>\n",
       "      <td>102</td>\n",
       "      <td>DET1- and DDB1-associated protein 1</td>\n",
       "      <td>False</td>\n",
       "      <td>0.232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MPSTLQVLAKKVLALEHKENDHISREYYYHILKCCGLWWHEAPIIL...</td>\n",
       "      <td>1</td>\n",
       "      <td>362</td>\n",
       "      <td>Protein MGF 360-19R</td>\n",
       "      <td>False</td>\n",
       "      <td>0.942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MASSAELDFNLQALLEQLSQDELSKFKSLIRTISLGKELQTVPQTE...</td>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>Pyrin domain-containing protein 2</td>\n",
       "      <td>False</td>\n",
       "      <td>0.134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAAWGKKHAGKDPVRDECEERNRFTETREEDVTDEHGEREPFAETD...</td>\n",
       "      <td>0</td>\n",
       "      <td>186</td>\n",
       "      <td>Protein FAM9B</td>\n",
       "      <td>False</td>\n",
       "      <td>0.153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MADPAAAQNYPLLKLLHSYTPTTPPRPIPKPAPWAPQKPRRQITND...</td>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "      <td>Protein E4</td>\n",
       "      <td>False</td>\n",
       "      <td>0.533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3677</th>\n",
       "      <td>MAAGDGDVKLGTLGSGSESSNDGGSESPGDAGAAAEGGGWAAAALA...</td>\n",
       "      <td>0</td>\n",
       "      <td>223</td>\n",
       "      <td>Membrane-associated progesterone receptor comp...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3678</th>\n",
       "      <td>MTLPASLKGFETLDIFNISSVRNICELNKHAEHVSIKNLPLIDISV...</td>\n",
       "      <td>1</td>\n",
       "      <td>256</td>\n",
       "      <td>Cytoplasmic envelopment protein 1</td>\n",
       "      <td>False</td>\n",
       "      <td>0.916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3679</th>\n",
       "      <td>MKWKALFTAAILQAQLPITEAQSFGLLDPKLCYLLDGILFIYGVIL...</td>\n",
       "      <td>0</td>\n",
       "      <td>164</td>\n",
       "      <td>T-cell surface glycoprotein CD3 zeta chain</td>\n",
       "      <td>False</td>\n",
       "      <td>0.089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3680</th>\n",
       "      <td>MAGVGDAAAPGEGGGGGVDGPQRDGRGEAEQPGGSGGQGPPPAPQL...</td>\n",
       "      <td>0</td>\n",
       "      <td>567</td>\n",
       "      <td>Transmembrane anterior posterior transformatio...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3681</th>\n",
       "      <td>MGASVSRGRAARVPAPEPEPEEALDLSQLPPELLLVVLSHVPPRTL...</td>\n",
       "      <td>0</td>\n",
       "      <td>283</td>\n",
       "      <td>F-box only protein 27</td>\n",
       "      <td>False</td>\n",
       "      <td>0.141</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19217 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Sequence  virus  Length  \\\n",
       "0     MADFLKGLPVYNKSNFSRFHADSVCKASNRRPSVYLPTREYPSEQI...      0     102   \n",
       "1     MPSTLQVLAKKVLALEHKENDHISREYYYHILKCCGLWWHEAPIIL...      1     362   \n",
       "2     MASSAELDFNLQALLEQLSQDELSKFKSLIRTISLGKELQTVPQTE...      0      97   \n",
       "3     MAAWGKKHAGKDPVRDECEERNRFTETREEDVTDEHGEREPFAETD...      0     186   \n",
       "4     MADPAAAQNYPLLKLLHSYTPTTPPRPIPKPAPWAPQKPRRQITND...      1      93   \n",
       "...                                                 ...    ...     ...   \n",
       "3677  MAAGDGDVKLGTLGSGSESSNDGGSESPGDAGAAAEGGGWAAAALA...      0     223   \n",
       "3678  MTLPASLKGFETLDIFNISSVRNICELNKHAEHVSIKNLPLIDISV...      1     256   \n",
       "3679  MKWKALFTAAILQAQLPITEAQSFGLLDPKLCYLLDGILFIYGVIL...      0     164   \n",
       "3680  MAGVGDAAAPGEGGGGGVDGPQRDGRGEAEQPGGSGGQGPPPAPQL...      0     567   \n",
       "3681  MGASVSRGRAARVPAPEPEPEEALDLSQLPPELLLVVLSHVPPRTL...      0     283   \n",
       "\n",
       "                                           Cluster name  mistake  pred_score_1  \n",
       "0                   DET1- and DDB1-associated protein 1    False         0.232  \n",
       "1                                   Protein MGF 360-19R    False         0.942  \n",
       "2                     Pyrin domain-containing protein 2    False         0.134  \n",
       "3                                         Protein FAM9B    False         0.153  \n",
       "4                                            Protein E4    False         0.533  \n",
       "...                                                 ...      ...           ...  \n",
       "3677  Membrane-associated progesterone receptor comp...    False         0.041  \n",
       "3678                  Cytoplasmic envelopment protein 1    False         0.916  \n",
       "3679         T-cell surface glycoprotein CD3 zeta chain    False         0.089  \n",
       "3680  Transmembrane anterior posterior transformatio...    False         0.047  \n",
       "3681                              F-box only protein 27    False         0.141  \n",
       "\n",
       "[19217 rows x 6 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sequence        0\n",
       "virus           0\n",
       "Length          0\n",
       "Cluster name    0\n",
       "mistake         0\n",
       "pred_score_1    0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pd.read_csv(\"SWP_human_or_virus-animal.tsv\",sep=\"\\t\",compression=\"gzip\").to_csv(\"SWP_human_or_virus-animal.csv.gz\",index=False,compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>virus</th>\n",
       "      <th>Length</th>\n",
       "      <th>Cluster name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MADFLKGLPVYNKSNFSRFHADSVCKASNRRPSVYLPTREYPSEQI...</td>\n",
       "      <td>0</td>\n",
       "      <td>102</td>\n",
       "      <td>DET1- and DDB1-associated protein 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MPSTLQVLAKKVLALEHKENDHISREYYYHILKCCGLWWHEAPIIL...</td>\n",
       "      <td>1</td>\n",
       "      <td>362</td>\n",
       "      <td>Protein MGF 360-19R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MASSAELDFNLQALLEQLSQDELSKFKSLIRTISLGKELQTVPQTE...</td>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>Pyrin domain-containing protein 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAAWGKKHAGKDPVRDECEERNRFTETREEDVTDEHGEREPFAETD...</td>\n",
       "      <td>0</td>\n",
       "      <td>186</td>\n",
       "      <td>Protein FAM9B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MADPAAAQNYPLLKLLHSYTPTTPPRPIPKPAPWAPQKPRRQITND...</td>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "      <td>Protein E4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3677</th>\n",
       "      <td>MAAGDGDVKLGTLGSGSESSNDGGSESPGDAGAAAEGGGWAAAALA...</td>\n",
       "      <td>0</td>\n",
       "      <td>223</td>\n",
       "      <td>Membrane-associated progesterone receptor comp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3678</th>\n",
       "      <td>MTLPASLKGFETLDIFNISSVRNICELNKHAEHVSIKNLPLIDISV...</td>\n",
       "      <td>1</td>\n",
       "      <td>256</td>\n",
       "      <td>Cytoplasmic envelopment protein 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3679</th>\n",
       "      <td>MKWKALFTAAILQAQLPITEAQSFGLLDPKLCYLLDGILFIYGVIL...</td>\n",
       "      <td>0</td>\n",
       "      <td>164</td>\n",
       "      <td>T-cell surface glycoprotein CD3 zeta chain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3680</th>\n",
       "      <td>MAGVGDAAAPGEGGGGGVDGPQRDGRGEAEQPGGSGGQGPPPAPQL...</td>\n",
       "      <td>0</td>\n",
       "      <td>567</td>\n",
       "      <td>Transmembrane anterior posterior transformatio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3681</th>\n",
       "      <td>MGASVSRGRAARVPAPEPEPEEALDLSQLPPELLLVVLSHVPPRTL...</td>\n",
       "      <td>0</td>\n",
       "      <td>283</td>\n",
       "      <td>F-box only protein 27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19217 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Sequence  virus  Length  \\\n",
       "0     MADFLKGLPVYNKSNFSRFHADSVCKASNRRPSVYLPTREYPSEQI...      0     102   \n",
       "1     MPSTLQVLAKKVLALEHKENDHISREYYYHILKCCGLWWHEAPIIL...      1     362   \n",
       "2     MASSAELDFNLQALLEQLSQDELSKFKSLIRTISLGKELQTVPQTE...      0      97   \n",
       "3     MAAWGKKHAGKDPVRDECEERNRFTETREEDVTDEHGEREPFAETD...      0     186   \n",
       "4     MADPAAAQNYPLLKLLHSYTPTTPPRPIPKPAPWAPQKPRRQITND...      1      93   \n",
       "...                                                 ...    ...     ...   \n",
       "3677  MAAGDGDVKLGTLGSGSESSNDGGSESPGDAGAAAEGGGWAAAALA...      0     223   \n",
       "3678  MTLPASLKGFETLDIFNISSVRNICELNKHAEHVSIKNLPLIDISV...      1     256   \n",
       "3679  MKWKALFTAAILQAQLPITEAQSFGLLDPKLCYLLDGILFIYGVIL...      0     164   \n",
       "3680  MAGVGDAAAPGEGGGGGVDGPQRDGRGEAEQPGGSGGQGPPPAPQL...      0     567   \n",
       "3681  MGASVSRGRAARVPAPEPEPEEALDLSQLPPELLLVVLSHVPPRTL...      0     283   \n",
       "\n",
       "                                           Cluster name  \n",
       "0                   DET1- and DDB1-associated protein 1  \n",
       "1                                   Protein MGF 360-19R  \n",
       "2                     Pyrin domain-containing protein 2  \n",
       "3                                         Protein FAM9B  \n",
       "4                                            Protein E4  \n",
       "...                                                 ...  \n",
       "3677  Membrane-associated progesterone receptor comp...  \n",
       "3678                  Cytoplasmic envelopment protein 1  \n",
       "3679         T-cell surface glycoprotein CD3 zeta chain  \n",
       "3680  Transmembrane anterior posterior transformatio...  \n",
       "3681                              F-box only protein 27  \n",
       "\n",
       "[19217 rows x 4 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence             25119\n",
      "Protein names        21394\n",
      "Organism              1136\n",
      "Mass                 22130\n",
      "Keywords             15980\n",
      "av_mw                 2110\n",
      "Virus hosts            291\n",
      "Taxonomic lineage      561\n",
      "UR50_Cluster ID      20886\n",
      "dtype: int64\n",
      "0.16851785501015168\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Mass</th>\n",
       "      <th>Keywords</th>\n",
       "      <th>av_mw</th>\n",
       "      <th>Virus hosts</th>\n",
       "      <th>Taxonomic lineage</th>\n",
       "      <th>UR50_Cluster ID</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGLEALVPLAMIVAIFLLLVDLMHRHQRWAARYPPGPLPLPGLGNL...</td>\n",
       "      <td>Putative cytochrome P450 2D7 (EC 1.14.14.1)</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>57489</td>\n",
       "      <td>Cytoplasm . Glycoprotein . Heme . Iron . Membr...</td>\n",
       "      <td>111.63</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Putative cytochrome P450 2D7 (EC 1.14.14.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFRRLTFAQLLFATVLGIAGGVYIFQPVFEQYAKDQKELKEKMQLV...</td>\n",
       "      <td>Protein PIGBOS1 (PIGB opposite strand protein 1)</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>6313</td>\n",
       "      <td>Direct protein sequencing . Membrane . Mitocho...</td>\n",
       "      <td>116.91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>A0A0B4J2F0</td>\n",
       "      <td>A0A0B4J2F0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MRWQEMGYIFYPRKLR</td>\n",
       "      <td>Mitochondrial-derived peptide MOTS-c (Mitochon...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>2175</td>\n",
       "      <td>DNA-binding . Mitochondrion . Nucleus . Osteog...</td>\n",
       "      <td>135.94</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MTQRAGAAMLPSALLLLCVPGCLTVSGPSTVMGAVGESLSVQCRYE...</td>\n",
       "      <td>Protein CD300H (CD300 antigen-like family memb...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>21806</td>\n",
       "      <td>Alternative splicing . Disulfide bond . Glycop...</td>\n",
       "      <td>108.49</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>A0A0K2S4Q6</td>\n",
       "      <td>A0A0K2S4Q6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGDQPCASGRSTLPPGNAREAKPPKKRCLLAPRWDYPEGTPNGGST...</td>\n",
       "      <td>Negative regulator of P-body association (P-bo...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>7025</td>\n",
       "      <td>Cytoplasm . mRNA processing . Reference proteome</td>\n",
       "      <td>103.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>A0A0U1RRE5</td>\n",
       "      <td>A0A0U1RRE5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25114</th>\n",
       "      <td>MKTSPRLSVIDGVGEEVRPDLFQLVHVVPPNTCEPPEQWLRVTVLL...</td>\n",
       "      <td>Uncharacterized protein ORF6</td>\n",
       "      <td>Beak and feather disease virus (BFDV)</td>\n",
       "      <td>9846</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111.89</td>\n",
       "      <td>Gracula [TaxID: 116991]; Psittaciformes [TaxID...</td>\n",
       "      <td>Viruses, Monodnaviria, Shotokuvirae (kingdom),...</td>\n",
       "      <td>Q9YUD1</td>\n",
       "      <td>Q9YUD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25115</th>\n",
       "      <td>MGGAYITSRYWLVPAHVTSRLRLTSFTGHRGVARVDGPMSSLGLNI...</td>\n",
       "      <td>Uncharacterized protein ORF5</td>\n",
       "      <td>Beak and feather disease virus (BFDV)</td>\n",
       "      <td>10863</td>\n",
       "      <td>NaN</td>\n",
       "      <td>107.55</td>\n",
       "      <td>Gracula [TaxID: 116991]; Psittaciformes [TaxID...</td>\n",
       "      <td>Viruses, Monodnaviria, Shotokuvirae (kingdom),...</td>\n",
       "      <td>Q9YUD2</td>\n",
       "      <td>Q9YUD2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25116</th>\n",
       "      <td>MPYVDFWELSRDFLHFASGHHGSNSSVERAVTVSRHNAQGEYNVPL...</td>\n",
       "      <td>Uncharacterized protein ORF3</td>\n",
       "      <td>Beak and feather disease virus (BFDV)</td>\n",
       "      <td>17738</td>\n",
       "      <td>NaN</td>\n",
       "      <td>110.86</td>\n",
       "      <td>Gracula [TaxID: 116991]; Psittaciformes [TaxID...</td>\n",
       "      <td>Viruses, Monodnaviria, Shotokuvirae (kingdom),...</td>\n",
       "      <td>Q9YUD4</td>\n",
       "      <td>Q9YUD4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25117</th>\n",
       "      <td>MNNGQATITRNGGRFEIRCRHLDRDYTMPLPNATSNDNFLDCIKFI...</td>\n",
       "      <td>Non-structural protein 3 (NS3)</td>\n",
       "      <td>Banna virus (strain Indonesia/JKT-6423/1980) (...</td>\n",
       "      <td>34991</td>\n",
       "      <td>Reference proteome</td>\n",
       "      <td>114.35</td>\n",
       "      <td>Bos taurus (Bovine) [TaxID: 9913]; Culex annul...</td>\n",
       "      <td>Viruses, Riboviria, Orthornavirae (kingdom), D...</td>\n",
       "      <td>Q9YWQ0</td>\n",
       "      <td>Q9YWQ0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25118</th>\n",
       "      <td>MGRKEHESPSQPHMCGWEDSQKPSVPSHGPKTPSCKGVKAPHSSRP...</td>\n",
       "      <td>Putative uncharacterized protein GUCA1ANB (GUC...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>14531</td>\n",
       "      <td>Reference proteome</td>\n",
       "      <td>114.42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms, Eukaryota, Opisthokonta, M...</td>\n",
       "      <td>X6R8D5</td>\n",
       "      <td>X6R8D5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25119 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Sequence  \\\n",
       "0      MGLEALVPLAMIVAIFLLLVDLMHRHQRWAARYPPGPLPLPGLGNL...   \n",
       "1      MFRRLTFAQLLFATVLGIAGGVYIFQPVFEQYAKDQKELKEKMQLV...   \n",
       "2                                       MRWQEMGYIFYPRKLR   \n",
       "3      MTQRAGAAMLPSALLLLCVPGCLTVSGPSTVMGAVGESLSVQCRYE...   \n",
       "4      MGDQPCASGRSTLPPGNAREAKPPKKRCLLAPRWDYPEGTPNGGST...   \n",
       "...                                                  ...   \n",
       "25114  MKTSPRLSVIDGVGEEVRPDLFQLVHVVPPNTCEPPEQWLRVTVLL...   \n",
       "25115  MGGAYITSRYWLVPAHVTSRLRLTSFTGHRGVARVDGPMSSLGLNI...   \n",
       "25116  MPYVDFWELSRDFLHFASGHHGSNSSVERAVTVSRHNAQGEYNVPL...   \n",
       "25117  MNNGQATITRNGGRFEIRCRHLDRDYTMPLPNATSNDNFLDCIKFI...   \n",
       "25118  MGRKEHESPSQPHMCGWEDSQKPSVPSHGPKTPSCKGVKAPHSSRP...   \n",
       "\n",
       "                                           Protein names  \\\n",
       "0            Putative cytochrome P450 2D7 (EC 1.14.14.1)   \n",
       "1       Protein PIGBOS1 (PIGB opposite strand protein 1)   \n",
       "2      Mitochondrial-derived peptide MOTS-c (Mitochon...   \n",
       "3      Protein CD300H (CD300 antigen-like family memb...   \n",
       "4      Negative regulator of P-body association (P-bo...   \n",
       "...                                                  ...   \n",
       "25114                       Uncharacterized protein ORF6   \n",
       "25115                       Uncharacterized protein ORF5   \n",
       "25116                       Uncharacterized protein ORF3   \n",
       "25117                     Non-structural protein 3 (NS3)   \n",
       "25118  Putative uncharacterized protein GUCA1ANB (GUC...   \n",
       "\n",
       "                                                Organism   Mass  \\\n",
       "0                                   Homo sapiens (Human)  57489   \n",
       "1                                   Homo sapiens (Human)   6313   \n",
       "2                                   Homo sapiens (Human)   2175   \n",
       "3                                   Homo sapiens (Human)  21806   \n",
       "4                                   Homo sapiens (Human)   7025   \n",
       "...                                                  ...    ...   \n",
       "25114              Beak and feather disease virus (BFDV)   9846   \n",
       "25115              Beak and feather disease virus (BFDV)  10863   \n",
       "25116              Beak and feather disease virus (BFDV)  17738   \n",
       "25117  Banna virus (strain Indonesia/JKT-6423/1980) (...  34991   \n",
       "25118                               Homo sapiens (Human)  14531   \n",
       "\n",
       "                                                Keywords   av_mw  \\\n",
       "0      Cytoplasm . Glycoprotein . Heme . Iron . Membr...  111.63   \n",
       "1      Direct protein sequencing . Membrane . Mitocho...  116.91   \n",
       "2      DNA-binding . Mitochondrion . Nucleus . Osteog...  135.94   \n",
       "3      Alternative splicing . Disulfide bond . Glycop...  108.49   \n",
       "4       Cytoplasm . mRNA processing . Reference proteome  103.31   \n",
       "...                                                  ...     ...   \n",
       "25114                                                NaN  111.89   \n",
       "25115                                                NaN  107.55   \n",
       "25116                                                NaN  110.86   \n",
       "25117                                 Reference proteome  114.35   \n",
       "25118                                 Reference proteome  114.42   \n",
       "\n",
       "                                             Virus hosts  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                                                    NaN   \n",
       "...                                                  ...   \n",
       "25114  Gracula [TaxID: 116991]; Psittaciformes [TaxID...   \n",
       "25115  Gracula [TaxID: 116991]; Psittaciformes [TaxID...   \n",
       "25116  Gracula [TaxID: 116991]; Psittaciformes [TaxID...   \n",
       "25117  Bos taurus (Bovine) [TaxID: 9913]; Culex annul...   \n",
       "25118                                                NaN   \n",
       "\n",
       "                                       Taxonomic lineage UR50_Cluster ID  \\\n",
       "0      cellular organisms, Eukaryota, Opisthokonta, M...             NaN   \n",
       "1      cellular organisms, Eukaryota, Opisthokonta, M...      A0A0B4J2F0   \n",
       "2      cellular organisms, Eukaryota, Opisthokonta, M...      A0A0C5B5G6   \n",
       "3      cellular organisms, Eukaryota, Opisthokonta, M...      A0A0K2S4Q6   \n",
       "4      cellular organisms, Eukaryota, Opisthokonta, M...      A0A0U1RRE5   \n",
       "...                                                  ...             ...   \n",
       "25114  Viruses, Monodnaviria, Shotokuvirae (kingdom),...          Q9YUD1   \n",
       "25115  Viruses, Monodnaviria, Shotokuvirae (kingdom),...          Q9YUD2   \n",
       "25116  Viruses, Monodnaviria, Shotokuvirae (kingdom),...          Q9YUD4   \n",
       "25117  Viruses, Riboviria, Orthornavirae (kingdom), D...          Q9YWQ0   \n",
       "25118  cellular organisms, Eukaryota, Opisthokonta, M...          X6R8D5   \n",
       "\n",
       "                                             group  \n",
       "0      Putative cytochrome P450 2D7 (EC 1.14.14.1)  \n",
       "1                                       A0A0B4J2F0  \n",
       "2                                       A0A0C5B5G6  \n",
       "3                                       A0A0K2S4Q6  \n",
       "4                                       A0A0U1RRE5  \n",
       "...                                            ...  \n",
       "25114                                       Q9YUD1  \n",
       "25115                                       Q9YUD2  \n",
       "25116                                       Q9YUD4  \n",
       "25117                                       Q9YWQ0  \n",
       "25118                                       X6R8D5  \n",
       "\n",
       "[25119 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21394"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta['Protein names'].str.split(\"_\",expand=True)[0].str.strip().nunique() # 21394 - same as protein name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20540"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta[\"Entry Name\"].str.split(\"_\",expand=True)[0].nunique() # 20540"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19217"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta.loc[df_meta[\"Sequence\"].isin(df_all[\"Sequence\"])].shape[0] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19606, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>virus</th>\n",
       "      <th>Length</th>\n",
       "      <th>Cluster name</th>\n",
       "      <th>mistake</th>\n",
       "      <th>pred_score_1</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Mass</th>\n",
       "      <th>Keywords</th>\n",
       "      <th>Virus hosts</th>\n",
       "      <th>Taxonomic lineage</th>\n",
       "      <th>av_mw</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MADFLKGLPVYNKSNFSRFHADSVCKASNRRPSVYLPTREYPSEQI...</td>\n",
       "      <td>0</td>\n",
       "      <td>102</td>\n",
       "      <td>DET1- and DDB1-associated protein 1</td>\n",
       "      <td>False</td>\n",
       "      <td>0.232</td>\n",
       "      <td>DET1- and DDB1-associated protein 1 (Placenta ...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>11835</td>\n",
       "      <td>3D-structure . Acetylation . Phosphoprotein . ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms , Eukaryota , Opisthokonta ...</td>\n",
       "      <td>116.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MPSTLQVLAKKVLALEHKENDHISREYYYHILKCCGLWWHEAPIIL...</td>\n",
       "      <td>1</td>\n",
       "      <td>362</td>\n",
       "      <td>Protein MGF 360-19R</td>\n",
       "      <td>False</td>\n",
       "      <td>0.942</td>\n",
       "      <td>Protein MGF 360-19R</td>\n",
       "      <td>African swine fever virus (isolate Tick/South ...</td>\n",
       "      <td>42605</td>\n",
       "      <td>ANK repeat</td>\n",
       "      <td>Ornithodoros (relapsing fever ticks) [TaxID: 6...</td>\n",
       "      <td>Viruses , Varidnaviria , Bamfordvirae (kingdom...</td>\n",
       "      <td>117.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MASSAELDFNLQALLEQLSQDELSKFKSLIRTISLGKELQTVPQTE...</td>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>Pyrin domain-containing protein 2</td>\n",
       "      <td>False</td>\n",
       "      <td>0.134</td>\n",
       "      <td>Pyrin domain-containing protein 2 (Pyrin-only ...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>10795</td>\n",
       "      <td>Cytoplasm . Immunity . Inflammatory response ....</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms , Eukaryota , Opisthokonta ...</td>\n",
       "      <td>111.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MAAWGKKHAGKDPVRDECEERNRFTETREEDVTDEHGEREPFAETD...</td>\n",
       "      <td>0</td>\n",
       "      <td>186</td>\n",
       "      <td>Protein FAM9B</td>\n",
       "      <td>False</td>\n",
       "      <td>0.153</td>\n",
       "      <td>Protein FAM9B</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>22438</td>\n",
       "      <td>Alternative splicing . Nucleus . Reference pro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms , Eukaryota , Opisthokonta ...</td>\n",
       "      <td>120.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MADPAAAQNYPLLKLLHSYTPTTPPRPIPKPAPWAPQKPRRQITND...</td>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "      <td>Protein E4</td>\n",
       "      <td>False</td>\n",
       "      <td>0.533</td>\n",
       "      <td>Protein E4</td>\n",
       "      <td>Human papillomavirus 35</td>\n",
       "      <td>10089</td>\n",
       "      <td>Early protein . Host cytoplasm . Host G2/M cel...</td>\n",
       "      <td>Homo sapiens (Human) [TaxID: 9606]</td>\n",
       "      <td>Viruses , Monodnaviria , Shotokuvirae (kingdom...</td>\n",
       "      <td>108.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19601</th>\n",
       "      <td>MAAGDGDVKLGTLGSGSESSNDGGSESPGDAGAAAEGGGWAAAALA...</td>\n",
       "      <td>0</td>\n",
       "      <td>223</td>\n",
       "      <td>Membrane-associated progesterone receptor comp...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.041</td>\n",
       "      <td>Membrane-associated progesterone receptor comp...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>23818</td>\n",
       "      <td>Alternative splicing . Endoplasmic reticulum ....</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms , Eukaryota , Opisthokonta ...</td>\n",
       "      <td>106.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19602</th>\n",
       "      <td>MTLPASLKGFETLDIFNISSVRNICELNKHAEHVSIKNLPLIDISV...</td>\n",
       "      <td>1</td>\n",
       "      <td>256</td>\n",
       "      <td>Cytoplasmic envelopment protein 1</td>\n",
       "      <td>False</td>\n",
       "      <td>0.916</td>\n",
       "      <td>Cytoplasmic envelopment protein 1</td>\n",
       "      <td>Human herpesvirus 7 (strain JI) (HHV-7) (Human...</td>\n",
       "      <td>29375</td>\n",
       "      <td>Host cytoplasm . Host Golgi apparatus . Refere...</td>\n",
       "      <td>Homo sapiens (Human) [TaxID: 9606]</td>\n",
       "      <td>Viruses , Duplodnaviria , Heunggongvirae (king...</td>\n",
       "      <td>114.75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19603</th>\n",
       "      <td>MKWKALFTAAILQAQLPITEAQSFGLLDPKLCYLLDGILFIYGVIL...</td>\n",
       "      <td>0</td>\n",
       "      <td>164</td>\n",
       "      <td>T-cell surface glycoprotein CD3 zeta chain</td>\n",
       "      <td>False</td>\n",
       "      <td>0.089</td>\n",
       "      <td>T-cell surface glycoprotein CD3 zeta chain (T-...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>18696</td>\n",
       "      <td>3D-structure . Adaptive immunity . Alternative...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms , Eukaryota , Opisthokonta ...</td>\n",
       "      <td>114.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19604</th>\n",
       "      <td>MAGVGDAAAPGEGGGGGVDGPQRDGRGEAEQPGGSGGQGPPPAPQL...</td>\n",
       "      <td>0</td>\n",
       "      <td>567</td>\n",
       "      <td>Transmembrane anterior posterior transformatio...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.047</td>\n",
       "      <td>Transmembrane anterior posterior transformatio...</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>64260</td>\n",
       "      <td>Acetylation . Alternative splicing . Cell proj...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms , Eukaryota , Opisthokonta ...</td>\n",
       "      <td>113.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19605</th>\n",
       "      <td>MGASVSRGRAARVPAPEPEPEEALDLSQLPPELLLVVLSHVPPRTL...</td>\n",
       "      <td>0</td>\n",
       "      <td>283</td>\n",
       "      <td>F-box only protein 27</td>\n",
       "      <td>False</td>\n",
       "      <td>0.141</td>\n",
       "      <td>F-box only protein 27 (F-box/G-domain protein 5)</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>31623</td>\n",
       "      <td>Reference proteome . Ubl conjugation pathway</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cellular organisms , Eukaryota , Opisthokonta ...</td>\n",
       "      <td>111.74</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19606 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Sequence  virus  Length  \\\n",
       "0      MADFLKGLPVYNKSNFSRFHADSVCKASNRRPSVYLPTREYPSEQI...      0     102   \n",
       "1      MPSTLQVLAKKVLALEHKENDHISREYYYHILKCCGLWWHEAPIIL...      1     362   \n",
       "2      MASSAELDFNLQALLEQLSQDELSKFKSLIRTISLGKELQTVPQTE...      0      97   \n",
       "3      MAAWGKKHAGKDPVRDECEERNRFTETREEDVTDEHGEREPFAETD...      0     186   \n",
       "4      MADPAAAQNYPLLKLLHSYTPTTPPRPIPKPAPWAPQKPRRQITND...      1      93   \n",
       "...                                                  ...    ...     ...   \n",
       "19601  MAAGDGDVKLGTLGSGSESSNDGGSESPGDAGAAAEGGGWAAAALA...      0     223   \n",
       "19602  MTLPASLKGFETLDIFNISSVRNICELNKHAEHVSIKNLPLIDISV...      1     256   \n",
       "19603  MKWKALFTAAILQAQLPITEAQSFGLLDPKLCYLLDGILFIYGVIL...      0     164   \n",
       "19604  MAGVGDAAAPGEGGGGGVDGPQRDGRGEAEQPGGSGGQGPPPAPQL...      0     567   \n",
       "19605  MGASVSRGRAARVPAPEPEPEEALDLSQLPPELLLVVLSHVPPRTL...      0     283   \n",
       "\n",
       "                                            Cluster name  mistake  \\\n",
       "0                    DET1- and DDB1-associated protein 1    False   \n",
       "1                                    Protein MGF 360-19R    False   \n",
       "2                      Pyrin domain-containing protein 2    False   \n",
       "3                                          Protein FAM9B    False   \n",
       "4                                             Protein E4    False   \n",
       "...                                                  ...      ...   \n",
       "19601  Membrane-associated progesterone receptor comp...    False   \n",
       "19602                  Cytoplasmic envelopment protein 1    False   \n",
       "19603         T-cell surface glycoprotein CD3 zeta chain    False   \n",
       "19604  Transmembrane anterior posterior transformatio...    False   \n",
       "19605                              F-box only protein 27    False   \n",
       "\n",
       "       pred_score_1                                      Protein names  \\\n",
       "0             0.232  DET1- and DDB1-associated protein 1 (Placenta ...   \n",
       "1             0.942                                Protein MGF 360-19R   \n",
       "2             0.134  Pyrin domain-containing protein 2 (Pyrin-only ...   \n",
       "3             0.153                                      Protein FAM9B   \n",
       "4             0.533                                         Protein E4   \n",
       "...             ...                                                ...   \n",
       "19601         0.041  Membrane-associated progesterone receptor comp...   \n",
       "19602         0.916                  Cytoplasmic envelopment protein 1   \n",
       "19603         0.089  T-cell surface glycoprotein CD3 zeta chain (T-...   \n",
       "19604         0.047  Transmembrane anterior posterior transformatio...   \n",
       "19605         0.141   F-box only protein 27 (F-box/G-domain protein 5)   \n",
       "\n",
       "                                                Organism   Mass  \\\n",
       "0                                   Homo sapiens (Human)  11835   \n",
       "1      African swine fever virus (isolate Tick/South ...  42605   \n",
       "2                                   Homo sapiens (Human)  10795   \n",
       "3                                   Homo sapiens (Human)  22438   \n",
       "4                                Human papillomavirus 35  10089   \n",
       "...                                                  ...    ...   \n",
       "19601                               Homo sapiens (Human)  23818   \n",
       "19602  Human herpesvirus 7 (strain JI) (HHV-7) (Human...  29375   \n",
       "19603                               Homo sapiens (Human)  18696   \n",
       "19604                               Homo sapiens (Human)  64260   \n",
       "19605                               Homo sapiens (Human)  31623   \n",
       "\n",
       "                                                Keywords  \\\n",
       "0      3D-structure . Acetylation . Phosphoprotein . ...   \n",
       "1                                             ANK repeat   \n",
       "2      Cytoplasm . Immunity . Inflammatory response ....   \n",
       "3      Alternative splicing . Nucleus . Reference pro...   \n",
       "4      Early protein . Host cytoplasm . Host G2/M cel...   \n",
       "...                                                  ...   \n",
       "19601  Alternative splicing . Endoplasmic reticulum ....   \n",
       "19602  Host cytoplasm . Host Golgi apparatus . Refere...   \n",
       "19603  3D-structure . Adaptive immunity . Alternative...   \n",
       "19604  Acetylation . Alternative splicing . Cell proj...   \n",
       "19605       Reference proteome . Ubl conjugation pathway   \n",
       "\n",
       "                                             Virus hosts  \\\n",
       "0                                                    NaN   \n",
       "1      Ornithodoros (relapsing fever ticks) [TaxID: 6...   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                     Homo sapiens (Human) [TaxID: 9606]   \n",
       "...                                                  ...   \n",
       "19601                                                NaN   \n",
       "19602                 Homo sapiens (Human) [TaxID: 9606]   \n",
       "19603                                                NaN   \n",
       "19604                                                NaN   \n",
       "19605                                                NaN   \n",
       "\n",
       "                                       Taxonomic lineage   av_mw  \n",
       "0      cellular organisms , Eukaryota , Opisthokonta ...  116.03  \n",
       "1      Viruses , Varidnaviria , Bamfordvirae (kingdom...  117.69  \n",
       "2      cellular organisms , Eukaryota , Opisthokonta ...  111.29  \n",
       "3      cellular organisms , Eukaryota , Opisthokonta ...  120.63  \n",
       "4      Viruses , Monodnaviria , Shotokuvirae (kingdom...  108.48  \n",
       "...                                                  ...     ...  \n",
       "19601  cellular organisms , Eukaryota , Opisthokonta ...  106.81  \n",
       "19602  Viruses , Duplodnaviria , Heunggongvirae (king...  114.75  \n",
       "19603  cellular organisms , Eukaryota , Opisthokonta ...  114.00  \n",
       "19604  cellular organisms , Eukaryota , Opisthokonta ...  113.33  \n",
       "19605  cellular organisms , Eukaryota , Opisthokonta ...  111.74  \n",
       "\n",
       "[19606 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_merged = df_all.merge(df_meta,on=\"Sequence\",how=\"left\")\n",
    "# df_merged[\"av_mw\"] = df_merged[\"Mass\"].div(df_merged[\"Length\"]).round(2)\n",
    "# df_merged[\"Keywords\"] = df_merged[\"Keywords\"].str.replace(\";\",\" . \").str.strip()\n",
    "# df_merged[\"Taxonomic lineage\"] = df_merged[\"Taxonomic lineage\"].str.replace(\"(no rank)\",\"\").str.replace(\"(superkingdom)\",\"\").str.strip()\n",
    "print(df_merged.shape)\n",
    "display(df_merged)\n",
    "df_merged.to_csv(\"data_with_mistakes_features_v1.csv.gz\",index=False,compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
